{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Bad key text.latex.unicode in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 112 ('text.latex.unicode : False # use \"ucs\" and \"inputenc\" LaTeX packages for handling')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key savefig.frameon in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 423 ('savefig.frameon : True')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key pgf.debug in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 444 ('pgf.debug           : False')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key verbose.level in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 475 ('verbose.level  : silent      # one of silent, helpful, debug, debug-annoying')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key verbose.fileo in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 476 ('verbose.fileo  : sys.stdout  # a log filename, sys.stdout or sys.stderr')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The text.latex.preview rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The mathtext.fallback_to_cm rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: Support for setting the 'mathtext.fallback_to_cm' rcParam is deprecated since 3.3 and will be removed two minor releases later; use 'mathtext.fallback : 'cm' instead.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The validate_bool_maybe_none function was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The savefig.jpeg_quality rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The keymap.all_axes rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The animation.avconv_path rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The animation.avconv_args rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "from collections import OrderedDict, Counter, defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import glob\n",
    "\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "\n",
    "import re\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "import pandas\n",
    "\n",
    "import pickle\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import cdist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "390\n"
     ]
    }
   ],
   "source": [
    "# open SET1 (PhyloSET) and SET2 (RiboSET)\n",
    "PhyloSET = pd.read_csv('../../tmp_res/SET1.txt', sep='\\t')\n",
    "\n",
    "\n",
    "\n",
    "RiboSET = pd.read_csv('../../tmp_res/SET2.txt', sep='\\t')\n",
    "\n",
    "RiboSET['tr_id1'] = [x.split('.')[0] for x in RiboSET['tr_id'].tolist()]\n",
    "\n",
    "RiboSET = RiboSET[(~RiboSET['Riboseq_Summary'].str.contains('ATG')) & \n",
    "                  (~RiboSET['tr_id'].str.contains('PAR'))]\n",
    "\n",
    "print (RiboSET.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_tr</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>cds_seq</th>\n",
       "      <th>cds_start_codon</th>\n",
       "      <th>cds_stop_codon</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>cds_stop_pos</th>\n",
       "      <th>utr5_start</th>\n",
       "      <th>utr5_end</th>\n",
       "      <th>record_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id   gene    gene_tr  \\\n",
       "0  ENST00000335137.3  OR4F5  OR4F5-001   \n",
       "\n",
       "                                      transcript_seq 5UTR_start_seq CDS_start  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...            NaN       ATG   \n",
       "\n",
       "                                             cds_seq cds_start_codon  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...             ATG   \n",
       "\n",
       "  cds_stop_codon  cds_start_pos  cds_stop_pos  utr5_start  utr5_end  \\\n",
       "0            TAG              0           918           0         0   \n",
       "\n",
       "                                           record_id  \n",
       "0  ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pc_g25 = pd.read_csv('../../tmp_res/metadata_pc_g25.txt', sep='\\t')\n",
    "metadata_pc_g25[0:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>...</th>\n",
       "      <th>max_sum_overlap_refseq_all_exons</th>\n",
       "      <th>max_sum_overlap_refseq_all_exons_strandless</th>\n",
       "      <th>ovlp</th>\n",
       "      <th>ovlp_strandless</th>\n",
       "      <th>ovlp2</th>\n",
       "      <th>ovlp2_strandless</th>\n",
       "      <th>PhyloCSF120score</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>number_of_records</th>\n",
       "      <th>records_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "      <td>-163.9926</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>85</td>\n",
       "      <td>Human; Gorilla; Orangutan; Golden_snub_nosed_m...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag global_coo_50_and_less  \\\n",
       "0          316        14.0  FO538757.2         274     chr1:182667-182708   \n",
       "\n",
       "   ... max_sum_overlap_refseq_all_exons  \\\n",
       "0  ...                                0   \n",
       "\n",
       "  max_sum_overlap_refseq_all_exons_strandless   ovlp ovlp_strandless    ovlp2  \\\n",
       "0                                           0  0;0;0           0;0;0  0;0;0;0   \n",
       "\n",
       "   ovlp2_strandless  PhyloCSF120score           tr_id1  number_of_records  \\\n",
       "0           0;0;0;0         -163.9926  ENST00000624431                 85   \n",
       "\n",
       "                                       records_names  \n",
       "0  Human; Gorilla; Orangutan; Golden_snub_nosed_m...  \n",
       "\n",
       "[1 rows x 30 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# open metadata with scores and metrics\n",
    "meta = pd.read_csv('../../tmp_res/local_and_global_df_g25ovlp_g35ovlp_g38ovlp_refseqovlp_strand_FRAMES_PhyloCSF.txt', sep='\\t')\n",
    "meta[0:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trips-viz pred genes: exclude them "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alla/.local/lib/python3.7/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  # This is added back by InteractiveShellApp.init_path()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3722\n",
      "3530\n",
      "final shape of RiboSET_EXT 3451 min CDS cov: 0.060654429369513166\n",
      "390\n"
     ]
    }
   ],
   "source": [
    "transl_with_all_codons = pd.read_csv('../../data/extensions_TripsViz_1perGene_AUG_NearCognates.txt', sep=',')\n",
    "transl_with_all_codons.columns = ['gene_ext', 'tr_id1', 'Start_ext', 'Stop_ext', 'len_ext', \n",
    "                                 'global_rank_ext', 'type_all_codons', 'trips_viz_link', 'Start_codon_ext', \n",
    "                                 'Highframe rank', 'Highframe value',\n",
    "       'Lowframe rank', 'Lowframe value', 'Stop rank', 'Stop value',\n",
    "       'Start rank', 'Start value', 'Coverage rank', 'Coverage_value_ext',\n",
    "       'Amino acid sequence', 'Proteomics_count_ext', 'CDS_ratio']\n",
    "transl_with_all_codons2 = transl_with_all_codons[['gene_ext', 'tr_id1', 'Start_ext', 'Stop_ext', \n",
    "                                                 'global_rank_ext', 'Start_codon_ext', \n",
    "                                                 'Proteomics_count_ext', 'Coverage_value_ext', 'CDS_ratio']]\n",
    "transl_with_all_codons2['Riboseq_Summary'] = transl_with_all_codons2['Start_codon_ext']+'; '+transl_with_all_codons2['Start_ext'].astype(str)+'-'+transl_with_all_codons2['Stop_ext'].astype(str)+'; Rank: '+transl_with_all_codons2['global_rank_ext'].astype(str)+'; cov: '+np.round(transl_with_all_codons2['Coverage_value_ext'],2).astype(str)\n",
    "\n",
    "top5000 = transl_with_all_codons2[0:5000]\n",
    "\n",
    "top5000_meta = top5000.merge(meta, on='tr_id1', how='inner')\n",
    "\n",
    "top5000_meta['len_codons'] = (top5000_meta['Stop_ext'] - top5000_meta['Start_ext']) // 3\n",
    "\n",
    "# theoretical extensions should be no less than 20 codons (works for initiatial filter and for PhyloSET)\n",
    "# but in practice it could be shorter\n",
    "# let's say 10aa\n",
    "\n",
    "# select 0 ovlps and where theoretical extension is at least 20 codons  \n",
    "RiboSET_EXT = top5000_meta[(top5000_meta['ovlp'] == '0;0;0') & \n",
    "                                  (top5000_meta['Start_codon_ext'] != 'ATG') & \n",
    "                                  (top5000_meta['tr_id1'].isin(meta[meta['len_codons'] >= 20]['tr_id1'].tolist()))]\n",
    "\n",
    "print (RiboSET_EXT.shape[0])\n",
    "\n",
    "# ~4k RiboSET ext with non-zero CDS cov in my data \n",
    "RiboSET_EXT_cds_cov_nonzero = pd.read_csv('../tmp_res/RiboSET_EXT_cds_cov_nonzero_theor_len_20codons_3663genes.txt', sep='\\t')\n",
    "RiboSET_EXT_cds_cov_nonzero = RiboSET_EXT_cds_cov_nonzero.merge(RiboSET_EXT, on='tr_id1', how='inner')\n",
    "RiboSET_EXT_cds_cov_nonzero = RiboSET_EXT_cds_cov_nonzero[RiboSET_EXT_cds_cov_nonzero['NTE_coverage'] > 0]\n",
    "RiboSET_EXT_cds_cov_nonzero = RiboSET_EXT_cds_cov_nonzero[~RiboSET_EXT_cds_cov_nonzero['tr_id'].str.contains('PAR')]\n",
    "RiboSET_EXT_cds_cov_nonzero.to_csv('../tmp_res/FINAL_RiboSET_EXT_cds_cov_nonzero_theor_len_20codons_3532genes.txt', \n",
    "                                  sep='\\t', index=False)\n",
    "\n",
    "print (RiboSET_EXT_cds_cov_nonzero.shape[0])\n",
    "\n",
    "RiboSET_EXT_cds_cov_nonzero = RiboSET_EXT_cds_cov_nonzero[RiboSET_EXT_cds_cov_nonzero['CDS_coverage'] > 0.06]\n",
    "\n",
    "RiboSET_EXT_cds_cov_nonzero.to_csv('../tmp_res/FINAL_RiboSET_EXT_cds_cov_nonzero_theor_len_20codons_3532genes.txt', \n",
    "                                  sep='\\t', index=False)\n",
    "\n",
    "print ('final shape of RiboSET_EXT', RiboSET_EXT_cds_cov_nonzero.shape[0], 'min CDS cov:', RiboSET_EXT_cds_cov_nonzero['CDS_coverage'].min())\n",
    "\n",
    "print (len(set(RiboSET_EXT_cds_cov_nonzero['gene'].tolist()).intersection(RiboSET['gene'].tolist())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NTE_nonzero_fraction</th>\n",
       "      <th>global_rank_ext</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1335</th>\n",
       "      <td>0.200000</td>\n",
       "      <td>1786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>0.319444</td>\n",
       "      <td>2030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2075</th>\n",
       "      <td>0.307018</td>\n",
       "      <td>2803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2078</th>\n",
       "      <td>0.120000</td>\n",
       "      <td>2807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2126</th>\n",
       "      <td>0.238095</td>\n",
       "      <td>2871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3621</th>\n",
       "      <td>0.031746</td>\n",
       "      <td>4904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3623</th>\n",
       "      <td>0.015152</td>\n",
       "      <td>4910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3640</th>\n",
       "      <td>0.009009</td>\n",
       "      <td>4942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3646</th>\n",
       "      <td>0.009524</td>\n",
       "      <td>4956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3656</th>\n",
       "      <td>0.063492</td>\n",
       "      <td>4977</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      NTE_nonzero_fraction  global_rank_ext\n",
       "1335              0.200000             1786\n",
       "1506              0.319444             2030\n",
       "2075              0.307018             2803\n",
       "2078              0.120000             2807\n",
       "2126              0.238095             2871\n",
       "...                    ...              ...\n",
       "3621              0.031746             4904\n",
       "3623              0.015152             4910\n",
       "3640              0.009009             4942\n",
       "3646              0.009524             4956\n",
       "3656              0.063492             4977\n",
       "\n",
       "[80 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RiboSET_EXT_cds_cov_nonzero[RiboSET_EXT_cds_cov_nonzero['Coverage_value_ext'] == 0][['NTE_nonzero_fraction', 'global_rank_ext']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alla/.local/lib/python3.7/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ext</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>Start_ext</th>\n",
       "      <th>Stop_ext</th>\n",
       "      <th>global_rank_ext</th>\n",
       "      <th>Start_codon_ext</th>\n",
       "      <th>Proteomics_count_ext</th>\n",
       "      <th>Coverage_value_ext</th>\n",
       "      <th>CDS_ratio</th>\n",
       "      <th>Riboseq_Summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASPH</td>\n",
       "      <td>ENST00000379454</td>\n",
       "      <td>144</td>\n",
       "      <td>189</td>\n",
       "      <td>1</td>\n",
       "      <td>CTG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.645069</td>\n",
       "      <td>CTG; 144-189; Rank: 1; cov: 100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HSP90AA1</td>\n",
       "      <td>ENST00000216281</td>\n",
       "      <td>168</td>\n",
       "      <td>207</td>\n",
       "      <td>2</td>\n",
       "      <td>GTG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.081078</td>\n",
       "      <td>GTG; 168-207; Rank: 2; cov: 100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CD24</td>\n",
       "      <td>ENST00000619133</td>\n",
       "      <td>408</td>\n",
       "      <td>441</td>\n",
       "      <td>3</td>\n",
       "      <td>CTG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>83.333333</td>\n",
       "      <td>11.578498</td>\n",
       "      <td>CTG; 408-441; Rank: 3; cov: 83.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HN1</td>\n",
       "      <td>ENST00000409753</td>\n",
       "      <td>242</td>\n",
       "      <td>287</td>\n",
       "      <td>4</td>\n",
       "      <td>TTG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.588389</td>\n",
       "      <td>TTG; 242-287; Rank: 4; cov: 100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YBX1</td>\n",
       "      <td>ENST00000321358</td>\n",
       "      <td>44</td>\n",
       "      <td>140</td>\n",
       "      <td>5</td>\n",
       "      <td>CTG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>85.185185</td>\n",
       "      <td>7.709763</td>\n",
       "      <td>CTG; 44-140; Rank: 5; cov: 85.19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ext           tr_id1  Start_ext  Stop_ext  global_rank_ext  \\\n",
       "0      ASPH  ENST00000379454        144       189                1   \n",
       "1  HSP90AA1  ENST00000216281        168       207                2   \n",
       "2      CD24  ENST00000619133        408       441                3   \n",
       "3       HN1  ENST00000409753        242       287                4   \n",
       "4      YBX1  ENST00000321358         44       140                5   \n",
       "\n",
       "  Start_codon_ext  Proteomics_count_ext  Coverage_value_ext  CDS_ratio  \\\n",
       "0             CTG                   0.0          100.000000   1.645069   \n",
       "1             GTG                   0.0          100.000000   3.081078   \n",
       "2             CTG                   0.0           83.333333  11.578498   \n",
       "3             TTG                   0.0          100.000000   0.588389   \n",
       "4             CTG                   0.0           85.185185   7.709763   \n",
       "\n",
       "                     Riboseq_Summary  \n",
       "0  CTG; 144-189; Rank: 1; cov: 100.0  \n",
       "1  GTG; 168-207; Rank: 2; cov: 100.0  \n",
       "2  CTG; 408-441; Rank: 3; cov: 83.33  \n",
       "3  TTG; 242-287; Rank: 4; cov: 100.0  \n",
       "4   CTG; 44-140; Rank: 5; cov: 85.19  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transl_with_all_codons = pd.read_csv('../../data/extensions_TripsViz_1perGene_AUG_NearCognates.txt', sep=',')\n",
    "\n",
    "transl_with_all_codons.columns = ['gene_ext', 'tr_id1', 'Start_ext', 'Stop_ext', 'len_ext', \n",
    "                                 'global_rank_ext', 'type_all_codons', 'trips_viz_link', 'Start_codon_ext', \n",
    "                                 'Highframe rank', 'Highframe value',\n",
    "       'Lowframe rank', 'Lowframe value', 'Stop rank', 'Stop value',\n",
    "       'Start rank', 'Start value', 'Coverage rank', 'Coverage_value_ext',\n",
    "       'Amino acid sequence', 'Proteomics_count_ext', 'CDS_ratio']\n",
    "\n",
    "transl_with_all_codons2 = transl_with_all_codons[['gene_ext', 'tr_id1', 'Start_ext', 'Stop_ext', \n",
    "                                                 'global_rank_ext', 'Start_codon_ext', \n",
    "                                                 'Proteomics_count_ext', 'Coverage_value_ext', 'CDS_ratio']]\n",
    "\n",
    "transl_with_all_codons2['Riboseq_Summary'] = transl_with_all_codons2['Start_codon_ext']+'; '+transl_with_all_codons2['Start_ext'].astype(str)+'-'+transl_with_all_codons2['Stop_ext'].astype(str)+'; Rank: '+transl_with_all_codons2['global_rank_ext'].astype(str)+'; cov: '+np.round(transl_with_all_codons2['Coverage_value_ext'],2).astype(str)\n",
    "\n",
    "\n",
    "transl_with_all_codons2[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ext</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>Start_ext</th>\n",
       "      <th>Stop_ext</th>\n",
       "      <th>len_ext</th>\n",
       "      <th>global_rank_ext</th>\n",
       "      <th>type_all_codons</th>\n",
       "      <th>trips_viz_link</th>\n",
       "      <th>Start_codon_ext</th>\n",
       "      <th>Highframe rank</th>\n",
       "      <th>...</th>\n",
       "      <th>Lowframe value</th>\n",
       "      <th>Stop rank</th>\n",
       "      <th>Stop value</th>\n",
       "      <th>Start rank</th>\n",
       "      <th>Start value</th>\n",
       "      <th>Coverage rank</th>\n",
       "      <th>Coverage_value_ext</th>\n",
       "      <th>Amino acid sequence</th>\n",
       "      <th>Proteomics_count_ext</th>\n",
       "      <th>CDS_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1785</th>\n",
       "      <td>TPM4</td>\n",
       "      <td>ENST00000344824</td>\n",
       "      <td>59</td>\n",
       "      <td>119</td>\n",
       "      <td>61</td>\n",
       "      <td>1786</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>CTG</td>\n",
       "      <td>749</td>\n",
       "      <td>...</td>\n",
       "      <td>-10</td>\n",
       "      <td>437</td>\n",
       "      <td>-240</td>\n",
       "      <td>212</td>\n",
       "      <td>1356</td>\n",
       "      <td>587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LPQPPQSPPRTPRPPRQRP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8498</th>\n",
       "      <td>DNAJC7</td>\n",
       "      <td>ENST00000457167</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8499</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>488</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>211</td>\n",
       "      <td>-1</td>\n",
       "      <td>857</td>\n",
       "      <td>-1</td>\n",
       "      <td>587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8497</th>\n",
       "      <td>UBXN4</td>\n",
       "      <td>ENST00000272638</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8498</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>488</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>211</td>\n",
       "      <td>-1</td>\n",
       "      <td>857</td>\n",
       "      <td>-1</td>\n",
       "      <td>587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8496</th>\n",
       "      <td>SUN2</td>\n",
       "      <td>ENST00000405510</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8497</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>488</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>211</td>\n",
       "      <td>-1</td>\n",
       "      <td>857</td>\n",
       "      <td>-1</td>\n",
       "      <td>587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8495</th>\n",
       "      <td>SLC36A1</td>\n",
       "      <td>ENST00000243389</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8496</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>488</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>211</td>\n",
       "      <td>-1</td>\n",
       "      <td>857</td>\n",
       "      <td>-1</td>\n",
       "      <td>587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4546</th>\n",
       "      <td>RAB40A</td>\n",
       "      <td>ENST00000372633</td>\n",
       "      <td>1985</td>\n",
       "      <td>2120</td>\n",
       "      <td>136</td>\n",
       "      <td>4547</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>ATT</td>\n",
       "      <td>487</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>855</td>\n",
       "      <td>1</td>\n",
       "      <td>538</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ISLYFTDFPISFFMFYANFSRRTGPAPPLRTTPRAWLRRECGAS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4626</th>\n",
       "      <td>ST8SIA3</td>\n",
       "      <td>ENST00000324000</td>\n",
       "      <td>1960</td>\n",
       "      <td>2035</td>\n",
       "      <td>76</td>\n",
       "      <td>4627</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>ATA</td>\n",
       "      <td>495</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>855</td>\n",
       "      <td>1</td>\n",
       "      <td>538</td>\n",
       "      <td>5.0</td>\n",
       "      <td>IYTPASCWPLNGPISPVSLNPAQP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4536</th>\n",
       "      <td>NECAB1</td>\n",
       "      <td>ENST00000417640</td>\n",
       "      <td>263</td>\n",
       "      <td>338</td>\n",
       "      <td>76</td>\n",
       "      <td>4537</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>CTG</td>\n",
       "      <td>487</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>854</td>\n",
       "      <td>2</td>\n",
       "      <td>538</td>\n",
       "      <td>5.0</td>\n",
       "      <td>LRPRPCQSRCVRLAPLRLRPSGLP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4557</th>\n",
       "      <td>BHMG1</td>\n",
       "      <td>ENST00000457052</td>\n",
       "      <td>342</td>\n",
       "      <td>417</td>\n",
       "      <td>76</td>\n",
       "      <td>4558</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>ATC</td>\n",
       "      <td>487</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>856</td>\n",
       "      <td>0</td>\n",
       "      <td>538</td>\n",
       "      <td>5.0</td>\n",
       "      <td>IGLPRVGPGETRTAGAASRHFARP</td>\n",
       "      <td>0.684211</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4601</th>\n",
       "      <td>SLC6A16</td>\n",
       "      <td>ENST00000335875</td>\n",
       "      <td>168</td>\n",
       "      <td>243</td>\n",
       "      <td>76</td>\n",
       "      <td>4602</td>\n",
       "      <td>extension</td>\n",
       "      <td>&lt;a href=\"https://trips.ucc.ie/homo_sapiens/Gen...</td>\n",
       "      <td>TTG</td>\n",
       "      <td>491</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>856</td>\n",
       "      <td>0</td>\n",
       "      <td>538</td>\n",
       "      <td>5.0</td>\n",
       "      <td>LGLGSPQEDEPSLAEAGAAPESLC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5929 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_ext           tr_id1  Start_ext  Stop_ext  len_ext  global_rank_ext  \\\n",
       "1785     TPM4  ENST00000344824         59       119       61             1786   \n",
       "8498   DNAJC7  ENST00000457167         -1         0        0             8499   \n",
       "8497    UBXN4  ENST00000272638         -1         0        0             8498   \n",
       "8496     SUN2  ENST00000405510         -1         0        0             8497   \n",
       "8495  SLC36A1  ENST00000243389         -1         0        0             8496   \n",
       "...       ...              ...        ...       ...      ...              ...   \n",
       "4546   RAB40A  ENST00000372633       1985      2120      136             4547   \n",
       "4626  ST8SIA3  ENST00000324000       1960      2035       76             4627   \n",
       "4536   NECAB1  ENST00000417640        263       338       76             4537   \n",
       "4557    BHMG1  ENST00000457052        342       417       76             4558   \n",
       "4601  SLC6A16  ENST00000335875        168       243       76             4602   \n",
       "\n",
       "     type_all_codons                                     trips_viz_link  \\\n",
       "1785       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "8498       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "8497       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "8496       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "8495       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "...              ...                                                ...   \n",
       "4546       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "4626       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "4536       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "4557       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "4601       extension  <a href=\"https://trips.ucc.ie/homo_sapiens/Gen...   \n",
       "\n",
       "     Start_codon_ext  Highframe rank  ...  Lowframe value  Stop rank  \\\n",
       "1785             CTG             749  ...             -10        437   \n",
       "8498             NaN             488  ...              -1        211   \n",
       "8497             NaN             488  ...              -1        211   \n",
       "8496             NaN             488  ...              -1        211   \n",
       "8495             NaN             488  ...              -1        211   \n",
       "...              ...             ...  ...             ...        ...   \n",
       "4546             ATT             487  ...               1        210   \n",
       "4626             ATA             495  ...              -1        210   \n",
       "4536             CTG             487  ...               0        210   \n",
       "4557             ATC             487  ...               0        210   \n",
       "4601             TTG             491  ...              -3        210   \n",
       "\n",
       "      Stop value  Start rank  Start value  Coverage rank  Coverage_value_ext  \\\n",
       "1785        -240         212         1356            587                 0.0   \n",
       "8498          -1         857           -1            587                 0.0   \n",
       "8497          -1         857           -1            587                 0.0   \n",
       "8496          -1         857           -1            587                 0.0   \n",
       "8495          -1         857           -1            587                 0.0   \n",
       "...          ...         ...          ...            ...                 ...   \n",
       "4546           0         855            1            538                 5.0   \n",
       "4626           0         855            1            538                 5.0   \n",
       "4536           0         854            2            538                 5.0   \n",
       "4557           0         856            0            538                 5.0   \n",
       "4601           0         856            0            538                 5.0   \n",
       "\n",
       "                               Amino acid sequence  Proteomics_count_ext  \\\n",
       "1785                           LPQPPQSPPRTPRPPRQRP              0.000000   \n",
       "8498                                           NaN              0.000000   \n",
       "8497                                           NaN              0.000000   \n",
       "8496                                           NaN              0.000000   \n",
       "8495                                           NaN              0.000000   \n",
       "...                                            ...                   ...   \n",
       "4546  ISLYFTDFPISFFMFYANFSRRTGPAPPLRTTPRAWLRRECGAS              0.000000   \n",
       "4626                      IYTPASCWPLNGPISPVSLNPAQP              0.000000   \n",
       "4536                      LRPRPCQSRCVRLAPLRLRPSGLP              0.000000   \n",
       "4557                      IGLPRVGPGETRTAGAASRHFARP              0.684211   \n",
       "4601                      LGLGSPQEDEPSLAEAGAAPESLC              0.000000   \n",
       "\n",
       "     CDS_ratio  \n",
       "1785  0.041187  \n",
       "8498  0.000000  \n",
       "8497  0.000000  \n",
       "8496  0.000000  \n",
       "8495  0.000000  \n",
       "...        ...  \n",
       "4546  0.000000  \n",
       "4626  0.000000  \n",
       "4536  0.000000  \n",
       "4557  0.000000  \n",
       "4601  0.000000  \n",
       "\n",
       "[5929 rows x 22 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transl_with_all_codons[transl_with_all_codons['Coverage_value_ext'] <= 5].sort_values(by='Coverage_value_ext')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genes with no translation upstream in theor.ext but with CDS translation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENST00000403106', 'ENST00000381916', 'ENST00000423831']\n",
      "76910 75232 75139\n"
     ]
    }
   ],
   "source": [
    "# get total list of transcripts \n",
    "\n",
    "amb_path = '../../data/counts_ext/download/N_term_amb/*'\n",
    "unamb_path = '../../data/counts_ext/download/N_term_unamb/*'\n",
    "\n",
    "list_of_tr = list(set([tr_path.split('/')[-1].split('_counts')[0] for tr_path in glob.glob(unamb_path)+glob.glob(amb_path)]))\n",
    "\n",
    "list_of_amb_tr = [tr_path.split('/')[-1].split('_counts')[0] for tr_path in glob.glob(amb_path)]\n",
    "\n",
    "list_of_unamb_tr = [tr_path.split('/')[-1].split('_counts')[0] for tr_path in glob.glob(unamb_path)]\n",
    "\n",
    "print (list_of_tr[0:3])\n",
    "\n",
    "print (len(list_of_tr), len(list_of_amb_tr), len(list_of_unamb_tr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ASPH',\n",
       " 'HSP90AA1',\n",
       " 'CD24',\n",
       " 'HN1',\n",
       " 'YBX1',\n",
       " 'HNRNPA0',\n",
       " 'NDUFA2',\n",
       " 'GPRC5A',\n",
       " 'SFPQ',\n",
       " 'NPM1',\n",
       " 'TK1',\n",
       " 'CCNB1',\n",
       " 'CD59',\n",
       " 'LRRC59',\n",
       " 'ATP5E',\n",
       " 'YWHAE',\n",
       " 'PRC1',\n",
       " 'NONO',\n",
       " 'MRFAP1',\n",
       " 'HDGF',\n",
       " 'NSUN2',\n",
       " 'STMN1',\n",
       " 'EZR',\n",
       " 'HM13',\n",
       " 'JUN',\n",
       " 'LY6K',\n",
       " 'NMU',\n",
       " 'FTL',\n",
       " 'NDUFS5',\n",
       " 'LGALS1',\n",
       " 'ID2',\n",
       " 'COPG1',\n",
       " 'ERVK3-1',\n",
       " 'DDIT4',\n",
       " 'DUSP5',\n",
       " 'COX6C',\n",
       " 'KDM1A',\n",
       " 'FSTL1',\n",
       " 'PSMD13',\n",
       " 'YBX3',\n",
       " 'PLEKHB2',\n",
       " 'COX7C',\n",
       " 'LARS',\n",
       " 'GSPT1',\n",
       " 'RAD1',\n",
       " 'PPP1R14B',\n",
       " 'ST13',\n",
       " 'LAMP2',\n",
       " 'PON2',\n",
       " 'SLC25A1',\n",
       " 'RBBP4',\n",
       " 'PLIN3',\n",
       " 'SYNCRIP',\n",
       " 'RPA1',\n",
       " 'KHDRBS1',\n",
       " 'LASP1',\n",
       " 'ASF1B',\n",
       " 'H1FX',\n",
       " 'F3',\n",
       " 'CD151',\n",
       " 'TGFB1',\n",
       " 'UCK2',\n",
       " 'LAPTM4B',\n",
       " 'SLC25A32',\n",
       " 'CDK2',\n",
       " 'NDUFB10',\n",
       " 'RRM1',\n",
       " 'EPRS',\n",
       " 'C1QBP',\n",
       " 'MAPKAPK2',\n",
       " 'TCEAL9',\n",
       " 'FADD',\n",
       " 'PPP1CB',\n",
       " 'AARS',\n",
       " 'SYAP1',\n",
       " 'C9ORF16',\n",
       " 'EREG',\n",
       " 'SUPT4H1',\n",
       " 'HMGCS1',\n",
       " 'PNPLA2',\n",
       " 'GALE',\n",
       " 'LDB1',\n",
       " 'BLMH',\n",
       " 'PDCL',\n",
       " 'HMGN1',\n",
       " 'PMAIP1',\n",
       " 'NELFB',\n",
       " 'DESI1',\n",
       " 'VMP1',\n",
       " 'RCN1',\n",
       " 'SUPT16H',\n",
       " 'LIPA',\n",
       " 'DYNLL2',\n",
       " 'PFKM',\n",
       " 'BAG6',\n",
       " 'SPCS1',\n",
       " 'PEG10',\n",
       " 'EIF2B1',\n",
       " 'SNRNP25',\n",
       " 'SLC39A14',\n",
       " 'NPTN',\n",
       " 'PPA1',\n",
       " 'TOMM20',\n",
       " 'PDIA6',\n",
       " 'UBE2D2',\n",
       " 'SRRT',\n",
       " 'TRIP13',\n",
       " 'CD46',\n",
       " 'CHTOP',\n",
       " 'FXR2',\n",
       " 'KDELR3',\n",
       " 'PAFAH1B2',\n",
       " 'RNF149',\n",
       " 'ATP6V0C',\n",
       " 'CKAP2',\n",
       " 'GNPNAT1',\n",
       " 'RAE1',\n",
       " 'PDCD6IP',\n",
       " 'FOSL1',\n",
       " 'BTF3L4',\n",
       " 'AP3S1',\n",
       " 'TXNDC9',\n",
       " 'PDCD6',\n",
       " 'G3BP2',\n",
       " 'TMEM9B',\n",
       " 'HNRNPA2B1',\n",
       " 'CTSZ',\n",
       " 'SPTSSA',\n",
       " 'KCTD9',\n",
       " 'SLC25A20',\n",
       " 'SERPINE2',\n",
       " 'MGEA5',\n",
       " 'DUSP6',\n",
       " 'CARD19',\n",
       " 'ATOX1',\n",
       " 'CTGF',\n",
       " 'UBE2M',\n",
       " 'SLBP',\n",
       " 'NACA',\n",
       " 'MARVELD1',\n",
       " 'SUN1',\n",
       " 'SND1',\n",
       " 'GADD45A',\n",
       " 'PTOV1',\n",
       " 'NUDT21',\n",
       " 'KRBOX4',\n",
       " 'RP4-613B23.5',\n",
       " 'KAT7',\n",
       " 'SPPL2A',\n",
       " 'HSPH1',\n",
       " 'MARCKS',\n",
       " 'GNAI1',\n",
       " 'HIGD1A',\n",
       " 'EEF1B2',\n",
       " 'TSPAN13',\n",
       " 'RPS11',\n",
       " 'TOP1',\n",
       " 'RCL1',\n",
       " 'ACLY',\n",
       " 'KLF5',\n",
       " 'CHAF1B',\n",
       " 'FGF2',\n",
       " 'TMEM123',\n",
       " 'MDH1',\n",
       " 'HMGXB3',\n",
       " 'DAPK3',\n",
       " 'CD99',\n",
       " 'PYCR1',\n",
       " 'DERL1',\n",
       " 'ERGIC1',\n",
       " 'MFSD12',\n",
       " 'CAPRIN1',\n",
       " 'AHCYL1',\n",
       " 'MRPS7',\n",
       " 'FAM64A',\n",
       " 'CNOT10',\n",
       " 'GMNN',\n",
       " 'RUVBL1',\n",
       " 'SUFU',\n",
       " 'UFD1L',\n",
       " 'CD276',\n",
       " 'ACYP1',\n",
       " 'MAGEF1',\n",
       " 'CDC42SE2',\n",
       " 'DDX46',\n",
       " 'RAB14',\n",
       " 'PDXK',\n",
       " 'NUP93',\n",
       " 'C1GALT1',\n",
       " 'WLS',\n",
       " 'ARMCX6',\n",
       " 'ATP1B3',\n",
       " 'THOP1',\n",
       " 'CCT5',\n",
       " 'SPINT1',\n",
       " 'RHOB',\n",
       " 'ZNF384',\n",
       " 'SAP30',\n",
       " 'NAT14',\n",
       " 'FBXL3',\n",
       " 'TBL3',\n",
       " 'PDP1',\n",
       " 'ARHGEF7',\n",
       " 'TIMM8B',\n",
       " 'AC013461.1',\n",
       " 'CRIPT',\n",
       " 'UBE2G1',\n",
       " 'TMEM41B',\n",
       " 'LAMC1',\n",
       " 'DAZAP2',\n",
       " 'ARPC1A',\n",
       " 'TWF2',\n",
       " 'GOT2',\n",
       " 'EIF5A2',\n",
       " 'INTS10',\n",
       " 'HEBP2',\n",
       " 'ZNF622',\n",
       " 'NXN',\n",
       " 'IFI27L1',\n",
       " 'CITED2',\n",
       " 'ATG12',\n",
       " 'ADAM15',\n",
       " 'ZIC2',\n",
       " 'PINX1',\n",
       " 'SLC25A17',\n",
       " 'ENDOG',\n",
       " 'USP33',\n",
       " 'ATF3',\n",
       " 'SMIM20',\n",
       " 'AURKAIP1',\n",
       " 'CTD-2135J3.4',\n",
       " 'PDIA3',\n",
       " 'CDR2',\n",
       " 'TXNL4A',\n",
       " 'RINT1',\n",
       " 'USP19',\n",
       " 'FAM168A',\n",
       " 'CARM1',\n",
       " 'NFKBIZ',\n",
       " 'ARPC5',\n",
       " 'RBM45',\n",
       " 'SEL1L',\n",
       " 'MCMBP',\n",
       " 'F2R',\n",
       " 'LAMTOR5',\n",
       " 'ZMYM1',\n",
       " 'FOSB',\n",
       " 'PARN',\n",
       " 'PTPA',\n",
       " 'GNPDA2',\n",
       " 'POLD4',\n",
       " 'TSN',\n",
       " 'CKB',\n",
       " 'CXXC1',\n",
       " 'CYTH2',\n",
       " 'SPHK1',\n",
       " 'DNER',\n",
       " 'NAA38',\n",
       " 'HYLS1',\n",
       " 'MXD4',\n",
       " 'ACTL6A',\n",
       " 'LOX',\n",
       " 'AMFR',\n",
       " 'PRSS8',\n",
       " 'TIMP3',\n",
       " 'LYRM7',\n",
       " 'KHDRBS3',\n",
       " 'DPH5',\n",
       " 'TRMT6',\n",
       " 'LMO4',\n",
       " 'RP3-468K18.7',\n",
       " 'OGT',\n",
       " 'FAM160B1',\n",
       " 'CDV3',\n",
       " 'INSL4',\n",
       " 'MFAP3',\n",
       " 'PBRM1',\n",
       " 'WDR26',\n",
       " 'PABPC4',\n",
       " 'NUFIP2',\n",
       " 'GGNBP2',\n",
       " 'TNS3',\n",
       " 'SMARCA1',\n",
       " 'AREG',\n",
       " 'GNS',\n",
       " 'FN1',\n",
       " 'UBA5',\n",
       " 'ISG15',\n",
       " 'TMEM165',\n",
       " 'PA2G4',\n",
       " 'DHFR',\n",
       " 'NUDC',\n",
       " 'CDC34',\n",
       " 'MAPK14',\n",
       " 'METTL15',\n",
       " 'PSMA7',\n",
       " 'KCTD3',\n",
       " 'TNC',\n",
       " 'PIM2',\n",
       " 'RNF167',\n",
       " 'ZNF316',\n",
       " 'TPBG',\n",
       " 'BLM',\n",
       " 'NPLOC4',\n",
       " 'GSTO1',\n",
       " 'HIST2H2AA4',\n",
       " 'SMIM12',\n",
       " 'DARS',\n",
       " 'CNP',\n",
       " 'UBALD2',\n",
       " 'ADRM1',\n",
       " 'CIT',\n",
       " 'IGF2R',\n",
       " 'MAGOHB',\n",
       " 'CHPT1',\n",
       " 'STK38',\n",
       " 'AK3',\n",
       " 'CARNMT1',\n",
       " 'GPRC5B',\n",
       " 'ADI1',\n",
       " 'MAD2L2',\n",
       " 'ALDOA',\n",
       " 'FKBP1A',\n",
       " 'RP11-314N13.10',\n",
       " 'MSL1',\n",
       " 'SAFB',\n",
       " 'VMA21',\n",
       " 'UBE2J2',\n",
       " 'EED',\n",
       " 'FLNA',\n",
       " 'COTL1',\n",
       " 'TMEM107',\n",
       " 'RPTOR',\n",
       " 'NSDHL',\n",
       " 'PCSK9',\n",
       " 'SOX2',\n",
       " 'SH3BP5',\n",
       " 'MTHFD1',\n",
       " 'C1D',\n",
       " 'HNRNPR',\n",
       " 'RND3',\n",
       " 'FAM102B',\n",
       " 'RRNAD1',\n",
       " 'SERPINE1',\n",
       " 'OAT',\n",
       " 'MFAP1',\n",
       " 'GORASP2',\n",
       " 'FAM92A1',\n",
       " 'NPC1',\n",
       " 'CHMP2B',\n",
       " 'CALU',\n",
       " 'COX7B',\n",
       " 'UBA52',\n",
       " 'BAG3',\n",
       " 'CLU',\n",
       " 'SNX5',\n",
       " 'UGCG',\n",
       " 'GCLM',\n",
       " 'C18ORF8',\n",
       " 'MCM5',\n",
       " 'POLR1E',\n",
       " 'REEP6',\n",
       " 'MTPN',\n",
       " 'EAPP',\n",
       " 'C16ORF72',\n",
       " 'USP14',\n",
       " 'RP11-849F2.7',\n",
       " 'POGZ',\n",
       " 'PTPRJ',\n",
       " 'TMEM189',\n",
       " 'MORN2',\n",
       " 'INTS8',\n",
       " 'MED15',\n",
       " 'RHOC',\n",
       " 'NARS',\n",
       " 'C14ORF119',\n",
       " 'TCF7L2',\n",
       " 'UBTD2',\n",
       " 'FZD7',\n",
       " 'TARBP2',\n",
       " 'SLC17A5',\n",
       " 'NIP7',\n",
       " 'NAGA',\n",
       " 'LRRFIP1',\n",
       " 'KPNA1',\n",
       " 'CCT6A',\n",
       " 'CLK2',\n",
       " 'RP4-777O23.3',\n",
       " 'MAD1L1',\n",
       " 'ARHGAP8',\n",
       " 'CTH',\n",
       " 'PLPP2',\n",
       " 'EGLN3',\n",
       " 'RIPK1',\n",
       " 'UBE2Z',\n",
       " 'CCDC134',\n",
       " 'RAB3B',\n",
       " 'GINS4',\n",
       " 'FAM168B',\n",
       " 'CCDC8',\n",
       " 'BRD7',\n",
       " 'ADO',\n",
       " 'CCDC127',\n",
       " 'WDR36',\n",
       " 'SH3D19',\n",
       " 'ZNF629',\n",
       " 'RDH11',\n",
       " 'CRELD2',\n",
       " 'DNAJC3',\n",
       " 'AEN',\n",
       " 'MAFK',\n",
       " 'FXYD3',\n",
       " 'TRMT10C',\n",
       " 'LPP',\n",
       " 'ZFPL1',\n",
       " 'EID2',\n",
       " 'CYB5R2',\n",
       " 'AC002398.9',\n",
       " 'AF011889.5',\n",
       " 'PSENEN',\n",
       " 'FAM193A',\n",
       " 'GDI1',\n",
       " 'PNKP',\n",
       " 'BCL7C',\n",
       " 'ALDH3A2',\n",
       " 'FOS',\n",
       " 'TNKS2',\n",
       " 'RGS10',\n",
       " 'PTN',\n",
       " 'SLC25A38',\n",
       " 'NCBP1',\n",
       " 'JADE2',\n",
       " 'TMX3',\n",
       " 'PTMS',\n",
       " 'YARS',\n",
       " 'RAB42',\n",
       " 'SNCG',\n",
       " 'CDCA3',\n",
       " 'CCDC50',\n",
       " 'KDSR',\n",
       " 'RFK',\n",
       " 'DNAJC25',\n",
       " 'OCIAD1',\n",
       " 'CSRNP1',\n",
       " 'S100A10',\n",
       " 'PPP4R2',\n",
       " 'SERINC5',\n",
       " 'RHOBTB3',\n",
       " 'RSRC1',\n",
       " 'FAM122B',\n",
       " 'NPNT',\n",
       " 'PRMT7',\n",
       " 'ZMYND19',\n",
       " 'GNA13',\n",
       " 'MAN2A1',\n",
       " 'ZSWIM7',\n",
       " 'GAL',\n",
       " 'NDRG3',\n",
       " 'GABARAP',\n",
       " 'DESI2',\n",
       " 'SERTAD3',\n",
       " 'STIM2',\n",
       " 'HSPB1',\n",
       " 'SLC19A2',\n",
       " 'PIEZO1',\n",
       " 'IGFBP2',\n",
       " 'SLC25A6',\n",
       " 'VASN',\n",
       " 'CYSTM1',\n",
       " 'INPP5K',\n",
       " 'EIF1',\n",
       " 'NAE1',\n",
       " 'SHTN1',\n",
       " 'BRAP',\n",
       " 'DHX33',\n",
       " 'RLIM',\n",
       " 'CDC6',\n",
       " 'ESCO2',\n",
       " 'RNPEPL1',\n",
       " 'SMURF1',\n",
       " 'MOV10',\n",
       " 'KRTAP2-3',\n",
       " 'C3ORF33',\n",
       " 'RAB32',\n",
       " 'KRT5',\n",
       " 'C1GALT1C1',\n",
       " 'CHMP1B',\n",
       " 'ZNF428',\n",
       " 'DHRS7',\n",
       " 'EDIL3',\n",
       " 'NFKBID',\n",
       " 'OPN3',\n",
       " 'RTCB',\n",
       " 'PUS7',\n",
       " 'CHRNA5',\n",
       " 'TAF7',\n",
       " 'KIF1BP',\n",
       " 'PRDX5',\n",
       " 'PRDX6',\n",
       " 'SLC5A6',\n",
       " 'KIF23',\n",
       " 'ADNP2',\n",
       " 'ZNF268',\n",
       " 'VGF',\n",
       " 'SNF8',\n",
       " 'HYOU1',\n",
       " 'SEPT7',\n",
       " 'ACACA',\n",
       " 'QKI',\n",
       " 'USP22',\n",
       " 'PHLDA3',\n",
       " 'XXBAC-BPG32J3.22',\n",
       " 'KCNN4',\n",
       " 'CSNK2B',\n",
       " 'ZNF3',\n",
       " 'C17ORF58',\n",
       " 'FBRS',\n",
       " 'POLR3F',\n",
       " 'RTN4RL2',\n",
       " 'SLC22A17',\n",
       " 'TMPO',\n",
       " 'ZNF75A',\n",
       " 'SAP130',\n",
       " 'CHTF8',\n",
       " 'ZNHIT6',\n",
       " 'C9ORF40',\n",
       " 'ZNF654',\n",
       " 'ID3',\n",
       " 'SNRPD3',\n",
       " 'HPRT1',\n",
       " 'ARHGAP1',\n",
       " 'ALCAM',\n",
       " 'ASXL2',\n",
       " 'ASXL1',\n",
       " 'CRYZL1',\n",
       " 'PAQR3',\n",
       " 'HADH',\n",
       " 'SF1',\n",
       " 'RP11-544M22.13',\n",
       " 'SLC12A2',\n",
       " 'ZC3H7B',\n",
       " 'PMM2',\n",
       " 'VDR',\n",
       " 'BASP1',\n",
       " 'BEX3',\n",
       " 'PPP1R35',\n",
       " 'IRX5',\n",
       " 'HIST3H2A',\n",
       " 'ELOF1',\n",
       " 'SSR2',\n",
       " 'PODXL',\n",
       " 'NR4A2',\n",
       " 'BTBD6',\n",
       " 'CCND3',\n",
       " 'TTC4',\n",
       " 'ZNF551',\n",
       " 'TRIM44',\n",
       " 'POLD3',\n",
       " 'RAB12',\n",
       " 'LARP1B',\n",
       " 'CIR1',\n",
       " 'FOSL2',\n",
       " 'YME1L1',\n",
       " 'DNMT3B',\n",
       " 'URGCP',\n",
       " 'HTRA1',\n",
       " 'TRIM7',\n",
       " 'MTHFD1L',\n",
       " 'MBP',\n",
       " 'HDAC2',\n",
       " 'YTHDC1',\n",
       " 'SPATS2L',\n",
       " 'TP53BP2',\n",
       " 'TSPAN4',\n",
       " 'TRIM33',\n",
       " 'SOD2',\n",
       " 'MSI2',\n",
       " 'DCAF1',\n",
       " 'HIST1H2BN',\n",
       " 'CECR2',\n",
       " 'IGFBP4',\n",
       " 'ZNF398',\n",
       " 'IL32',\n",
       " 'MLF1',\n",
       " 'EFNA1',\n",
       " 'FAM107B',\n",
       " 'EIF2AK1',\n",
       " 'ATP6V0A1',\n",
       " 'CYB5R3',\n",
       " 'C6ORF106',\n",
       " 'ARID1A',\n",
       " 'NUPL2',\n",
       " 'ARF3',\n",
       " 'RP11-302B13.5',\n",
       " 'DENND6A',\n",
       " 'PLP2',\n",
       " 'SSFA2',\n",
       " 'RTN3',\n",
       " 'PLPP1',\n",
       " 'TRAPPC12',\n",
       " 'SRSF2',\n",
       " 'DCUN1D4',\n",
       " 'TRIQK',\n",
       " 'RAB1B',\n",
       " 'SNRPB2',\n",
       " 'RBM43',\n",
       " 'TFAP2C',\n",
       " 'AAAS',\n",
       " 'INO80E',\n",
       " 'TOLLIP',\n",
       " 'BMP1',\n",
       " 'C4ORF32',\n",
       " 'PITHD1',\n",
       " 'TGIF1',\n",
       " 'TSNAX',\n",
       " 'TSNAX-DISC1',\n",
       " 'GNL2',\n",
       " 'PFKFB3',\n",
       " 'PRMT3',\n",
       " 'TMEM161B',\n",
       " 'WEE1',\n",
       " 'MIEF2',\n",
       " 'CCT2',\n",
       " 'FBXW4',\n",
       " 'CST6',\n",
       " 'OXCT1',\n",
       " 'TCAF1',\n",
       " 'RMND5A',\n",
       " 'ZNF281',\n",
       " 'MAPRE2',\n",
       " 'SLC25A29',\n",
       " 'FAM96A',\n",
       " 'METTL12',\n",
       " 'PIP4K2B',\n",
       " 'MGST2',\n",
       " 'MDK',\n",
       " 'ORC5',\n",
       " 'YBEY',\n",
       " 'ESM1',\n",
       " 'PELI2',\n",
       " 'COPRS',\n",
       " 'DHCR7',\n",
       " 'RAPGEF5',\n",
       " 'FAM117A',\n",
       " 'RNASET2',\n",
       " 'DCUN1D5',\n",
       " 'RERE',\n",
       " 'TRIM16',\n",
       " 'PPTC7',\n",
       " 'HSPA1B',\n",
       " 'BRI3BP',\n",
       " 'SAMD8',\n",
       " 'CMTR1',\n",
       " 'CNOT11',\n",
       " 'MYL9',\n",
       " 'HAUS4',\n",
       " 'LSM8',\n",
       " 'KIF15',\n",
       " 'NTSR1',\n",
       " 'GIMAP5',\n",
       " 'BRMS1L',\n",
       " 'RFX7',\n",
       " 'KLHL42',\n",
       " 'CACYBP',\n",
       " 'VAT1',\n",
       " 'TSPAN9',\n",
       " 'HSPA1A',\n",
       " 'TMEM206',\n",
       " 'HSPA12A',\n",
       " 'RRAGA',\n",
       " 'ARRB2',\n",
       " 'CDK2AP1',\n",
       " 'PYGB',\n",
       " 'TSTD1',\n",
       " 'RIMKLB',\n",
       " 'GPC1',\n",
       " 'POLR2E',\n",
       " 'CSDC2',\n",
       " 'SCNN1A',\n",
       " 'CUL1',\n",
       " 'ZNF101',\n",
       " 'PDIA4',\n",
       " 'SLC22A18',\n",
       " 'MTMR12',\n",
       " 'SYNPO',\n",
       " 'AK4',\n",
       " 'TRA2B',\n",
       " 'ZW10',\n",
       " 'TMEM39A',\n",
       " 'CXORF57',\n",
       " 'DDX3Y',\n",
       " 'DARS2',\n",
       " 'EIF5A',\n",
       " 'QSER1',\n",
       " 'TXN',\n",
       " 'PDK1',\n",
       " 'PAQR7',\n",
       " 'COG3',\n",
       " 'MRPL34',\n",
       " 'MRPL4',\n",
       " 'BPNT1',\n",
       " 'MTA3',\n",
       " 'PPP2R2D',\n",
       " 'CHEK1',\n",
       " 'CNOT9',\n",
       " 'DIS3L2',\n",
       " 'CAMK2N1',\n",
       " 'RGS19',\n",
       " 'DIAPH1',\n",
       " 'SMTN',\n",
       " 'XRRA1',\n",
       " 'HOOK1',\n",
       " 'SCML2',\n",
       " 'NUDT4',\n",
       " 'TNPO2',\n",
       " 'AHCYL2',\n",
       " 'ING3',\n",
       " 'ARAP1',\n",
       " 'TKT',\n",
       " 'NCBP2-AS2',\n",
       " 'RPL7L1',\n",
       " 'USP27X',\n",
       " 'ARGLU1',\n",
       " 'NECTIN3',\n",
       " 'MIPOL1',\n",
       " 'WDR5',\n",
       " 'UQCRFS1',\n",
       " 'CDKN2B',\n",
       " 'RAB11FIP5',\n",
       " 'COX5B',\n",
       " 'HIST1H4E',\n",
       " 'MRPL38',\n",
       " 'RPL7',\n",
       " 'ARHGAP11A',\n",
       " 'RPL23',\n",
       " 'SLC35F3',\n",
       " 'NPRL2',\n",
       " 'ZNF275',\n",
       " 'PSCA',\n",
       " 'TINAGL1',\n",
       " 'TSPAN6',\n",
       " 'FAM217B',\n",
       " 'CEBPG',\n",
       " 'LMNB1',\n",
       " 'MAT2A',\n",
       " 'MED20',\n",
       " 'CDKN1C',\n",
       " 'TM7SF3',\n",
       " 'CDCA8',\n",
       " 'PXMP4',\n",
       " 'HMGN4',\n",
       " 'RHOV',\n",
       " 'CETN3',\n",
       " 'C21ORF91',\n",
       " 'NELFE',\n",
       " 'ATF6',\n",
       " 'ITGB1',\n",
       " 'LCOR',\n",
       " 'CCDC93',\n",
       " 'NUDCD1',\n",
       " 'ZNF593',\n",
       " 'C17ORF51',\n",
       " 'NDUFS3',\n",
       " 'NCS1',\n",
       " 'MAN1A1',\n",
       " 'WWC3',\n",
       " 'MTCH2',\n",
       " 'PHACTR4',\n",
       " 'SETD7',\n",
       " 'UFSP1',\n",
       " 'SNRNP35',\n",
       " 'KSR1',\n",
       " 'TNFAIP2',\n",
       " 'TRIM6',\n",
       " 'CBR4',\n",
       " 'TNFAIP1',\n",
       " 'SLAIN1',\n",
       " 'FGD4',\n",
       " 'FBLIM1',\n",
       " 'BCL7A',\n",
       " 'TULP3',\n",
       " 'NDUFB6',\n",
       " 'RPL22L1',\n",
       " 'NUMBL',\n",
       " 'GLRX5',\n",
       " 'GPAT3',\n",
       " 'VANGL2',\n",
       " 'ZNF789',\n",
       " 'KHK',\n",
       " 'NUP155',\n",
       " 'SEMA4B',\n",
       " 'OGFOD3',\n",
       " 'PSMG1',\n",
       " 'HSPA4L',\n",
       " 'S100PBP',\n",
       " 'UTP18',\n",
       " 'SESTD1',\n",
       " 'ENOPH1',\n",
       " 'TRIML2',\n",
       " 'YTHDC2',\n",
       " 'TUBA1C',\n",
       " 'COQ10B',\n",
       " 'SRPK1',\n",
       " 'IP6K2',\n",
       " 'CGA',\n",
       " 'ABHD14B',\n",
       " 'RFC3',\n",
       " 'HOXC6',\n",
       " 'DGAT1',\n",
       " 'CENPE',\n",
       " 'VPS26A',\n",
       " 'PEX11A',\n",
       " 'POLR2I',\n",
       " 'GNAS',\n",
       " 'RBM24',\n",
       " 'REST',\n",
       " 'FAM50A',\n",
       " 'IFIT5',\n",
       " 'PHF8',\n",
       " 'COPS8',\n",
       " 'IFITM2',\n",
       " 'RBM34',\n",
       " 'FAM129B',\n",
       " 'METTL23',\n",
       " 'MTA1',\n",
       " 'LIN37',\n",
       " 'KLHDC4',\n",
       " 'MRPL43',\n",
       " 'CPSF1',\n",
       " 'JTB',\n",
       " 'DPYSL5',\n",
       " 'VASP',\n",
       " 'GAN',\n",
       " 'MRPS2',\n",
       " 'CBR1',\n",
       " 'RGS16',\n",
       " 'NGFR',\n",
       " 'DNAJB11',\n",
       " 'IL18',\n",
       " 'FAHD2A',\n",
       " 'RALB',\n",
       " 'HNRNPD',\n",
       " 'TGFB1I1',\n",
       " 'HSPBAP1',\n",
       " 'ERF',\n",
       " 'TPRG1L',\n",
       " 'SDC1',\n",
       " 'SCNM1',\n",
       " 'FAM105A',\n",
       " 'RAMP1',\n",
       " 'PFAS',\n",
       " 'FADS3',\n",
       " 'PIP5K1A',\n",
       " 'TRMU',\n",
       " 'RNH1',\n",
       " 'MRPL2',\n",
       " 'TWIST2',\n",
       " 'UBE2R2',\n",
       " 'ZNF440',\n",
       " 'NXT2',\n",
       " 'SORD',\n",
       " 'DNAJA4',\n",
       " 'ERI2',\n",
       " 'UBR1',\n",
       " 'TSC22D1',\n",
       " 'RRP9',\n",
       " 'VEGFB',\n",
       " 'DYNC2LI1',\n",
       " 'GMDS',\n",
       " 'MT1F',\n",
       " 'MCTS1',\n",
       " 'NRSN2',\n",
       " 'VKORC1L1',\n",
       " 'MAOB',\n",
       " 'TMEM14A',\n",
       " 'TSPYL4',\n",
       " 'GOLGB1',\n",
       " 'GBP2',\n",
       " 'GTF2I',\n",
       " 'SPATA2L',\n",
       " 'CDC26',\n",
       " 'PRKAR2A',\n",
       " 'MARCKSL1',\n",
       " 'HERC5',\n",
       " 'PTPN12',\n",
       " 'ECHDC3',\n",
       " 'PNMA1',\n",
       " 'HNRNPDL',\n",
       " 'PCNP',\n",
       " 'UROS',\n",
       " 'RPS2',\n",
       " 'RAP1GAP',\n",
       " 'CCT4',\n",
       " 'UBL7',\n",
       " 'PGK1',\n",
       " 'KIAA0895',\n",
       " 'HID1',\n",
       " 'SPTY2D1',\n",
       " 'EMC6',\n",
       " 'XKR8',\n",
       " 'TBC1D13',\n",
       " 'MAPKAPK5',\n",
       " 'RBM10',\n",
       " 'MSLN',\n",
       " 'DRAM1',\n",
       " 'QSOX1',\n",
       " 'FLNC',\n",
       " 'CARD10',\n",
       " 'WAPL',\n",
       " 'FBN2',\n",
       " 'PCDH1',\n",
       " 'TMEM8B',\n",
       " 'RBM8A',\n",
       " 'KBTBD4',\n",
       " 'NQO2',\n",
       " 'CAMKMT',\n",
       " 'STK11IP',\n",
       " 'PPP1R21',\n",
       " 'NDUFV3',\n",
       " 'ZNF823',\n",
       " 'TMEM260',\n",
       " 'SLC25A15',\n",
       " 'C4ORF48',\n",
       " 'IFFO2',\n",
       " 'NCAPD3',\n",
       " 'ARMCX2',\n",
       " 'C1ORF112',\n",
       " 'C7ORF25',\n",
       " 'YIPF1',\n",
       " 'EFCAB14',\n",
       " 'ZNF485',\n",
       " 'BOD1L1',\n",
       " 'ACSF3',\n",
       " 'CCDC28A',\n",
       " 'ECD',\n",
       " 'WDR18',\n",
       " 'GATAD1',\n",
       " 'KIFC3',\n",
       " 'GTF2IRD1',\n",
       " 'SLC25A33',\n",
       " 'BUB1B',\n",
       " 'KANSL2',\n",
       " 'AMOT',\n",
       " 'SEC16A',\n",
       " 'CERS4',\n",
       " 'GNAI2',\n",
       " 'DUSP9',\n",
       " 'ARHGEF2',\n",
       " 'AMPD3',\n",
       " 'PRRG2',\n",
       " 'TP63',\n",
       " 'FOXO1',\n",
       " 'IMMP2L',\n",
       " 'RHEBL1',\n",
       " 'RPP25',\n",
       " 'CRYM',\n",
       " 'USP7',\n",
       " 'SIGIRR',\n",
       " 'WDR76',\n",
       " 'CAST',\n",
       " 'TRIM52',\n",
       " 'FBXL18',\n",
       " 'UQCC3',\n",
       " 'LPCAT4',\n",
       " 'PET100',\n",
       " 'SLC33A1',\n",
       " 'SHPRH',\n",
       " 'MIDN',\n",
       " 'FAM127B',\n",
       " 'DZIP1L',\n",
       " 'PRIMPOL',\n",
       " 'PPIF',\n",
       " 'ARL13B',\n",
       " 'STX4',\n",
       " 'MTERF3',\n",
       " 'ITM2A',\n",
       " 'SOD3',\n",
       " 'SLC35A3',\n",
       " 'FAM118A',\n",
       " 'DNTTIP1',\n",
       " 'LAMB2',\n",
       " 'SH2D4A',\n",
       " 'CXCL16',\n",
       " 'KATNA1',\n",
       " 'ESYT1',\n",
       " 'CSNK1D',\n",
       " 'SRSF1',\n",
       " 'TSNARE1',\n",
       " 'SWSAP1',\n",
       " 'BET1',\n",
       " 'TMED8',\n",
       " 'FZR1',\n",
       " 'ZBTB9',\n",
       " 'BARD1',\n",
       " 'LINC00998',\n",
       " 'TACSTD2',\n",
       " 'ZNF235',\n",
       " 'TMEM116',\n",
       " 'CPNE7',\n",
       " 'ZNF589',\n",
       " ...]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transl_with_all_codons2['gene_ext'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "69997 76910\n",
      "['ENST00000403106', 'ENST00000381916']\n"
     ]
    }
   ],
   "source": [
    "# choose transcripts with no rank at al  \n",
    "\n",
    "SEL_TR = list(set(list_of_tr) - set(transl_with_all_codons2[0:7000]['tr_id1'].tolist()))\n",
    "\n",
    "print (len(SEL_TR), len(list_of_tr))\n",
    "print (SEL_TR[0:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full CDS len: 94359\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "33089"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pc_g25['tr_len'] = [len(x) for x in metadata_pc_g25['transcript_seq'].tolist()]\n",
    "\n",
    "full_cds = [x.split('.')[0] for x in metadata_pc_g25[(metadata_pc_g25['cds_stop_pos'] <= metadata_pc_g25['tr_len'])].tr_id.tolist()]\n",
    "\n",
    "print ('full CDS len:', len(full_cds))\n",
    "\n",
    "# length of theor ext is >=20 codons!!! \n",
    "SEL_TR_full_CDS = list(set(full_cds).intersection(SEL_TR).intersection(meta[(meta['len_codons'] >= 19)].tr_id1.tolist()))\n",
    "\n",
    "len(SEL_TR_full_CDS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12934, 39121)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = pd.DataFrame(set(full_cds).intersection(meta[(meta['len_codons'] >= 19)].tr_id1.tolist()))\n",
    "a.columns = ['tr_id1']\n",
    "a.merge(meta[['tr_id1', 'gene']]).gene.nunique(), a.merge(meta[['tr_id1', 'gene']]).tr_id1.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3523, 8577)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b = pd.DataFrame(set(full_cds).intersection(meta[(meta['len_codons'] >= 20)].tr_id1.tolist()) - set(transl_with_all_codons2[0:5000]['tr_id1'].tolist()))\n",
    "b.columns = ['tr_id1']\n",
    "b = b.merge(meta[['tr_id1', 'gene']])\n",
    "\n",
    "# exclude genes \n",
    "b = b[~b['gene'].isin(transl_with_all_codons2['gene_ext'].tolist())]\n",
    "\n",
    "b.gene.nunique(), b.tr_id1.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3485, 8485)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b = pd.DataFrame(set(full_cds).intersection(meta[(meta['len_codons'] >= 20)].tr_id1.tolist()) - set(transl_with_all_codons2[0:]['tr_id1'].tolist()))\n",
    "b.columns = ['tr_id1']\n",
    "b = b.merge(meta[['tr_id1', 'gene']])\n",
    "\n",
    "# exclude genes \n",
    "b = b[~b['gene'].isin(transl_with_all_codons2['gene_ext'].tolist())]\n",
    "\n",
    "b.gene.nunique(), b.tr_id1.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of genes with non-truncated CDS and with theretical ext length >= 20 codons is 11813 (28336 transcripts)\n",
    "# number of genes with all of that exclusing genes predicted with Trips-viz (all 10k): 2819 (5686 transcripts) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "LIST_OF_FILT = b['tr_id1'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transcript 0 is processed out of 8485\n",
      "transcript 100 is processed out of 8485\n",
      "transcript 200 is processed out of 8485\n",
      "transcript 300 is processed out of 8485\n",
      "transcript 400 is processed out of 8485\n",
      "transcript 500 is processed out of 8485\n",
      "transcript 600 is processed out of 8485\n",
      "transcript 700 is processed out of 8485\n",
      "transcript 800 is processed out of 8485\n",
      "transcript 900 is processed out of 8485\n",
      "transcript 1000 is processed out of 8485\n",
      "transcript 1100 is processed out of 8485\n",
      "transcript 1200 is processed out of 8485\n",
      "transcript 1300 is processed out of 8485\n",
      "transcript 1400 is processed out of 8485\n",
      "transcript 1500 is processed out of 8485\n",
      "transcript 1600 is processed out of 8485\n",
      "transcript 1700 is processed out of 8485\n",
      "transcript 1800 is processed out of 8485\n",
      "transcript 1900 is processed out of 8485\n",
      "transcript 2000 is processed out of 8485\n",
      "transcript 2100 is processed out of 8485\n",
      "transcript 2200 is processed out of 8485\n",
      "transcript 2300 is processed out of 8485\n",
      "transcript 2400 is processed out of 8485\n",
      "transcript 2500 is processed out of 8485\n",
      "transcript 2600 is processed out of 8485\n",
      "transcript 2700 is processed out of 8485\n",
      "transcript 2800 is processed out of 8485\n",
      "transcript 2900 is processed out of 8485\n",
      "transcript 3000 is processed out of 8485\n",
      "transcript 3100 is processed out of 8485\n",
      "transcript 3200 is processed out of 8485\n",
      "transcript 3300 is processed out of 8485\n",
      "transcript 3400 is processed out of 8485\n",
      "transcript 3500 is processed out of 8485\n",
      "transcript 3600 is processed out of 8485\n",
      "transcript 3700 is processed out of 8485\n",
      "transcript 3800 is processed out of 8485\n",
      "transcript 3900 is processed out of 8485\n",
      "transcript 4000 is processed out of 8485\n",
      "transcript 4100 is processed out of 8485\n",
      "transcript 4200 is processed out of 8485\n",
      "transcript 4300 is processed out of 8485\n",
      "transcript 4400 is processed out of 8485\n",
      "transcript 4500 is processed out of 8485\n",
      "transcript 4600 is processed out of 8485\n",
      "transcript 4700 is processed out of 8485\n",
      "transcript 4800 is processed out of 8485\n",
      "transcript 4900 is processed out of 8485\n",
      "transcript 5000 is processed out of 8485\n",
      "transcript 5100 is processed out of 8485\n",
      "transcript 5200 is processed out of 8485\n",
      "transcript 5300 is processed out of 8485\n",
      "transcript 5400 is processed out of 8485\n",
      "transcript 5500 is processed out of 8485\n",
      "transcript 5600 is processed out of 8485\n",
      "transcript 5700 is processed out of 8485\n",
      "transcript 5800 is processed out of 8485\n",
      "transcript 5900 is processed out of 8485\n",
      "transcript 6000 is processed out of 8485\n",
      "transcript 6100 is processed out of 8485\n",
      "transcript 6200 is processed out of 8485\n",
      "transcript 6300 is processed out of 8485\n",
      "transcript 6400 is processed out of 8485\n",
      "transcript 6500 is processed out of 8485\n",
      "transcript 6600 is processed out of 8485\n",
      "transcript 6700 is processed out of 8485\n",
      "transcript 6800 is processed out of 8485\n",
      "transcript 6900 is processed out of 8485\n",
      "transcript 7000 is processed out of 8485\n",
      "transcript 7100 is processed out of 8485\n",
      "transcript 7200 is processed out of 8485\n",
      "transcript 7300 is processed out of 8485\n",
      "transcript 7400 is processed out of 8485\n",
      "transcript 7500 is processed out of 8485\n",
      "transcript 7600 is processed out of 8485\n",
      "transcript 7700 is processed out of 8485\n",
      "transcript 7800 is processed out of 8485\n",
      "transcript 7900 is processed out of 8485\n",
      "transcript 8000 is processed out of 8485\n",
      "transcript 8100 is processed out of 8485\n",
      "transcript 8200 is processed out of 8485\n",
      "transcript 8300 is processed out of 8485\n",
      "transcript 8400 is processed out of 8485\n"
     ]
    }
   ],
   "source": [
    "out_counts = {'amb':{}, 'unamb':{}, 'all':{}}\n",
    "\n",
    "for i, tr_id in enumerate(LIST_OF_FILT):   \n",
    "    \n",
    "    if i % 100 == 0:\n",
    "        print ('transcript %s is processed out of 8485' % i)\n",
    "    # end of transcript = its length\n",
    "    tr_length = len(metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(tr_id)].iloc[0].transcript_seq)\n",
    "    tr_seq = metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(tr_id)].iloc[0].transcript_seq\n",
    "    gene = metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(tr_id)].iloc[0].gene\n",
    "            \n",
    "    if tr_id in list_of_unamb_tr:    \n",
    "            \n",
    "        tr_path_unamb = '../../data/counts_ext/download/N_term_unamb/%s_counts.txt' % tr_id\n",
    "        df_unamb = pd.read_csv(tr_path_unamb, sep='\\t')\n",
    "        d_unamb = dict(zip(df_unamb['pos'].tolist(), df_unamb['Total'].tolist()))\n",
    "            \n",
    "        for i in range(0, tr_length):\n",
    "            if i not in d_unamb:\n",
    "                d_unamb[i] = 0\n",
    "                \n",
    "        out_counts['unamb'][tr_id] = d_unamb \n",
    "        \n",
    "    else: # if tr_id is not in amb list\n",
    "        continue\n",
    "        #print ('%s is not in unamb list' % tr_id)\n",
    "           \n",
    "            \n",
    "# save it to tmp_res            \n",
    "with open('RiboSET_3485_genes.pickle', 'wb') as handle:\n",
    "    pickle.dump(out_counts, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘profiles’: File exists\r\n"
     ]
    }
   ],
   "source": [
    "!mkdir profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n"
     ]
    }
   ],
   "source": [
    "out_df_frames_un = {}\n",
    "\n",
    "for i, tr_id in enumerate(list(out_counts['unamb'].keys())):\n",
    "    \n",
    "    if i % 1000 == 0:\n",
    "        print (i)\n",
    "        \n",
    "    d = out_counts['unamb'][tr_id]\n",
    "    # d = dict with positioons and counts \n",
    "    frames = {'Frame_1': {}, 'Frame_2': {}, 'Frame_3': {} }\n",
    "    for pos, counts in d.items():\n",
    "        if pos % 3 == 2:\n",
    "            frames['Frame_1'][pos] = counts\n",
    "        elif pos % 3 == 0:\n",
    "            frames['Frame_2'][pos] = counts\n",
    "        else:\n",
    "            frames['Frame_3'][pos] = counts\n",
    "    # make df \n",
    "    li1 = []\n",
    "    li2 = []\n",
    "    li3 = []\n",
    "    \n",
    "    # end of transcript = its length\n",
    "    tr_length = len(metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(tr_id)].iloc[0].transcript_seq)\n",
    "    tr_seq = metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(tr_id)].iloc[0].transcript_seq\n",
    "    gene = metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(tr_id)].iloc[0].gene\n",
    "    \n",
    "    for i in range(1, tr_length+1):\n",
    "        if i in frames['Frame_1']:\n",
    "            li1.append([i, int(frames['Frame_1'][i])])\n",
    "        else:\n",
    "            li1.append([i, int(0)])\n",
    "    \n",
    "    df1 = pd.DataFrame(li1, columns = ['Position', 'Frame 1']).sort_values(by='Position')\n",
    "    \n",
    "    for i in range(1, tr_length+1):\n",
    "        if i in frames['Frame_2']:\n",
    "            li2.append([i, int(frames['Frame_2'][i])])\n",
    "        else:\n",
    "            li2.append([i, int(0)])\n",
    "    \n",
    "    df2 = pd.DataFrame(li2, columns = ['Position', 'Frame 2']).sort_values(by='Position')\n",
    "    \n",
    "    for i in range(1, tr_length+1):\n",
    "        if i in frames['Frame_3']:\n",
    "            li3.append([i, int(frames['Frame_3'][i])])\n",
    "        else:\n",
    "            li3.append([i, int(0)])\n",
    "    \n",
    "    df3 = pd.DataFrame(li3, columns = ['Position', 'Frame 3']).sort_values(by='Position')\n",
    "    \n",
    "    df_f = df1.merge(df2, on='Position', how='inner').merge(df3, on='Position', how='inner')\n",
    "    \n",
    "    df_f['Sequence'] = list(tr_seq)\n",
    "    \n",
    "    df_f['Position'] = df_f['Position'] -1 #-3\n",
    "    \n",
    "    df_f['rna_seq'] = None\n",
    "    \n",
    "    df_f[['Position', 'Sequence', 'Frame 1', 'Frame 2', 'Frame 3', 'rna_seq']].to_csv('profiles/%s_%s_unamb.csv' % (gene, tr_id), index=False)\n",
    "\n",
    "            \n",
    "    out_df_frames_un[tr_id] = df_f\n",
    "    \n",
    "# save it out_df_frames_un \n",
    "\n",
    "with open('RiboSET_3485_genes_frames.pickle', 'wb') as handle:\n",
    "    pickle.dump(out_df_frames_un, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alla/.local/lib/python3.7/site-packages/ipykernel_launcher.py:47: RuntimeWarning: invalid value encountered in double_scalars\n",
      "/home/alla/.local/lib/python3.7/site-packages/ipykernel_launcher.py:47: RuntimeWarning: divide by zero encountered in double_scalars\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n"
     ]
    }
   ],
   "source": [
    "li = []\n",
    "\n",
    "for i,el in enumerate(list(out_df_frames_un.keys())):\n",
    "    \n",
    "    if i % 1000 == 0:\n",
    "        print (i)\n",
    "    N_term_start = meta[meta['tr_id'].str.contains(el)].iloc[0].N_term_start\n",
    "    N_term_end = meta[meta['tr_id'].str.contains(el)].iloc[0].N_term_end\n",
    "    \n",
    "    cds_start = metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(el)].iloc[0].cds_start_pos\n",
    "    cds_stop = metadata_pc_g25[metadata_pc_g25['tr_id'].str.contains(el)].iloc[0].cds_stop_pos \n",
    "\n",
    "    \n",
    "    # get CDS frame and CDS array \n",
    "    if cds_start % 3 == 0:\n",
    "        frame = 'Frame 1'        \n",
    "    elif cds_start % 3 == 1:\n",
    "        frame = 'Frame 2'        \n",
    "    else:\n",
    "        frame = 'Frame 3'\n",
    "        \n",
    "    # extract counts for that frame \n",
    "    counts = out_df_frames_un[el][['Position', frame]]\n",
    "    \n",
    "    # CDS_array and CDS_coverage per codon and nonzero pos fraction \n",
    "    CDS_array = counts[cds_start:cds_stop][frame].tolist() \n",
    "    CDS_coverage = np.sum(CDS_array) / ((cds_stop - cds_start)//3)\n",
    "    \n",
    "    # number of non-zero pos / len of CDS in codons \n",
    "    CDS_nonzero_fraction = 3*np.count_nonzero(CDS_array) / (len(CDS_array)) \n",
    "    \n",
    "    #if frame == 'Frame 1':\n",
    "    #    CDS_array1 = [CDS_array[i] for i in range(0, len(CDS_array), 3)]\n",
    "    #elif frame == 'Frame 2':\n",
    "    CDS_array1 = [CDS_array[i] for i in range(1, len(CDS_array), 3)]\n",
    "    #else:\n",
    "    #    CDS_array1 = [CDS_array[i] for i in range(2, len(CDS_array), 3)]\n",
    "        \n",
    "    \n",
    "    theor_ext_array = counts[N_term_start:N_term_end][frame].tolist()\n",
    "    NTE_coverage = np.sum(theor_ext_array) / ((N_term_end - N_term_start) // 3)\n",
    "    NTE_nonzero_fraction = 3*np.count_nonzero(theor_ext_array) / (len(theor_ext_array)) \n",
    "    \n",
    "    \n",
    "    CDS_std = np.std(CDS_array1)\n",
    "    \n",
    "    Z_score = (NTE_coverage - CDS_coverage) / CDS_std\n",
    "    \n",
    "    \n",
    "    \n",
    "    li.append([el, Z_score, \n",
    "               CDS_coverage, NTE_coverage, \n",
    "               CDS_nonzero_fraction, NTE_nonzero_fraction,\n",
    "              ])\n",
    "    \n",
    "    \n",
    "df = pd.DataFrame(li, columns = ['tr_id', 'Z_score', \n",
    "               'CDS_coverage', 'NTE_coverage', \n",
    "               'CDS_nonzero_fraction', 'NTE_nonzero_fraction'])\n",
    "\n",
    "\n",
    "df.to_csv('3485_genes_df_coverages.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset = df[(df['NTE_nonzero_fraction'] < 0.0005) & \n",
    "               (df['NTE_coverage'] < 0.005) & \n",
    "               (df['CDS_coverage'] > 0.1) & \n",
    "               (df['CDS_nonzero_fraction'] > 0.3)].sort_values(by='CDS_coverage', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>Z_score</th>\n",
       "      <th>CDS_coverage</th>\n",
       "      <th>NTE_coverage</th>\n",
       "      <th>CDS_nonzero_fraction</th>\n",
       "      <th>NTE_nonzero_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6019</th>\n",
       "      <td>ENST00000264203</td>\n",
       "      <td>-0.857147</td>\n",
       "      <td>543.402299</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.873563</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008</th>\n",
       "      <td>ENST00000276682</td>\n",
       "      <td>-1.039719</td>\n",
       "      <td>435.678474</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.956403</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5076</th>\n",
       "      <td>ENST00000520813</td>\n",
       "      <td>-0.838518</td>\n",
       "      <td>286.958763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.811644</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3826</th>\n",
       "      <td>ENST00000409752</td>\n",
       "      <td>-0.692804</td>\n",
       "      <td>262.121053</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.878947</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7609</th>\n",
       "      <td>ENST00000522794</td>\n",
       "      <td>-0.945111</td>\n",
       "      <td>219.881188</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.968699</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4153</th>\n",
       "      <td>ENST00000615889</td>\n",
       "      <td>-0.616316</td>\n",
       "      <td>0.786517</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.393258</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>ENST00000589327</td>\n",
       "      <td>-0.616316</td>\n",
       "      <td>0.786517</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.393258</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6540</th>\n",
       "      <td>ENST00000572449</td>\n",
       "      <td>-0.375367</td>\n",
       "      <td>0.740668</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.324165</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503</th>\n",
       "      <td>ENST00000442111</td>\n",
       "      <td>-0.400687</td>\n",
       "      <td>0.651297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.314121</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7324</th>\n",
       "      <td>ENST00000416284</td>\n",
       "      <td>-0.449056</td>\n",
       "      <td>0.628788</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>374 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                tr_id   Z_score  CDS_coverage  NTE_coverage  \\\n",
       "6019  ENST00000264203 -0.857147    543.402299           0.0   \n",
       "1008  ENST00000276682 -1.039719    435.678474           0.0   \n",
       "5076  ENST00000520813 -0.838518    286.958763           0.0   \n",
       "3826  ENST00000409752 -0.692804    262.121053           0.0   \n",
       "7609  ENST00000522794 -0.945111    219.881188           0.0   \n",
       "...               ...       ...           ...           ...   \n",
       "4153  ENST00000615889 -0.616316      0.786517           0.0   \n",
       "939   ENST00000589327 -0.616316      0.786517           0.0   \n",
       "6540  ENST00000572449 -0.375367      0.740668           0.0   \n",
       "1503  ENST00000442111 -0.400687      0.651297           0.0   \n",
       "7324  ENST00000416284 -0.449056      0.628788           0.0   \n",
       "\n",
       "      CDS_nonzero_fraction  NTE_nonzero_fraction  \n",
       "6019              0.873563                   0.0  \n",
       "1008              0.956403                   0.0  \n",
       "5076              0.811644                   0.0  \n",
       "3826              0.878947                   0.0  \n",
       "7609              0.968699                   0.0  \n",
       "...                    ...                   ...  \n",
       "4153              0.393258                   0.0  \n",
       "939               0.393258                   0.0  \n",
       "6540              0.324165                   0.0  \n",
       "1503              0.314121                   0.0  \n",
       "7324              0.318182                   0.0  \n",
       "\n",
       "[374 rows x 6 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
