{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Bad key text.latex.unicode in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 112 ('text.latex.unicode : False # use \"ucs\" and \"inputenc\" LaTeX packages for handling')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key savefig.frameon in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 423 ('savefig.frameon : True')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key pgf.debug in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 444 ('pgf.debug           : False')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key verbose.level in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 475 ('verbose.level  : silent      # one of silent, helpful, debug, debug-annoying')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "\n",
      "Bad key verbose.fileo in file /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 476 ('verbose.fileo  : sys.stdout  # a log filename, sys.stdout or sys.stderr')\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.3.2/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The text.latex.preview rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The mathtext.fallback_to_cm rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: Support for setting the 'mathtext.fallback_to_cm' rcParam is deprecated since 3.3 and will be removed two minor releases later; use 'mathtext.fallback : 'cm' instead.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The validate_bool_maybe_none function was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The savefig.jpeg_quality rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The keymap.all_axes rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The animation.avconv_path rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/alla/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The animation.avconv_args rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "from collections import OrderedDict, Counter, defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import glob\n",
    "\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "\n",
    "import re\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "import pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook plan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* extraction of extension sequences and transcriptomic (local) coordinates for gencode 25 .fasta files\n",
    "---\n",
    "* local coordinates of 50-codons-long parts of extensions \n",
    "---\n",
    "* global coordinates of extensions/50-codons-long parts \n",
    "---\n",
    "* overlaps of 50-codons with coding exons: gencode25/35/refseq, overlaps frame comparison (separate notebooks)\n",
    "---\n",
    "* add PhyloCSF 120-mammals pipeline (separate scripts most likely)\n",
    "---\n",
    "* save to df with local/global coo/PhyloCSF score/overlaps. Next step is to add Proteomics/Ribo-seq data in separate notebook "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Metadata dataframe for gencode 25/35 releases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extrac_metadata(filepath, pc=True):\n",
    "    if pc==True:\n",
    "        troubles = []\n",
    "        li = []\n",
    "        with open(filepath, \"r\") as handle:\n",
    "            for record in SeqIO.parse(handle, \"fasta\"):\n",
    "                record_id = str(record.id)\n",
    "                tr_id = record_id.split('|')[0]\n",
    "                seq = str(record.seq)\n",
    "                \n",
    "                gene = record_id.split('|')[5]\n",
    "                gene_tr = record_id.split('|')[4]\n",
    "                \n",
    "                # CDS extraction\n",
    "                m_cds = re.finditer(r'(CDS:)[0-9]*-[0-9]*', record_id)\n",
    "                cds = [x.group(0) for x in m_cds][0] #UTR3:955-1054\n",
    "                cds_start = int(cds.split(':')[1].split('-')[0]) - 1\n",
    "                cds_end = int(cds.split(':')[1].split('-')[1]) \n",
    "                cds_seq = seq[cds_start:cds_end]     \n",
    "                cds_start_codon = cds_seq[0:3]\n",
    "                cds_stop_codon = cds_seq[-3:]\n",
    "        \n",
    "                \n",
    "                # UTR5 extraction\n",
    "                # some transcript do not have UTR5 annotated \n",
    "                # but we still want to save such trascripts\n",
    "                m_utr5 = re.finditer(r'(UTR5:)[0-9]*-[0-9]*', record_id)\n",
    "                t = [x.group(0) for x in m_utr5]\n",
    "                \n",
    "                \n",
    "                if t:\n",
    "                    utr5 = t[0] #UTR5:1-1054\n",
    "                    utr5_start = int(utr5.split(':')[1].split('-')[0]) - 1\n",
    "                    utr5_end = int(utr5.split(':')[1].split('-')[1]) \n",
    "        \n",
    "                    if utr5_start == utr5_end: #1=1\n",
    "                        print ('utr5_start == utr5_end', record_id)\n",
    "                        utr5_start_seq = seq[utr5_start-1:(utr5_end)]\n",
    "                    else: \n",
    "                        utr5_start_seq = seq[utr5_start:(utr5_end)]\n",
    "                        \n",
    "                    li.append([tr_id, gene, gene_tr, seq, utr5_start_seq, \n",
    "                   utr5_start_seq[-3:], cds_seq,\n",
    "                  cds_start_codon, cds_stop_codon, cds_start, cds_end,\n",
    "                  utr5_start, utr5_end, record_id])\n",
    "                    \n",
    "                    \n",
    "                else:\n",
    "                    utr5_start = 0\n",
    "                    utr5_end = 0\n",
    "                    utr5_start_seq = None\n",
    "                    cds_start = 0\n",
    "                    \n",
    "                    li.append([tr_id, gene, gene_tr, seq, utr5_start_seq, \n",
    "                   cds_seq[:3], cds_seq,\n",
    "                  cds_start_codon, cds_stop_codon, cds_start, cds_end,\n",
    "                  utr5_start, utr5_end, record_id])\n",
    "        \n",
    "        metadata_pc_g25 = pd.DataFrame(li, columns = ['tr_id', 'gene' ,'gene_tr', \n",
    "                                 'transcript_seq', '5UTR_start_seq', \n",
    "                                 'CDS_start', 'cds_seq', 'cds_start_codon', \n",
    "                                 'cds_stop_codon', 'cds_start_pos',  'cds_stop_pos',\n",
    "                                 'utr5_start', 'utr5_end', 'record_id'])\n",
    "        return (metadata_pc_g25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# trs with 5UTR: 78873 ; # all trs 94359\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_tr</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>cds_seq</th>\n",
       "      <th>cds_start_codon</th>\n",
       "      <th>cds_stop_codon</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>cds_stop_pos</th>\n",
       "      <th>utr5_start</th>\n",
       "      <th>utr5_end</th>\n",
       "      <th>record_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>None</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>FO538757.2-201</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>AAG</td>\n",
       "      <td>ATGAGTGACAGCATCAACTTCTCTCACAACCTAGGCCAGCTCCTGT...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>316</td>\n",
       "      <td>718</td>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>ENST00000624431.1|ENSG00000279928.1|-|-|FO5387...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id        gene         gene_tr  \\\n",
       "0  ENST00000335137.3       OR4F5       OR4F5-001   \n",
       "1  ENST00000624431.1  FO538757.2  FO538757.2-201   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...   \n",
       "1  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "\n",
       "                                      5UTR_start_seq CDS_start  \\\n",
       "0                                               None       ATG   \n",
       "1  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...       AAG   \n",
       "\n",
       "                                             cds_seq cds_start_codon  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...             ATG   \n",
       "1  ATGAGTGACAGCATCAACTTCTCTCACAACCTAGGCCAGCTCCTGT...             ATG   \n",
       "\n",
       "  cds_stop_codon  cds_start_pos  cds_stop_pos  utr5_start  utr5_end  \\\n",
       "0            TAG              0           918           0         0   \n",
       "1            TAG            316           718           0       316   \n",
       "\n",
       "                                           record_id  \n",
       "0  ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...  \n",
       "1  ENST00000624431.1|ENSG00000279928.1|-|-|FO5387...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pc_g25 = extrac_metadata(filepath=\"data/gencode.v25.pc_transcripts.fa\", pc=True)\n",
    "print ('# trs with 5UTR:', metadata_pc_g25[metadata_pc_g25['5UTR_start_seq'].notna()].shape[0], '; # all trs', metadata_pc_g25.shape[0])\n",
    "metadata_pc_g25.to_csv('tmp_res/metadata_pc_g25.txt', sep='\\t', index=False)\n",
    "\n",
    "metadata_pc_g25[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# trs with 5UTR: 85118 ; # all trs 101486\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_tr</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>cds_seq</th>\n",
       "      <th>cds_start_codon</th>\n",
       "      <th>cds_stop_codon</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>cds_stop_pos</th>\n",
       "      <th>utr5_start</th>\n",
       "      <th>utr5_end</th>\n",
       "      <th>record_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5-202</td>\n",
       "      <td>CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...</td>\n",
       "      <td>CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...</td>\n",
       "      <td>GTT</td>\n",
       "      <td>ATGAAGAAGGTAACTGCAGAGGCTATTTCCTGGAATGAATCAACGA...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>60</td>\n",
       "      <td>1041</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>ENST00000641515.2|ENSG00000186092.6|OTTHUMG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000335137.4</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>TCCTGGAATGAATCAACGAGTGAAACGAATAACTCTATGGTGACTG...</td>\n",
       "      <td>TCCTGGAATGAATCAACGAGTGAAACGAATAACTCT</td>\n",
       "      <td>TCT</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>36</td>\n",
       "      <td>954</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>ENST00000335137.4|ENSG00000186092.6|OTTHUMG000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id   gene    gene_tr  \\\n",
       "0  ENST00000641515.2  OR4F5  OR4F5-202   \n",
       "1  ENST00000335137.4  OR4F5  OR4F5-201   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...   \n",
       "1  TCCTGGAATGAATCAACGAGTGAAACGAATAACTCTATGGTGACTG...   \n",
       "\n",
       "                                      5UTR_start_seq CDS_start  \\\n",
       "0  CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...       GTT   \n",
       "1               TCCTGGAATGAATCAACGAGTGAAACGAATAACTCT       TCT   \n",
       "\n",
       "                                             cds_seq cds_start_codon  \\\n",
       "0  ATGAAGAAGGTAACTGCAGAGGCTATTTCCTGGAATGAATCAACGA...             ATG   \n",
       "1  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...             ATG   \n",
       "\n",
       "  cds_stop_codon  cds_start_pos  cds_stop_pos  utr5_start  utr5_end  \\\n",
       "0            TAG             60          1041           0        60   \n",
       "1            TAG             36           954           0        36   \n",
       "\n",
       "                                           record_id  \n",
       "0  ENST00000641515.2|ENSG00000186092.6|OTTHUMG000...  \n",
       "1  ENST00000335137.4|ENSG00000186092.6|OTTHUMG000...  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pc_g35 = extrac_metadata(filepath=\"data/gencode.v35.pc_transcripts.fa\", pc=True)\n",
    "print ('# trs with 5UTR:', metadata_pc_g35[metadata_pc_g35['5UTR_start_seq'].notna()].shape[0], '; # all trs', metadata_pc_g35.shape[0])\n",
    "metadata_pc_g35.to_csv('tmp_res/metadata_pc_g35.txt', sep='\\t', index=False)\n",
    "\n",
    "metadata_pc_g35[0:2]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "metadata_pc_g37 = extrac_metadata(filepath=\"data/gencode.v37.pc_transcripts.fa\", pc=True)\n",
    "print ('# trs with 5UTR:', metadata_pc_g35[metadata_pc_g35['5UTR_start_seq'].notna()].shape[0], '; # all trs', metadata_pc_g35.shape[0])\n",
    "metadata_pc_g37.to_csv('tmp_res/metadata_pc_g37.txt', sep='\\t', index=False)\n",
    "\n",
    "metadata_pc_g37[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# local coordinates of N-terminal-Extensions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(77820, 8)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extracting full NTEs\n",
    "\n",
    "def get_pos_of_1st_stop_inframe(seq):\n",
    "    '''\n",
    "    seq = seq of 5'leader\n",
    "    '''\n",
    "    \n",
    "    # reverse seq\n",
    "    seq_r = seq[::-1]\n",
    "    \n",
    "    i = 0  \n",
    "    li = []\n",
    "    while i <= len(seq_r):    \n",
    "    \n",
    "        codon = seq_r[i:i+3][::-1] # read by codon and reverse it like it's read on forward strand\n",
    "        if (codon == 'TAG') | (codon == 'TGA') | (codon == 'TAA'):\n",
    "            li.append(i) \n",
    "        i = i + 3\n",
    "        \n",
    "    if li:\n",
    "        pos = (np.min(li))\n",
    "        pos_f = len(seq) - pos\n",
    "    else:\n",
    "        pos_f = 0\n",
    "    \n",
    "    return (pos_f)\n",
    "\n",
    "\n",
    "def cut_nt(x):\n",
    "    \n",
    "    if x['len'] % 3 == 0:\n",
    "        x['N_term_start1'] = x['N_term_start']\n",
    "        x['N_term_end1'] = x['N_term_end']\n",
    "        \n",
    "    elif x['len'] %3 == 1:\n",
    "        x['N_term_start1'] = x['N_term_start'] + 1\n",
    "        x['N_term_end1'] = x['N_term_end']\n",
    "        \n",
    "        \n",
    "    elif x['len'] %3 == 2:\n",
    "        x['N_term_start1'] = x['N_term_start'] + 2\n",
    "        x['N_term_end1'] = x['N_term_end']\n",
    "        \n",
    "    else:\n",
    "        pass\n",
    "        \n",
    "    return x    \n",
    "\n",
    "def local_coo_df(metadata):\n",
    "    new_li = []\n",
    "    \n",
    "    for row in metadata.dropna().to_numpy():\n",
    "        tr_seq = row[3]\n",
    "        cds_start = row[9]\n",
    "        cds_end = row[10]\n",
    "        utr5_start = int(row[11])\n",
    "        utr5_end = int(row[12])\n",
    "        tr_id = row[0]\n",
    "        gene = row[1]\n",
    "\n",
    "        seq = tr_seq[utr5_start:utr5_end]\n",
    "        pos_stop = get_pos_of_1st_stop_inframe(seq)\n",
    "    \n",
    "        if (pos_stop is not None):\n",
    "            if (pos_stop != cds_start):\n",
    "                pos_stop_real = pos_stop + utr5_start\n",
    "                # transcriptomic pos of N-terminal extension\n",
    "                tr_pos = [pos_stop_real, cds_start]\n",
    "                new_li.append([tr_id, pos_stop_real, cds_start])\n",
    "        \n",
    "        else:\n",
    "            tr_pos = [utr5_start+0, cds_start]\n",
    "            new_li.append([tr_id, utr5_start+0, cds_start])\n",
    "        \n",
    "    local_coo = pd.DataFrame(new_li, columns = ['tr_id', 'N_term_start', 'N_term_end'])\n",
    "    local_coo['len'] = local_coo['N_term_end'] - local_coo['N_term_start'] \n",
    "    \n",
    "    # check frame: NTE should have length multiple of 3nt = codon\n",
    "    \n",
    "    local_coo = local_coo.apply(cut_nt, axis=1)\n",
    "    local_coo['len_codons'] = (local_coo['N_term_end1'] - local_coo['N_term_start1']) / 3 \n",
    "    local_coo = local_coo.merge(metadata[['tr_id', 'gene']], on='tr_id', how='inner')\n",
    "    local_coo.to_csv('tmp_res/NTE_local_coordinates.txt', sep='\\t', index=False)\n",
    "    \n",
    "    return (local_coo)\n",
    "\n",
    "local_coo = local_coo_df(metadata=metadata_pc_g25)\n",
    "local_coo.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TAA', 'TAG', 'TGA'}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# why local_coo.shape = 77820, and 78873 trs in total have UTR5\n",
    "\n",
    "set_of_closest_to_CDS_codons = set()\n",
    "\n",
    "for el in metadata_pc_g25[(metadata_pc_g25['5UTR_start_seq'].notna()) & (metadata_pc_g25['tr_id'].isin(list(set(metadata_pc_g25['tr_id'].tolist()) - set(local_coo['tr_id'].tolist()))))].to_numpy():\n",
    "    tr = el[0]\n",
    "    utr5_seq = el[4]\n",
    "    \n",
    "    set_of_closest_to_CDS_codons.add(utr5_seq[-3:])\n",
    "\n",
    "# stop codons right before CDS => no NTEs are found \n",
    "set_of_closest_to_CDS_codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Local cordinates for 50-codons-long parts of extensions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_new_NTE_start_50_codons(x):\n",
    "    '''\n",
    "    return N_term_start_50\n",
    "    if there are >50 codons = all 50\n",
    "    if less = all we have \n",
    "    '''\n",
    "    if x['len_codons'] >= 50:\n",
    "        New_nte_start = x['N_term_end1'] - 150  \n",
    "    else:\n",
    "        New_nte_start = x['N_term_start1']\n",
    "        \n",
    "    return New_nte_start\n",
    "        \n",
    "\n",
    "local_coo['50len_flag'] = local_coo.apply(add_new_NTE_start_50_codons, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "local_coo[['tr_id', 'N_term_end1', '50len_flag']].to_csv('tmp_res/local_coo_50codons_and_less.txt', sep='\\t', index=False)\n",
    "local_coo[['tr_id', 'N_term_start1', 'N_term_end1']].to_csv('tmp_res/local_coo_primary_NTE.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Global coordinates "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Separate R-script is used to extract genomic coordinates using .gtf files, GenomicFeatures library: transcriptomic_to_genomic_coordunates_example.R\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_global_coo(path_to_file_with_global_coo, colname):\n",
    "    global_coo_g25 = pd.read_csv(path_to_file_with_global_coo, sep='\\t')\n",
    "    global_coo_g25_f = global_coo_g25[global_coo_g25['hit'] == True]\n",
    "    global_coo_g25_f = global_coo_g25_f.sort_values(by=['seqnames', 'start'])\n",
    "    \n",
    "    global_coo_g25_f['global_coo'] = global_coo_g25_f['seqnames']+':'+global_coo_g25_f['start'].astype(str) +'-'+global_coo_g25_f['end'].astype(str)\n",
    "    global_coo_g25_f = global_coo_g25_f[['group_name', 'strand', 'global_coo']].groupby(['group_name', 'strand']).agg('+'.join).reset_index()\n",
    "    global_coo_g25_f.columns = ['tr_id', 'strand', colname]\n",
    "    return global_coo_g25_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-818a915f4f83>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m primary_global = prepare_global_coo(path_to_file_with_global_coo='tmp_res/primary_NTE_global_coo.txt',\n\u001b[0;32m----> 2\u001b[0;31m                                    colname = 'global_coo_primary')\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m codons50_global = prepare_global_coo(path_to_file_with_global_coo='tmp_res/global_coo_50codons_and_less.txt',\n\u001b[1;32m      5\u001b[0m                                     colname = 'global_coo_50_and_less')\n",
      "\u001b[0;32m<ipython-input-1-7e475adf428f>\u001b[0m in \u001b[0;36mprepare_global_coo\u001b[0;34m(path_to_file_with_global_coo, colname)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mprepare_global_coo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath_to_file_with_global_coo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mglobal_coo_g25\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath_to_file_with_global_coo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mglobal_coo_g25_f\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mglobal_coo_g25\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mglobal_coo_g25\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'hit'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mglobal_coo_g25_f\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mglobal_coo_g25_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'seqnames'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'start'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mglobal_coo_g25_f\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mglobal_coo_g25_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop_duplicates\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "primary_global = prepare_global_coo(path_to_file_with_global_coo='tmp_res/primary_NTE_global_coo.txt',\n",
    "                                   colname = 'global_coo_primary')\n",
    "\n",
    "codons50_global = prepare_global_coo(path_to_file_with_global_coo='tmp_res/global_coo_50codons_and_less.txt',\n",
    "                                    colname = 'global_coo_50_and_less')\n",
    "\n",
    "print ('tr # check:',codons50_global.tr_id.nunique(), primary_global.tr_id.nunique(), local_coo.tr_id.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(77820, 14) 77820\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag global_coo_50_and_less  \\\n",
       "0          316        14.0  FO538757.2         274     chr1:182667-182708   \n",
       "1          236        78.0  FO538757.1          86     chr1:200087-200236   \n",
       "2          504       148.0      SAMD11         354     chr1:924282-924431   \n",
       "\n",
       "  strand  global_coo_primary  \\\n",
       "0      +  chr1:182667-182708   \n",
       "1      -  chr1:200087-200320   \n",
       "2      +  chr1:923988-924431   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...  \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...  \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add global coo to local df\n",
    "\n",
    "local_and_global_df = local_coo.merge(codons50_global[['global_coo_50_and_less', 'tr_id']], \n",
    "                                      on='tr_id', how='inner').merge(primary_global, on='tr_id', how='inner').merge(metadata_pc_g25[['tr_id', 'transcript_seq', '5UTR_start_seq']], on='tr_id', how='inner')\n",
    "\n",
    "\n",
    "print (local_and_global_df.shape, local_and_global_df.tr_id.nunique())\n",
    "\n",
    "local_and_global_df[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "local_and_global_df.to_csv('tmp_res/local_and_global_coo_50_and_full.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exclude cases with coding exons overlaps "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bed(filepath, df):\n",
    "    '''\n",
    "    df: tr_id, strand, coo\n",
    "    '''\n",
    "    f = open(filepath, 'w')\n",
    "\n",
    "    for el in df.to_numpy():\n",
    "        coo = el[2].split('+')\n",
    "        strand = el[1]\n",
    "        tr_id = el[0]\n",
    "    \n",
    "        for coo_ in coo:\n",
    "            chrom = coo_.split(':')[0]\n",
    "            start = coo_.split(':')[1].split('-')[0]\n",
    "            stop = coo_.split(':')[1].split('-')[1]\n",
    "        \n",
    "        \n",
    "            if int(start) - int(stop) == 1:\n",
    "                f.write(chrom+'\\t'+start+'\\t'+str(int(stop)+1)+'\\t'+tr_id+'\\t1\\t'+strand+'\\n') \n",
    "                # where the length of the region is 0 nt\n",
    "            else:\n",
    "                f.write(chrom+'\\t'+start+'\\t'+stop+'\\t'+tr_id+'\\t1\\t'+strand+'\\n')\n",
    "            \n",
    "    f.close()  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "make .bed file of 50-codons-long parts of extensions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_bed(filepath='tmp_res/50_codons_NTE.bed', df=local_and_global_df[['tr_id', 'strand', 'global_coo_50_and_less']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "make .bed file of Full parts of extensions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_bed(filepath='tmp_res/full_NTE.bed', df=local_and_global_df[['tr_id', 'strand', 'global_coo_primary']])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# some magic with exon's numeration \n",
    "\n",
    "f = pd.read_csv('tmp_res/50_codons_NTE.bed', sep='\\t', header=None)\n",
    "f.columns = ['chr_ext', 'start_ext', 'end_ext', 'tr_id_ext', 'exon_ext', 'strand_ext']\n",
    "\n",
    "\n",
    "#tmp1 = f[f['strand_ext'] == '+']\n",
    "\n",
    "res1 = dict(tuple(f.groupby('tr_id_ext')))\n",
    "\n",
    "li = []\n",
    "for k,v in res1.items():\n",
    "    if v.iloc[0].strand_ext == '+':\n",
    "        v1 = v.sort_values(by='start_ext', ascending=True)\n",
    "        v1['exon_ext'] = [i for i in range(1, v1.shape[0]+1)]\n",
    "    else:\n",
    "        v1 = v.sort_values(by='start_ext', ascending=False)\n",
    "        v1['exon_ext'] = [i for i in range(1, v1.shape[0]+1)]\n",
    "    li.append(v1)\n",
    "    \n",
    "out = pd.concat(li)\n",
    "\n",
    "# out['start_ext'] = out['start_ext'] - 1\n",
    "\n",
    "out.to_csv('tmp_res/NTE_50codons_corrected_start.bed', sep='\\t', index=False, header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gencode25"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "calculate global coordinates for coding exons in R script using metadata_pc_g25 file as supply of local coordinates of CDS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "cod_ex_25 = pd.read_csv('tmp_res/coding_exons_g25_global.txt', sep='\\t')\n",
    "cod_ex_25 = cod_ex_25[cod_ex_25['hit'] == True]\n",
    "cod_ex_25[['seqnames', 'start', 'end', 'group_name', 'exon_rank', 'strand']].to_csv('tmp_res/coding_exons_g25_global.bed', sep='\\t', index=False, header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* correct starts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_cod</th>\n",
       "      <th>start_cod</th>\n",
       "      <th>end_cod</th>\n",
       "      <th>tr_id_cod</th>\n",
       "      <th>exon_cod</th>\n",
       "      <th>strand_cod</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182708</td>\n",
       "      <td>182746</td>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_cod  start_cod  end_cod          tr_id_cod  exon_cod strand_cod\n",
       "0    chr1      69090    70008  ENST00000335137.3         1          +\n",
       "1    chr1     182708   182746  ENST00000624431.1         1          +"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cod_ex_25 = pd.read_csv('tmp_res/coding_exons_g25_global.bed', sep='\\t', header=None)\n",
    "cod_ex_25.columns = ['chr_cod', 'start_cod', 'end_cod', 'tr_id_cod', 'exon_cod', 'strand_cod']\n",
    "cod_ex_25['start_cod'] = cod_ex_25['start_cod'] - 1 #correct start now in GWIPs\n",
    "\n",
    "cod_ex_25.to_csv('tmp_res/cod_ex_25_corrected.bed', sep='\\t', index=False, header=None)\n",
    "cod_ex_25[0:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gencode 35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "cod_ex_35 = pd.read_csv('tmp_res/coding_exons_g35_global.txt', sep='\\t')\n",
    "cod_ex_35 = cod_ex_35[cod_ex_35['hit'] == True]\n",
    "cod_ex_35[['seqnames', 'start', 'end', 'group_name', 'exon_rank', 'strand']].to_csv('tmp_res/coding_exons_g35_global.bed', sep='\\t', index=False, header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* correct starts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_cod</th>\n",
       "      <th>start_cod</th>\n",
       "      <th>end_cod</th>\n",
       "      <th>tr_id_cod</th>\n",
       "      <th>exon_cod</th>\n",
       "      <th>strand_cod</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65564</td>\n",
       "      <td>65573</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69036</td>\n",
       "      <td>70008</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_cod  start_cod  end_cod          tr_id_cod  exon_cod strand_cod\n",
       "0    chr1      65564    65573  ENST00000641515.2         2          +\n",
       "1    chr1      69036    70008  ENST00000641515.2         3          +"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cod_ex_35 = pd.read_csv('tmp_res/coding_exons_g35_global.bed', sep='\\t', header=None)\n",
    "cod_ex_35.columns = ['chr_cod', 'start_cod', 'end_cod', 'tr_id_cod', 'exon_cod', 'strand_cod']\n",
    "cod_ex_35['start_cod'] = cod_ex_35['start_cod'] - 1 #correct start now in GWIPs\n",
    "\n",
    "cod_ex_35.to_csv('tmp_res/cod_ex_35_corrected.bed', sep='\\t', index=False, header=None)\n",
    "cod_ex_35[0:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Refseq\n",
    "\n",
    "- one option - use latest .gff file and parse it, extract CDS coordinates (5May, https://www.ncbi.nlm.nih.gov/projects/genome/guide/human/index.shtml) \n",
    "- another option - use UCSC Table Browser (I was not sure that it provides the latest version, so I used 1st option)\n",
    "\n",
    "- GRCh38_RefSeq2UCSC.txt: chromosome ids conversion file, https://github.com/dpryan79/ChromosomeMappings/blob/master/GRCh38_RefSeq2UCSC.txt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "list index out of range\n",
      "{'NM', 'XM'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seqid   chr\n",
       "0  NC_000001.11  chr1\n",
       "1  NC_000002.12  chr2"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys                                                                         \n",
    "from gffutils.iterators import DataIterator                                        \n",
    "\n",
    "li = []\n",
    "\n",
    "# extract gene and \n",
    "                                                                                \n",
    "for feature in DataIterator('data/GRCh38_latest_genomic.gff'):                                                                                                 \n",
    "    if feature.featuretype == \"CDS\":   \n",
    "        #print (feature.seqid, feature.featuretype, len(feature), feature.chrom, feature.start, feature.stop, feature.end, feature.strand)\n",
    "        try:\n",
    "            #print (feature.attributes['Parent'])\n",
    "            li.append([feature.attributes['Parent'][0].split('rna-')[1], feature.attributes['gene'][0], feature.seqid, feature.featuretype, feature.chrom, feature.start, feature.stop, feature.end, feature.strand])\n",
    "        except Exception as e:\n",
    "            print (e)\n",
    "df = pd.DataFrame(li, columns = ['tr_id', 'gene', 'seqid', 'type', 'chrom', 'start', 'stop', 'end', 'strand'])\n",
    "df[0:4]\n",
    "\n",
    "\n",
    "print (set([x.split('_')[0] for x in df['tr_id'].tolist()]))\n",
    "\n",
    "chr_map = pd.read_csv('data/GRCh38_RefSeq2UCSC.txt', sep='\\t', header=None)\n",
    "chr_map.columns = ['seqid', 'chr']\n",
    "chr_map[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1514439, 10) (1514439, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>seqid</th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>65573</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69037</td>\n",
       "      <td>70008</td>\n",
       "      <td>70008</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tr_id   gene         seqid type         chrom  start   stop  \\\n",
       "0  NM_001005484.2  OR4F5  NC_000001.11  CDS  NC_000001.11  65565  65573   \n",
       "1  NM_001005484.2  OR4F5  NC_000001.11  CDS  NC_000001.11  69037  70008   \n",
       "\n",
       "     end strand  \n",
       "0  65573      +  \n",
       "1  70008      +  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df.merge(chr_map, on='seqid', how='left')\n",
    "print (df2.shape, df.shape)\n",
    "df[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq = df2[(df2['tr_id'].str.contains('NM_')) & (df2['chr'].isin(['chr'+str(i) for i in range(1, 23)]+['chrX', 'chrY']))][['chr', 'start', 'end', 'tr_id', 'strand', 'gene']]\n",
    "refseq['tmp'] = 1\n",
    "refseq[['chr', 'start', 'end', 'tr_id', 'tmp', 'strand']].to_csv('tmp_res/refseq_coding_exons_chr_NM.bed', sep='\\t', index=False, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gencode 38 \n",
    "\n",
    "May5, for some reason, there is gencode38 release in Hiller lab Genome browser mirror. Why not? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67093004</td>\n",
       "      <td>67093604</td>\n",
       "      <td>ENST00000684719.1_cds_0_0_chr1_67093005_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67095234</td>\n",
       "      <td>67095421</td>\n",
       "      <td>ENST00000684719.1_cds_1_0_chr1_67095235_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0         1         2                                          3  4  5\n",
       "0  chr1  67093004  67093604  ENST00000684719.1_cds_0_0_chr1_67093005_r  0  -\n",
       "1  chr1  67095234  67095421  ENST00000684719.1_cds_1_0_chr1_67095235_r  0  -"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_g38_CDS = pd.read_csv('data/Gencode38_HillerLab_table_browser_CDS.bed', sep='\\t', header=None)\n",
    "bed_g38_CDS[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlap between theoretical extensions & coding exons of gencode25"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* bedtools sort -i tmp_res/cod_ex_25_corrected.bed > tmp_res/cod_ex_25_corrected_sorted.bed\n",
    "\n",
    "* bedtools sort -i tmp_res/full_NTE.bed > tmp_res/full_NTE_sorted.bed\n",
    "\n",
    "* bedtools intersect -a tmp_res/full_NTE_sorted.bed -b tmp_res/cod_ex_25_corrected_sorted.bed -wao -s > tmp_res/NTE_full_start_with_all_coding_exons_g25.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "* bedtools sort -i tmp_res/cod_ex_25_corrected.bed > tmp_res/cod_ex_25_corrected_sorted.bed\n",
    "\n",
    "* bedtools sort -i tmp_res/full_NTE.bed > tmp_res/full_NTE_sorted.bed\n",
    "\n",
    "* bedtools intersect -a tmp_res/full_NTE_sorted.bed -b tmp_res/cod_ex_25_corrected_sorted.bed -wao -s > tmp_res/NTE_full_start_with_all_coding_exons_g25.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse it "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 15) (77820, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less strand  \\\n",
       "0                     chr1:182667-182708      +   \n",
       "1                     chr1:200087-200236      -   \n",
       "2                     chr1:924282-924431      +   \n",
       "3  chr1:925150-925189+chr1:925922-925941      +   \n",
       "4  chr1:925740-925800+chr1:925922-925941      +   \n",
       "\n",
       "                      global_coo_primary  \\\n",
       "0                     chr1:182667-182708   \n",
       "1                     chr1:200087-200320   \n",
       "2                     chr1:923988-924431   \n",
       "3  chr1:925150-925189+chr1:925922-925941   \n",
       "4  chr1:925740-925800+chr1:925922-925941   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "   max_sum_overlap_g25_all_exons  \n",
       "0                              0  \n",
       "1                              0  \n",
       "2                              0  \n",
       "3                             19  \n",
       "4                             19  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_g25 = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_g25.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_g25.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, 'tr_id_cod', 10, 11, 'max_overlap_g25_all_exons']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_g25_sum = overlap_all_exons_g25.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_g25_all_exons'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_g25_sum_max = overlap_all_exons_g25_sum.loc[overlap_all_exons_g25_sum.groupby('tr_id_ext')['max_overlap_g25_all_exons'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_g25_sum_max.tr_id_ext.nunique(), overlap_all_exons_g25_sum_max.shape)\n",
    "\n",
    "overlap_all_exons_g25_sum_max.columns = ['tr_id', 'tr_id_cod', 'max_sum_overlap_g25_all_exons']\n",
    "\n",
    "#overlap_all_exons_g25_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_g25ovlp = local_and_global_df.merge(overlap_all_exons_g25_sum_max[['tr_id', 'max_sum_overlap_g25_all_exons']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_g25ovlp.shape, local_and_global_df.shape)\n",
    "\n",
    "local_and_global_df_g25ovlp[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Well, for the sake of DHFR gene for which there are ovlps with coding exons in another strand... "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* bedtools intersect -a tmp_res/full_NTE_sorted.bed -b tmp_res/cod_ex_25_corrected_sorted.bed -wao > tmp_res/NTE_full_start_with_all_coding_exons_g25_strandless.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 16) (77820, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less strand  \\\n",
       "0                     chr1:182667-182708      +   \n",
       "1                     chr1:200087-200236      -   \n",
       "2                     chr1:924282-924431      +   \n",
       "3  chr1:925150-925189+chr1:925922-925941      +   \n",
       "4  chr1:925740-925800+chr1:925922-925941      +   \n",
       "\n",
       "                      global_coo_primary  \\\n",
       "0                     chr1:182667-182708   \n",
       "1                     chr1:200087-200320   \n",
       "2                     chr1:923988-924431   \n",
       "3  chr1:925150-925189+chr1:925922-925941   \n",
       "4  chr1:925740-925800+chr1:925922-925941   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "   max_sum_overlap_g25_all_exons  max_overlap_g25_all_exons_strandless  \n",
       "0                              0                                     0  \n",
       "1                              0                                     0  \n",
       "2                              0                                     0  \n",
       "3                             19                                    19  \n",
       "4                             19                                    19  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_g25_s = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_g25_strandless.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_g25_s.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, 'tr_id_cod', 10, 11, 'max_overlap_g25_all_exons_strandless']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_g25_sum_s = overlap_all_exons_g25_s.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_g25_all_exons_strandless'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_g25_sum_max_s = overlap_all_exons_g25_sum_s.loc[overlap_all_exons_g25_sum_s.groupby('tr_id_ext')['max_overlap_g25_all_exons_strandless'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_g25_sum_max_s.tr_id_ext.nunique(), overlap_all_exons_g25_sum_max_s.shape)\n",
    "\n",
    "overlap_all_exons_g25_sum_max_s.columns = ['tr_id', 'tr_id_cod', 'max_overlap_g25_all_exons_strandless']\n",
    "\n",
    "#overlap_all_exons_g25_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_g25ovlp_s = local_and_global_df_g25ovlp.merge(overlap_all_exons_g25_sum_max_s[['tr_id', 'max_overlap_g25_all_exons_strandless']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_g25ovlp_s.shape, local_and_global_df_g25ovlp_s.shape)\n",
    "\n",
    "local_and_global_df_g25ovlp_s[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20599</th>\n",
       "      <td>ENST00000439211.6</td>\n",
       "      <td>0</td>\n",
       "      <td>494</td>\n",
       "      <td>494</td>\n",
       "      <td>2</td>\n",
       "      <td>494</td>\n",
       "      <td>164.0</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>344</td>\n",
       "      <td>chr5:80654490-80654639</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5:80654490-80654981</td>\n",
       "      <td>AGTCCCAGACAGAACCTACTATGTGCGGCGGCAGCTGGGGCGGGAA...</td>\n",
       "      <td>AGTCCCAGACAGAACCTACTATGTGCGGCGGCAGCTGGGGCGGGAA...</td>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "20599  ENST00000439211.6             0         494  494              2   \n",
       "\n",
       "       N_term_end1  len_codons  gene  50len_flag  global_coo_50_and_less  \\\n",
       "20599          494       164.0  DHFR         344  chr5:80654490-80654639   \n",
       "\n",
       "      strand      global_coo_primary  \\\n",
       "20599      -  chr5:80654490-80654981   \n",
       "\n",
       "                                          transcript_seq  \\\n",
       "20599  AGTCCCAGACAGAACCTACTATGTGCGGCGGCAGCTGGGGCGGGAA...   \n",
       "\n",
       "                                          5UTR_start_seq  \\\n",
       "20599  AGTCCCAGACAGAACCTACTATGTGCGGCGGCAGCTGGGGCGGGAA...   \n",
       "\n",
       "       max_sum_overlap_g25_all_exons  max_overlap_g25_all_exons_strandless  \n",
       "20599                              0                                   237  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "local_and_global_df_g25ovlp_s[local_and_global_df_g25ovlp_s['tr_id'] == 'ENST00000439211.6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlap between theoretical extensions & coding exons of gencode35"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "bedtools sort -i tmp_res/cod_ex_35_corrected.bed > tmp_res/cod_ex_35_corrected_sorted.bed\n",
    "\n",
    "bedtools sort -i tmp_res/full_NTE.bed > tmp_res/full_NTE_sorted.bed\n",
    "\n",
    "bedtools intersect -a tmp_res/full_NTE_sorted.bed -b tmp_res/cod_ex_35_corrected_sorted.bed -wao -s > tmp_res/NTE_full_start_with_all_coding_exons_g35.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### parse it "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 17) (77820, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g35_all_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less strand  \\\n",
       "0                     chr1:182667-182708      +   \n",
       "1                     chr1:200087-200236      -   \n",
       "2                     chr1:924282-924431      +   \n",
       "3  chr1:925150-925189+chr1:925922-925941      +   \n",
       "4  chr1:925740-925800+chr1:925922-925941      +   \n",
       "\n",
       "                      global_coo_primary  \\\n",
       "0                     chr1:182667-182708   \n",
       "1                     chr1:200087-200320   \n",
       "2                     chr1:923988-924431   \n",
       "3  chr1:925150-925189+chr1:925922-925941   \n",
       "4  chr1:925740-925800+chr1:925922-925941   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "   max_sum_overlap_g25_all_exons  max_overlap_g25_all_exons_strandless  \\\n",
       "0                              0                                     0   \n",
       "1                              0                                     0   \n",
       "2                              0                                     0   \n",
       "3                             19                                    19   \n",
       "4                             19                                    19   \n",
       "\n",
       "   max_sum_overlap_g35_all_exons  \n",
       "0                              0  \n",
       "1                              0  \n",
       "2                              0  \n",
       "3                             19  \n",
       "4                             19  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_g35 = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_g35.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_g35.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, 'tr_id_cod', 10, 11, 'max_overlap_g35_all_exons']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_g35_sum = overlap_all_exons_g35.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_g35_all_exons'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_g35_sum_max = overlap_all_exons_g35_sum.loc[overlap_all_exons_g35_sum.groupby('tr_id_ext')['max_overlap_g35_all_exons'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_g35_sum_max.tr_id_ext.nunique(), overlap_all_exons_g35_sum_max.shape)\n",
    "\n",
    "overlap_all_exons_g35_sum_max.columns = ['tr_id', 'tr_id_cod', 'max_sum_overlap_g35_all_exons']\n",
    "\n",
    "#overlap_all_exons_g35_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_g35ovlp = local_and_global_df_g25ovlp_s.merge(overlap_all_exons_g35_sum_max[['tr_id', 'max_sum_overlap_g35_all_exons']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_g35ovlp.shape, local_and_global_df.shape)\n",
    "\n",
    "local_and_global_df_g35ovlp[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gencode35 strandless "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "bedtools intersect -a tmp_res/full_NTE_sorted.bed -b tmp_res/cod_ex_35_corrected_sorted.bed -wao > tmp_res/NTE_full_start_with_all_coding_exons_g35_strandless.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 18) (77820, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g35_all_exons</th>\n",
       "      <th>max_sum_overlap_g35_all_exons_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less strand  \\\n",
       "0                     chr1:182667-182708      +   \n",
       "1                     chr1:200087-200236      -   \n",
       "2                     chr1:924282-924431      +   \n",
       "3  chr1:925150-925189+chr1:925922-925941      +   \n",
       "4  chr1:925740-925800+chr1:925922-925941      +   \n",
       "\n",
       "                      global_coo_primary  \\\n",
       "0                     chr1:182667-182708   \n",
       "1                     chr1:200087-200320   \n",
       "2                     chr1:923988-924431   \n",
       "3  chr1:925150-925189+chr1:925922-925941   \n",
       "4  chr1:925740-925800+chr1:925922-925941   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "   max_sum_overlap_g25_all_exons  max_overlap_g25_all_exons_strandless  \\\n",
       "0                              0                                     0   \n",
       "1                              0                                     0   \n",
       "2                              0                                     0   \n",
       "3                             19                                    19   \n",
       "4                             19                                    19   \n",
       "\n",
       "   max_sum_overlap_g35_all_exons  max_sum_overlap_g35_all_exons_strandless  \n",
       "0                              0                                         0  \n",
       "1                              0                                         0  \n",
       "2                              0                                         0  \n",
       "3                             19                                        19  \n",
       "4                             19                                        19  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_g35_s = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_g35_strandless.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_g35_s.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, 'tr_id_cod', 10, 11, \n",
    "                                   'max_overlap_g35_all_exons_strandless']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_g35_sum_s = overlap_all_exons_g35_s.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_g35_all_exons_strandless'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_g35_sum_max_s = overlap_all_exons_g35_sum_s.loc[overlap_all_exons_g35_sum_s.groupby('tr_id_ext')['max_overlap_g35_all_exons_strandless'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_g35_sum_max_s.tr_id_ext.nunique(), overlap_all_exons_g35_sum_max_s.shape)\n",
    "\n",
    "overlap_all_exons_g35_sum_max_s.columns = ['tr_id', 'tr_id_cod', 'max_sum_overlap_g35_all_exons_strandless']\n",
    "\n",
    "#overlap_all_exons_g35_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_g35ovlp_s = local_and_global_df_g35ovlp.merge(overlap_all_exons_g35_sum_max_s[['tr_id', 'max_sum_overlap_g35_all_exons_strandless']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_g35ovlp_s.shape, local_and_global_df.shape)\n",
    "\n",
    "local_and_global_df_g35ovlp_s[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlap between theoretical extensions & coding exons of gencode38"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "bedtools sort -i data/Gencode38_HillerLab_table_browser_CDS.bed > data/Gencode38_HillerLab_table_browser_CDS_sorted.bed\n",
    "\n",
    "bedtools sort -i tmp_res/full_NTE.bed > tmp_res/full_NTE_sorted.bed\n",
    "\n",
    "bedtools intersect -a tmp_res/full_NTE_sorted.bed -b data/Gencode38_HillerLab_table_browser_CDS_sorted.bed -wao -s > tmp_res/NTE_full_start_with_all_coding_exons_g38.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### parse it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 19) (77820, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g35_all_exons</th>\n",
       "      <th>max_sum_overlap_g35_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g38_all_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less strand  \\\n",
       "0                     chr1:182667-182708      +   \n",
       "1                     chr1:200087-200236      -   \n",
       "2                     chr1:924282-924431      +   \n",
       "3  chr1:925150-925189+chr1:925922-925941      +   \n",
       "4  chr1:925740-925800+chr1:925922-925941      +   \n",
       "\n",
       "                      global_coo_primary  \\\n",
       "0                     chr1:182667-182708   \n",
       "1                     chr1:200087-200320   \n",
       "2                     chr1:923988-924431   \n",
       "3  chr1:925150-925189+chr1:925922-925941   \n",
       "4  chr1:925740-925800+chr1:925922-925941   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "   max_sum_overlap_g25_all_exons  max_overlap_g25_all_exons_strandless  \\\n",
       "0                              0                                     0   \n",
       "1                              0                                     0   \n",
       "2                              0                                     0   \n",
       "3                             19                                    19   \n",
       "4                             19                                    19   \n",
       "\n",
       "   max_sum_overlap_g35_all_exons  max_sum_overlap_g35_all_exons_strandless  \\\n",
       "0                              0                                         0   \n",
       "1                              0                                         0   \n",
       "2                              0                                         0   \n",
       "3                             19                                        19   \n",
       "4                             19                                        19   \n",
       "\n",
       "   max_sum_overlap_g38_all_exons  \n",
       "0                              0  \n",
       "1                              0  \n",
       "2                              0  \n",
       "3                             19  \n",
       "4                             19  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_g38 = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_g38.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_g38.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, 'tr_id_cod', 10, 11, 'max_overlap_g38_all_exons']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_g38_sum = overlap_all_exons_g38.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_g38_all_exons'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_g38_sum_max = overlap_all_exons_g38_sum.loc[overlap_all_exons_g38_sum.groupby('tr_id_ext')['max_overlap_g38_all_exons'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_g38_sum_max.tr_id_ext.nunique(), overlap_all_exons_g38_sum_max.shape)\n",
    "\n",
    "overlap_all_exons_g38_sum_max.columns = ['tr_id', 'tr_id_cod', 'max_sum_overlap_g38_all_exons']\n",
    "\n",
    "#overlap_all_exons_g35_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_g38ovlp = local_and_global_df_g35ovlp_s.merge(overlap_all_exons_g38_sum_max[['tr_id', 'max_sum_overlap_g38_all_exons']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_g38ovlp.shape, local_and_global_df.shape)\n",
    "\n",
    "local_and_global_df_g38ovlp[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Strandless gencode38"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "bedtools intersect -a tmp_res/full_NTE_sorted.bed -b data/Gencode38_HillerLab_table_browser_CDS_sorted.bed -wao > tmp_res/NTE_full_start_with_all_coding_exons_g38_strandless.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 20) (77820, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g35_all_exons</th>\n",
       "      <th>max_sum_overlap_g35_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g38_all_exons</th>\n",
       "      <th>max_sum_overlap_g38_all_exons_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less strand  \\\n",
       "0                     chr1:182667-182708      +   \n",
       "1                     chr1:200087-200236      -   \n",
       "2                     chr1:924282-924431      +   \n",
       "3  chr1:925150-925189+chr1:925922-925941      +   \n",
       "4  chr1:925740-925800+chr1:925922-925941      +   \n",
       "\n",
       "                      global_coo_primary  \\\n",
       "0                     chr1:182667-182708   \n",
       "1                     chr1:200087-200320   \n",
       "2                     chr1:923988-924431   \n",
       "3  chr1:925150-925189+chr1:925922-925941   \n",
       "4  chr1:925740-925800+chr1:925922-925941   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "   max_sum_overlap_g25_all_exons  max_overlap_g25_all_exons_strandless  \\\n",
       "0                              0                                     0   \n",
       "1                              0                                     0   \n",
       "2                              0                                     0   \n",
       "3                             19                                    19   \n",
       "4                             19                                    19   \n",
       "\n",
       "   max_sum_overlap_g35_all_exons  max_sum_overlap_g35_all_exons_strandless  \\\n",
       "0                              0                                         0   \n",
       "1                              0                                         0   \n",
       "2                              0                                         0   \n",
       "3                             19                                        19   \n",
       "4                             19                                        19   \n",
       "\n",
       "   max_sum_overlap_g38_all_exons  max_sum_overlap_g38_all_exons_strandless  \n",
       "0                              0                                         0  \n",
       "1                              0                                         0  \n",
       "2                              0                                         0  \n",
       "3                             19                                        19  \n",
       "4                             19                                        19  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_g38_s = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_g38_strandless.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_g38_s.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, 'tr_id_cod', \n",
    "                                 10, 11, 'max_overlap_g38_all_exons_strandless']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_g38_sum_s = overlap_all_exons_g38_s.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_g38_all_exons_strandless'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_g38_sum_max_s = overlap_all_exons_g38_sum_s.loc[overlap_all_exons_g38_sum_s.groupby('tr_id_ext')['max_overlap_g38_all_exons_strandless'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_g38_sum_max_s.tr_id_ext.nunique(), overlap_all_exons_g38_sum_max_s.shape)\n",
    "\n",
    "overlap_all_exons_g38_sum_max_s.columns = ['tr_id', 'tr_id_cod', 'max_sum_overlap_g38_all_exons_strandless']\n",
    "\n",
    "#overlap_all_exons_g35_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_g38ovlp_s = local_and_global_df_g38ovlp.merge(overlap_all_exons_g38_sum_max_s[['tr_id', 'max_sum_overlap_g38_all_exons_strandless']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_g38ovlp_s.shape, local_and_global_df.shape)\n",
    "\n",
    "local_and_global_df_g38ovlp_s[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlap between theoretical extensions & coding exons of refseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_cod</th>\n",
       "      <th>start_cod</th>\n",
       "      <th>end_cod</th>\n",
       "      <th>tr_id_cod</th>\n",
       "      <th>exon_cod</th>\n",
       "      <th>strand_cod</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69037</td>\n",
       "      <td>70008</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_cod  start_cod  end_cod       tr_id_cod  exon_cod strand_cod\n",
       "0    chr1      65565    65573  NM_001005484.2         1          +\n",
       "1    chr1      69037    70008  NM_001005484.2         1          +"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_cod_ex = pd.read_csv('tmp_res/refseq_coding_exons_chr_NM.bed', sep='\\t', header=None)\n",
    "\n",
    "refseq_cod_ex.columns = ['chr_cod', 'start_cod', 'end_cod', 'tr_id_cod', 'exon_cod', 'strand_cod']\n",
    "\n",
    "refseq_cod_ex[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "res1 = dict(tuple(refseq_cod_ex.groupby('tr_id_cod')))\n",
    "\n",
    "li = []\n",
    "for k,v in res1.items():\n",
    "    if v.iloc[0].strand_cod == '+':\n",
    "        v1 = v.sort_values(by='start_cod', ascending=True)\n",
    "        v1['exon_cod'] = [i for i in range(1, v1.shape[0]+1)]\n",
    "    else:\n",
    "        v1 = v.sort_values(by='start_cod', ascending=False)\n",
    "        v1['exon_cod'] = [i for i in range(1, v1.shape[0]+1)]\n",
    "    li.append(v1)\n",
    "    \n",
    "refseq_cod_ex = pd.concat(li)\n",
    "\n",
    "refseq_cod_ex['start_cod'] = refseq_cod_ex['start_cod'] - 1\n",
    "\n",
    "refseq_cod_ex = refseq_cod_ex.sort_values(by=['tr_id_cod', 'exon_cod'])\n",
    "\n",
    "refseq_cod_ex.to_csv('tmp_res/RefSeq_cod_ex_corrected_start.bed', sep='\\t', index=False, header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "bedtools sort -i tmp_res/RefSeq_cod_ex_corrected_start.bed > tmp_res/RefSeq_cod_ex_corrected_start_sorted.bed\n",
    "\n",
    "bedtools sort -i tmp_res/full_NTE.bed > tmp_res/full_NTE_sorted.bed\n",
    "\n",
    "bedtools intersect -a tmp_res/full_NTE_sorted.bed -b tmp_res/RefSeq_cod_ex_corrected_start_sorted.bed -wao -s > tmp_res/NTE_full_start_with_all_coding_exons_refseq.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 21) (77820, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>...</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g35_all_exons</th>\n",
       "      <th>max_sum_overlap_g35_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g38_all_exons</th>\n",
       "      <th>max_sum_overlap_g38_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_refseq_all_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:200087-200320</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:923988-924431</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less  ...  \\\n",
       "0                     chr1:182667-182708  ...   \n",
       "1                     chr1:200087-200236  ...   \n",
       "2                     chr1:924282-924431  ...   \n",
       "3  chr1:925150-925189+chr1:925922-925941  ...   \n",
       "4  chr1:925740-925800+chr1:925922-925941  ...   \n",
       "\n",
       "                      global_coo_primary  \\\n",
       "0                     chr1:182667-182708   \n",
       "1                     chr1:200087-200320   \n",
       "2                     chr1:923988-924431   \n",
       "3  chr1:925150-925189+chr1:925922-925941   \n",
       "4  chr1:925740-925800+chr1:925922-925941   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "  max_sum_overlap_g25_all_exons  max_overlap_g25_all_exons_strandless  \\\n",
       "0                             0                                     0   \n",
       "1                             0                                     0   \n",
       "2                             0                                     0   \n",
       "3                            19                                    19   \n",
       "4                            19                                    19   \n",
       "\n",
       "   max_sum_overlap_g35_all_exons  max_sum_overlap_g35_all_exons_strandless  \\\n",
       "0                              0                                         0   \n",
       "1                              0                                         0   \n",
       "2                              0                                         0   \n",
       "3                             19                                        19   \n",
       "4                             19                                        19   \n",
       "\n",
       "   max_sum_overlap_g38_all_exons  max_sum_overlap_g38_all_exons_strandless  \\\n",
       "0                              0                                         0   \n",
       "1                              0                                         0   \n",
       "2                              0                                         0   \n",
       "3                             19                                        19   \n",
       "4                             19                                        19   \n",
       "\n",
       "   max_sum_overlap_refseq_all_exons  \n",
       "0                                 0  \n",
       "1                                 0  \n",
       "2                                 0  \n",
       "3                                19  \n",
       "4                                19  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_refseq = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_refseq.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_refseq.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, 'tr_id_cod', 10, 11, 'max_overlap_refseq_all_exons']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_refseq_sum = overlap_all_exons_refseq.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_refseq_all_exons'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_refseq_sum_max = overlap_all_exons_refseq_sum.loc[overlap_all_exons_refseq_sum.groupby('tr_id_ext')['max_overlap_refseq_all_exons'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_refseq_sum_max.tr_id_ext.nunique(), overlap_all_exons_refseq_sum_max.shape)\n",
    "\n",
    "overlap_all_exons_refseq_sum_max.columns = ['tr_id', 'tr_id_cod', 'max_sum_overlap_refseq_all_exons']\n",
    "\n",
    "#overlap_all_exons_refseq_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_all_ovlp_refseq = local_and_global_df_g38ovlp_s.merge(overlap_all_exons_refseq_sum_max[['tr_id', 'max_sum_overlap_refseq_all_exons']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_all_ovlp_refseq.shape, local_and_global_df.shape)\n",
    "\n",
    "local_and_global_df_all_ovlp_refseq[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Strandless Refseq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "bedtools intersect -a tmp_res/full_NTE_sorted.bed -b tmp_res/RefSeq_cod_ex_corrected_start_sorted.bed -wao > tmp_res/NTE_full_start_with_all_coding_exons_refseq_strandless.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77820 (77820, 3)\n",
      "(77820, 22) (77820, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>max_sum_overlap_g25_all_exons</th>\n",
       "      <th>max_overlap_g25_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g35_all_exons</th>\n",
       "      <th>max_sum_overlap_g35_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g38_all_exons</th>\n",
       "      <th>max_sum_overlap_g38_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_refseq_all_exons</th>\n",
       "      <th>max_sum_overlap_refseq_all_exons_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>444</td>\n",
       "      <td>60</td>\n",
       "      <td>504</td>\n",
       "      <td>148.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>354</td>\n",
       "      <td>chr1:924282-924431</td>\n",
       "      <td>...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:925150-925189+chr1:925922-925941</td>\n",
       "      <td>...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:925740-925800+chr1:925922-925941</td>\n",
       "      <td>...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "2  ENST00000420190.6            60         504  444             60   \n",
       "3  ENST00000437963.5             0          60   60              0   \n",
       "4  ENST00000342066.7             0          83   83              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag  \\\n",
       "0          316        14.0  FO538757.2         274   \n",
       "1          236        78.0  FO538757.1          86   \n",
       "2          504       148.0      SAMD11         354   \n",
       "3           60        20.0      SAMD11           0   \n",
       "4           83        27.0      SAMD11           2   \n",
       "\n",
       "                  global_coo_50_and_less  ...  \\\n",
       "0                     chr1:182667-182708  ...   \n",
       "1                     chr1:200087-200236  ...   \n",
       "2                     chr1:924282-924431  ...   \n",
       "3  chr1:925150-925189+chr1:925922-925941  ...   \n",
       "4  chr1:925740-925800+chr1:925922-925941  ...   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "                                      5UTR_start_seq  \\\n",
       "0  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1  GGAAAGCGGGTCAAGGCGTAGGGCTGGAGGGCAGGGGCGGGCCCTG...   \n",
       "2  CGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGC...   \n",
       "3  CAGCGCTTGGGGCTCGCGGGCCGCTCCCTCCGCTCGGAAGGGAAAA...   \n",
       "4  CCAGCAGATCCCTGCGGCGTTCGCGAGGGTGGGACGGGAAGCGGGC...   \n",
       "\n",
       "  max_sum_overlap_g25_all_exons max_overlap_g25_all_exons_strandless  \\\n",
       "0                             0                                    0   \n",
       "1                             0                                    0   \n",
       "2                             0                                    0   \n",
       "3                            19                                   19   \n",
       "4                            19                                   19   \n",
       "\n",
       "   max_sum_overlap_g35_all_exons  max_sum_overlap_g35_all_exons_strandless  \\\n",
       "0                              0                                         0   \n",
       "1                              0                                         0   \n",
       "2                              0                                         0   \n",
       "3                             19                                        19   \n",
       "4                             19                                        19   \n",
       "\n",
       "   max_sum_overlap_g38_all_exons  max_sum_overlap_g38_all_exons_strandless  \\\n",
       "0                              0                                         0   \n",
       "1                              0                                         0   \n",
       "2                              0                                         0   \n",
       "3                             19                                        19   \n",
       "4                             19                                        19   \n",
       "\n",
       "   max_sum_overlap_refseq_all_exons  \\\n",
       "0                                 0   \n",
       "1                                 0   \n",
       "2                                 0   \n",
       "3                                19   \n",
       "4                                19   \n",
       "\n",
       "   max_sum_overlap_refseq_all_exons_strandless  \n",
       "0                                            0  \n",
       "1                                            0  \n",
       "2                                            0  \n",
       "3                                           19  \n",
       "4                                           19  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_refseq_s = pd.read_csv('tmp_res/NTE_full_start_with_all_coding_exons_refseq_strandless.bed', sep='\\t', header=None) \n",
    "overlap_all_exons_refseq_s.columns = [0, 1, 2, 'tr_id_ext', 4, 5, 6, 7, 8, \n",
    "                                    'tr_id_cod', 10, 11, 'max_overlap_refseq_all_exons_strandless']\n",
    "\n",
    "# find sum overlap per pair of transcript with extension and transcript with coding exons \n",
    "#overlap_all_exons_g25_ = overlap_all_exons_g25[overlap_all_exons_g25[10] != -1]\n",
    "overlap_all_exons_refseq_sum_s = overlap_all_exons_refseq_s.groupby(by=['tr_id_ext', 'tr_id_cod'])['max_overlap_refseq_all_exons_strandless'].sum().reset_index()\n",
    "\n",
    "# find max overall (regardkess of frame)\n",
    "overlap_all_exons_refseq_sum_max_s = overlap_all_exons_refseq_sum_s.loc[overlap_all_exons_refseq_sum_s.groupby('tr_id_ext')['max_overlap_refseq_all_exons_strandless'].idxmax()]\n",
    "\n",
    "print (overlap_all_exons_refseq_sum_max_s.tr_id_ext.nunique(), overlap_all_exons_refseq_sum_max_s.shape)\n",
    "\n",
    "overlap_all_exons_refseq_sum_max_s.columns = ['tr_id', 'tr_id_cod', 'max_sum_overlap_refseq_all_exons_strandless']\n",
    "\n",
    "#overlap_all_exons_refseq_sum_max[0:3]\n",
    "\n",
    "local_and_global_df_all_ovlp = local_and_global_df_all_ovlp_refseq.merge(overlap_all_exons_refseq_sum_max_s[['tr_id', 'max_sum_overlap_refseq_all_exons_strandless']], on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_all_ovlp.shape, local_and_global_df.shape)\n",
    "\n",
    "local_and_global_df_all_ovlp[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ENST00000439211.6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>tr_id_ext</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>tr_id_cod</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>max_overlap_refseq_all_exons_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133334</th>\n",
       "      <td>chr5</td>\n",
       "      <td>80654490</td>\n",
       "      <td>80654981</td>\n",
       "      <td>ENST00000439211.6</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5</td>\n",
       "      <td>80654727</td>\n",
       "      <td>80654964</td>\n",
       "      <td>NM_002439.5</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2          tr_id_ext  4  5     6         7  \\\n",
       "133334  chr5  80654490  80654981  ENST00000439211.6  1  -  chr5  80654727   \n",
       "\n",
       "               8    tr_id_cod  10 11  max_overlap_refseq_all_exons_strandless  \n",
       "133334  80654964  NM_002439.5   1  +                                      237  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_refseq_s[overlap_all_exons_refseq_s['tr_id_ext'] == 'ENST00000439211.6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>tr_id_ext</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>tr_id_cod</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>max_overlap_refseq_all_exons_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [0, 1, 2, tr_id_ext, 4, 5, 6, 7, 8, tr_id_cod, 10, 11, max_overlap_refseq_all_exons_strandless]\n",
       "Index: []"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_all_exons_refseq_s[overlap_all_exons_refseq_s['tr_id_cod'] == '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combine overlaps into 1 column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>ovlp</th>\n",
       "      <th>ovlp2</th>\n",
       "      <th>ovlp_strandless</th>\n",
       "      <th>ovlp2_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000420190.6</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>0;0;0;0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000437963.5</td>\n",
       "      <td>19;19;19</td>\n",
       "      <td>19;19;19;19</td>\n",
       "      <td>19;19;19</td>\n",
       "      <td>19;19;19;19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000342066.7</td>\n",
       "      <td>19;19;19</td>\n",
       "      <td>19;19;19;19</td>\n",
       "      <td>19;19;19</td>\n",
       "      <td>19;19;19;19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id      ovlp        ovlp2 ovlp_strandless ovlp2_strandless\n",
       "0  ENST00000624431.1     0;0;0      0;0;0;0           0;0;0          0;0;0;0\n",
       "1  ENST00000624735.1     0;0;0      0;0;0;0           0;0;0          0;0;0;0\n",
       "2  ENST00000420190.6     0;0;0      0;0;0;0           0;0;0          0;0;0;0\n",
       "3  ENST00000437963.5  19;19;19  19;19;19;19        19;19;19      19;19;19;19\n",
       "4  ENST00000342066.7  19;19;19  19;19;19;19        19;19;19      19;19;19;19"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "local_and_global_df_all_ovlp['ovlp'] = local_and_global_df_all_ovlp['max_sum_overlap_g25_all_exons'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_g35_all_exons'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_refseq_all_exons'].astype(str)\n",
    "\n",
    "local_and_global_df_all_ovlp['ovlp_strandless'] = local_and_global_df_all_ovlp['max_overlap_g25_all_exons_strandless'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_g35_all_exons_strandless'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_refseq_all_exons_strandless'].astype(str)\n",
    "\n",
    "local_and_global_df_all_ovlp['ovlp2'] = local_and_global_df_all_ovlp['max_sum_overlap_g25_all_exons'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_g35_all_exons'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_g38_all_exons'].astype(str) +';'+local_and_global_df_all_ovlp['max_sum_overlap_refseq_all_exons'].astype(str)\n",
    "\n",
    "local_and_global_df_all_ovlp['ovlp2_strandless'] = local_and_global_df_all_ovlp['max_overlap_g25_all_exons_strandless'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_g35_all_exons_strandless'].astype(str) + ';' + local_and_global_df_all_ovlp['max_sum_overlap_g38_all_exons_strandless'].astype(str) +';'+local_and_global_df_all_ovlp['max_sum_overlap_refseq_all_exons_strandless'].astype(str)\n",
    "\n",
    "\n",
    "local_and_global_df_all_ovlp[['tr_id', 'ovlp', 'ovlp2', 'ovlp_strandless', 'ovlp2_strandless']][0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>...</th>\n",
       "      <th>max_sum_overlap_g35_all_exons</th>\n",
       "      <th>max_sum_overlap_g35_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_g38_all_exons</th>\n",
       "      <th>max_sum_overlap_g38_all_exons_strandless</th>\n",
       "      <th>max_sum_overlap_refseq_all_exons</th>\n",
       "      <th>max_sum_overlap_refseq_all_exons_strandless</th>\n",
       "      <th>ovlp</th>\n",
       "      <th>ovlp_strandless</th>\n",
       "      <th>ovlp2</th>\n",
       "      <th>ovlp2_strandless</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20599</th>\n",
       "      <td>ENST00000439211.6</td>\n",
       "      <td>0</td>\n",
       "      <td>494</td>\n",
       "      <td>494</td>\n",
       "      <td>2</td>\n",
       "      <td>494</td>\n",
       "      <td>164.0</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>344</td>\n",
       "      <td>chr5:80654490-80654639</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>237;237;237</td>\n",
       "      <td>0;0;0;0</td>\n",
       "      <td>237;237;237;237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "20599  ENST00000439211.6             0         494  494              2   \n",
       "\n",
       "       N_term_end1  len_codons  gene  50len_flag  global_coo_50_and_less  ...  \\\n",
       "20599          494       164.0  DHFR         344  chr5:80654490-80654639  ...   \n",
       "\n",
       "      max_sum_overlap_g35_all_exons max_sum_overlap_g35_all_exons_strandless  \\\n",
       "20599                             0                                      237   \n",
       "\n",
       "      max_sum_overlap_g38_all_exons max_sum_overlap_g38_all_exons_strandless  \\\n",
       "20599                             0                                      237   \n",
       "\n",
       "       max_sum_overlap_refseq_all_exons  \\\n",
       "20599                                 0   \n",
       "\n",
       "       max_sum_overlap_refseq_all_exons_strandless   ovlp  ovlp_strandless  \\\n",
       "20599                                          237  0;0;0      237;237;237   \n",
       "\n",
       "         ovlp2  ovlp2_strandless  \n",
       "20599  0;0;0;0   237;237;237;237  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # g25 - g35 - g38 - refseq\n",
    "# 5'UTR of DHFR gene has overlap with coding exons of MSH3 in a different strand\n",
    "\n",
    "local_and_global_df_all_ovlp[local_and_global_df_all_ovlp['tr_id'] == 'ENST00000439211.6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add PhyloCSF score results "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* note: if length of extension is <=1 codon - no PhyloCSF score was calculated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(77513, 28) (77820, 26)\n"
     ]
    }
   ],
   "source": [
    "li = []\n",
    "\n",
    "for path in glob.glob('data/PhyloCSF_50_NTE/*/*.txt'):\n",
    "    try: \n",
    "        tmp = pd.read_csv(path, sep='\\t', header=None)\n",
    "        li.append(tmp)\n",
    "        \n",
    "    except Exception as e:     \n",
    "        with open(path, 'r') as f:\n",
    "            li1 = []\n",
    "            for l in f.readlines():\n",
    "                if (l.find('Failure(\"invalid MFA alignment: Invalid_argument(\\\\\"Option.get\\\\\")\")') == -1) | (l.find('Failure(\"no regions successfully evaluated\")') == -1):\n",
    "                    li1.append(l.strip().split('\\t'))\n",
    "                else:\n",
    "                    print (l.split('\\t')[0])\n",
    "            tmp2 = pd.DataFrame(li1)\n",
    "            li.append(tmp2)\n",
    "\n",
    "PhyloCSF_50 = pd.concat(li)[[0, 1, 2]]\n",
    "\n",
    "PhyloCSF_50['tr_id'] = [x.split('/')[-1].split('.fasta')[0] for x in PhyloCSF_50[0].tolist()]\n",
    "\n",
    "PhyloCSF_50 = PhyloCSF_50[['tr_id', 2]]\n",
    "\n",
    "PhyloCSF_50 = PhyloCSF_50[(PhyloCSF_50[2] != 'Failure(\"no regions successfully evaluated\")') & (PhyloCSF_50[2] != 'Failure(\"invalid MFA alignment: Invalid_argument(\\\\\"Option.get\\\\\")\")')]\n",
    "\n",
    "PhyloCSF_50[2] = PhyloCSF_50[2].astype(float)\n",
    "\n",
    "PhyloCSF_50.columns = ['tr_id', 'PhyloCSF120score']\n",
    "\n",
    "PhyloCSF_50['tr_id1'] = [x.split('.')[0] for x in PhyloCSF_50['tr_id'].tolist()]\n",
    "\n",
    "local_and_global_df_all_ovlp_PhyloCSF = local_and_global_df_all_ovlp.merge(PhyloCSF_50, on='tr_id', how='inner')\n",
    "\n",
    "print (local_and_global_df_all_ovlp_PhyloCSF.shape, local_and_global_df_all_ovlp.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# local_and_global_df_all_ovlp_PhyloCSF.to_csv('tmp_res/local_and_global_df_g25ovlp_g35ovlp_refseqovlp_FRAMES_PhyloCSF.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add number of records for PhyloCSF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* file number_of_records_in_alignments_PCSF.txt was calculated on a server and transfered locally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(77513, 28) (77513, 30)\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "size_of_al = pd.read_csv('tmp_res/number_of_records_in_alignments_PCSF.txt', sep='\\t')\n",
    "\n",
    "size_of_al['tr_id'] = size_of_al['name'].str.split('.fasta').str[0]\n",
    "\n",
    "local_and_global_df_all_ovlp_PhyloCSF_rec = local_and_global_df_all_ovlp_PhyloCSF.merge(size_of_al[['tr_id', 'number_of_records', 'records_names']], how='left')\n",
    "\n",
    "print (local_and_global_df_all_ovlp_PhyloCSF.shape, local_and_global_df_all_ovlp_PhyloCSF_rec.shape)\n",
    "\n",
    "print (local_and_global_df_all_ovlp_PhyloCSF_rec[local_and_global_df_all_ovlp_PhyloCSF_rec['number_of_records'].isna()].tr_id.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "local_and_global_df_all_ovlp_PhyloCSF_rec.to_csv('tmp_res/local_and_global_df_g25ovlp_g35ovlp_g38ovlp_refseqovlp_strand_FRAMES_PhyloCSF.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
