{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alla/anaconda3/lib/python3.7/site-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import collections\n",
    "from collections import OrderedDict, Counter, defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from bokeh.plotting import figure, output_file, show \n",
    "import matplotlib.patches as patches    \n",
    "from matplotlib import colors\n",
    "\n",
    "\n",
    "import glob\n",
    "\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "\n",
    "import re\n",
    "\n",
    "import pickle\n",
    "\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "import Bio.Data.CodonTable\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "import Bio.Data.CodonTable\n",
    "\n",
    "import joypy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>PhyloCSF120score</th>\n",
       "      <th>number_of_records</th>\n",
       "      <th>ovlp</th>\n",
       "      <th>Riboseq_Summary</th>\n",
       "      <th>Coverage_value_ext</th>\n",
       "      <th>Proteomics_count_ext</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>CDS_ratio</th>\n",
       "      <th>pos_of_codon</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000379389.4</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>1</td>\n",
       "      <td>151</td>\n",
       "      <td>50.0</td>\n",
       "      <td>chr1:1013424-1013573</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1013424-1013573</td>\n",
       "      <td>-2443.9971</td>\n",
       "      <td>111</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>ATC; 98-152; Rank: 288; cov: 53.85</td>\n",
       "      <td>53.846154</td>\n",
       "      <td>0.0</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.327240</td>\n",
       "      <td>97</td>\n",
       "      <td>ATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000349431.10</td>\n",
       "      <td>UBE2J2</td>\n",
       "      <td>28</td>\n",
       "      <td>220</td>\n",
       "      <td>64.0</td>\n",
       "      <td>chr1:1273666-1273815</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:1273666-1273857</td>\n",
       "      <td>-1285.8186</td>\n",
       "      <td>78</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>ACG; 173-221; Rank: 328; cov: 90.91</td>\n",
       "      <td>90.909091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.762814</td>\n",
       "      <td>172</td>\n",
       "      <td>ACG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                tr_id    gene  N_term_start1  N_term_end1  len_codons  \\\n",
       "0   ENST00000379389.4   ISG15              1          151        50.0   \n",
       "1  ENST00000349431.10  UBE2J2             28          220        64.0   \n",
       "\n",
       "  global_coo_50_and_less strand    global_coo_primary  PhyloCSF120score  \\\n",
       "0   chr1:1013424-1013573      +  chr1:1013424-1013573        -2443.9971   \n",
       "1   chr1:1273666-1273815      -  chr1:1273666-1273857        -1285.8186   \n",
       "\n",
       "   number_of_records   ovlp                      Riboseq_Summary  \\\n",
       "0                111  0;0;0   ATC; 98-152; Rank: 288; cov: 53.85   \n",
       "1                 78  0;0;0  ACG; 173-221; Rank: 328; cov: 90.91   \n",
       "\n",
       "   Coverage_value_ext  Proteomics_count_ext                            tag  \\\n",
       "0           53.846154                   0.0  basic,appris_principal_1,CCDS   \n",
       "1           90.909091                   0.0  basic,appris_principal_1,CCDS   \n",
       "\n",
       "  transcript_type  CDS_ratio  pos_of_codon codon  \n",
       "0  protein_coding   0.327240            97   ATC  \n",
       "1  protein_coding   0.762814           172   ACG  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# open metadata with scores and metrics\n",
    "meta = pd.read_csv('tmp_res/METADATA_CURRENT.txt', sep='\\t')\n",
    "\n",
    "# gencode 25\n",
    "metadata_pc_g25 = pd.read_csv('tmp_res/metadata_pc_g25.txt', sep='\\t')\n",
    "\n",
    "\n",
    "P_Set = pd.read_csv('tmp_res/SET1.txt', sep='\\t')\n",
    "\n",
    "R_Set = pd.read_csv('tmp_res/SET2.txt', sep='\\t')\n",
    "\n",
    "R_Set['pos_of_codon'] = [int(x.split('; ')[1].split('-')[0])-1 for x in R_Set.Riboseq_Summary.tolist()]\n",
    "\n",
    "R_Set['codon'] = [x.split('; ')[0] for x in R_Set.Riboseq_Summary.tolist()]\n",
    "\n",
    "R_Set[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Some important functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_aln(fasta_input):\n",
    "    aln_col = []\n",
    "    ids_li = []\n",
    "    \n",
    "    with open(fasta_input, 'r') as input_handle:\n",
    "        for record in SeqIO.parse(input_handle, \"fasta\"):\n",
    "            seq = str(record.seq)\n",
    "            ids = str(record.id)\n",
    "            aln_col.append([seq[i:i+1] for i in range(0, len(seq), 1)])\n",
    "            ids_li.append(ids)\n",
    "    return aln_col, ids_li\n",
    "\n",
    "def re_index_with_gaps(aln_col):\n",
    "    new_col_ind = []\n",
    "\n",
    "    # take ref seq\n",
    "    counter_pos = 0 # to exclude '-'\n",
    "\n",
    "    for col in pd.DataFrame(aln_col).to_numpy()[0]:\n",
    "        if (col == '-') | (col == '.'): # UCC UGA C = TCC TGA C\n",
    "            new_col_ind.append(counter_pos)\n",
    "        else: \n",
    "            counter_pos += 1\n",
    "            new_col_ind.append(counter_pos)\n",
    "    \n",
    "    return new_col_ind\n",
    "\n",
    "\n",
    "def slice_aln(aln_col, new_col_ind, a, b):\n",
    "    fr_stop_motif_array = []\n",
    "    for row in pd.DataFrame(aln_col).to_numpy():\n",
    "        tmp = []\n",
    "        for ind, col in zip(new_col_ind, row):\n",
    "            if (ind >= a) & (ind <= b):\n",
    "                tmp.append(col)\n",
    "            \n",
    "        fr_stop_motif_array.append(tmp)\n",
    "    \n",
    "    return fr_stop_motif_array\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "10\n",
      "20\n",
      "30\n",
      "40\n",
      "50\n",
      "60\n",
      "70\n",
      "80\n",
      "90\n",
      "100\n",
      "110\n",
      "120\n",
      "130\n",
      "140\n",
      "150\n",
      "160\n",
      "170\n",
      "180\n",
      "190\n",
      "200\n",
      "210\n",
      "220\n",
      "230\n",
      "240\n",
      "250\n",
      "260\n",
      "270\n",
      "280\n",
      "290\n",
      "300\n",
      "310\n",
      "320\n",
      "330\n",
      "340\n",
      "350\n",
      "360\n",
      "370\n",
      "380\n",
      "390\n"
     ]
    }
   ],
   "source": [
    "li = []\n",
    "\n",
    "kek = {}\n",
    "\n",
    "for i, el in enumerate(R_Set[['tr_id', 'gene', 'pos_of_codon']].merge(meta[['tr_id', 'N_term_start1', 'global_coo_primary']], on='tr_id', how='inner').to_numpy()):\n",
    "    tr_id = el[0]\n",
    "    \n",
    "    #print (tr_id)\n",
    "    chrom = el[-1].split(':')[0]\n",
    "    nte_start = el[-2]\n",
    "    pos_of_codon = el[2]\n",
    "    \n",
    "    a = pos_of_codon - nte_start + 1\n",
    "    \n",
    "    b = a + 2\n",
    "    \n",
    "    fasta_input = 'data/nte_full_aln/out/%s/fasta3/%s.fasta' % (chrom, tr_id)\n",
    "\n",
    "    aln_col, ids_li = get_aln(fasta_input)\n",
    "    new_col_ind = re_index_with_gaps(aln_col)\n",
    "    sliced_array = slice_aln(aln_col, new_col_ind, a, b)\n",
    "    \n",
    "    a1 = pos_of_codon - nte_start + 1 - 4\n",
    "    b1 = pos_of_codon - nte_start + 1 + 4\n",
    "    sliced_array_TIS = slice_aln(aln_col, new_col_ind, a1, b1)\n",
    "    sliced_array_TIS_and_surroundings = slice_aln(aln_col, new_col_ind, a1-15, b1+15)\n",
    "    \n",
    "    # check conservation\n",
    "    # fraction of the exact reference codon \n",
    "    # Counter of near-cognate codons \n",
    "    ref_codon = ''.join(pd.DataFrame(sliced_array).to_numpy()[0]).replace('-', '')\n",
    "    \n",
    "    #if tr_id == 'ENST00000341772.4':\n",
    "        #print (pd.DataFrame(sliced_array))\n",
    "    \n",
    "    number_of_species  = len(aln_col)\n",
    "    \n",
    "    codons = [''.join(list(x)).replace('-', '') for x in pd.DataFrame(sliced_array).to_numpy()[0:]]\n",
    "    \n",
    "    ref_codon_count = codons.count(ref_codon)\n",
    "    \n",
    "    # count of all non-cognate + AUG\n",
    "    non_cognates = ['ATG', 'CTG', 'TTG', 'GTG', 'ATC', 'ATT', 'ATA', 'AGG', 'AAG', 'ACG']\n",
    "    \n",
    "    non_cognates_codon_count = 0\n",
    "    for codon in non_cognates:\n",
    "        non_cognates_codon_count += codons.count(codon)\n",
    "    \n",
    "    ref_codon_fraction = ref_codon_count / number_of_species\n",
    "    near_cognate_codong_fraction = non_cognates_codon_count / number_of_species\n",
    "    \n",
    "    counter_codon = Counter(codons)\n",
    "    \n",
    "    if i % 10 == 0:\n",
    "        print (i)\n",
    "        \n",
    "    kek[tr_id] = [sliced_array, aln_col,sliced_array_TIS,sliced_array_TIS_and_surroundings,ids_li]\n",
    "    \n",
    "    dict_animals = dict(zip(ids_li[1:], codons))\n",
    "    \n",
    "    li.append([tr_id, el[2], nte_start, ref_codon, number_of_species, ref_codon_count, ref_codon_fraction, counter_codon,\n",
    "              non_cognates_codon_count, near_cognate_codong_fraction, dict_animals, ids_li])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(395, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>nte_start</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>number_of_species</th>\n",
       "      <th>ref_codon_count</th>\n",
       "      <th>ref_codon_fraction</th>\n",
       "      <th>counter_codon</th>\n",
       "      <th>near_cognate_codon_count</th>\n",
       "      <th>near_cognate_codon_fraction</th>\n",
       "      <th>specie_codon_dict</th>\n",
       "      <th>animals_li</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000379389.4</td>\n",
       "      <td>97</td>\n",
       "      <td>1</td>\n",
       "      <td>ATC</td>\n",
       "      <td>111</td>\n",
       "      <td>33</td>\n",
       "      <td>0.297297</td>\n",
       "      <td>{'ATC': 33, 'TTC': 2, '': 7, 'GTCTTCGCCCGACCGG...</td>\n",
       "      <td>36</td>\n",
       "      <td>0.324324</td>\n",
       "      <td>{'Chimp': 'ATC', 'Bonobo': 'ATC', 'Gorilla': '...</td>\n",
       "      <td>[Human, Chimp, Bonobo, Gorilla, Orangutan, Gib...</td>\n",
       "      <td>ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000349431.10</td>\n",
       "      <td>172</td>\n",
       "      <td>28</td>\n",
       "      <td>ACG</td>\n",
       "      <td>82</td>\n",
       "      <td>58</td>\n",
       "      <td>0.707317</td>\n",
       "      <td>{'ACG': 58, 'ACA': 1, 'AGG': 2, 'GCA': 1, 'GCG...</td>\n",
       "      <td>60</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>{'Chimp': 'ACG', 'Bonobo': 'ACG', 'Gorilla': '...</td>\n",
       "      <td>[Human, Chimp, Bonobo, Gorilla, Orangutan, Gib...</td>\n",
       "      <td>UBE2J2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                tr_id  codon_start  nte_start ref_codon  number_of_species  \\\n",
       "0   ENST00000379389.4           97          1       ATC                111   \n",
       "1  ENST00000349431.10          172         28       ACG                 82   \n",
       "\n",
       "   ref_codon_count  ref_codon_fraction  \\\n",
       "0               33            0.297297   \n",
       "1               58            0.707317   \n",
       "\n",
       "                                       counter_codon  \\\n",
       "0  {'ATC': 33, 'TTC': 2, '': 7, 'GTCTTCGCCCGACCGG...   \n",
       "1  {'ACG': 58, 'ACA': 1, 'AGG': 2, 'GCA': 1, 'GCG...   \n",
       "\n",
       "   near_cognate_codon_count  near_cognate_codon_fraction  \\\n",
       "0                        36                     0.324324   \n",
       "1                        60                     0.731707   \n",
       "\n",
       "                                   specie_codon_dict  \\\n",
       "0  {'Chimp': 'ATC', 'Bonobo': 'ATC', 'Gorilla': '...   \n",
       "1  {'Chimp': 'ACG', 'Bonobo': 'ACG', 'Gorilla': '...   \n",
       "\n",
       "                                          animals_li    gene  \n",
       "0  [Human, Chimp, Bonobo, Gorilla, Orangutan, Gib...   ISG15  \n",
       "1  [Human, Chimp, Bonobo, Gorilla, Orangutan, Gib...  UBE2J2  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conservation_df = pd.DataFrame(li, columns = ['tr_id', 'codon_start', 'nte_start', 'ref_codon', \n",
    "                                  'number_of_species', 'ref_codon_count', \n",
    "                                  'ref_codon_fraction', 'counter_codon', \n",
    "                                 'near_cognate_codon_count', 'near_cognate_codon_fraction',\n",
    "                                             'specie_codon_dict', 'animals_li'])\n",
    "\n",
    "conservation_df = conservation_df.merge(metadata_pc_g25[['tr_id', 'gene']], on='tr_id', how='inner')\n",
    "\n",
    "# exclude ATG extensions\n",
    "conservation_df = conservation_df[conservation_df['codon_start'] != 'ATG']\n",
    "\n",
    "print (conservation_df.shape)\n",
    "conservation_df[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>specie_codon_dict</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000379389.4</td>\n",
       "      <td>ATC</td>\n",
       "      <td>{'Chimp': 'ATC', 'Bonobo': 'ATC', 'Gorilla': '...</td>\n",
       "      <td>ISG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000349431.10</td>\n",
       "      <td>ACG</td>\n",
       "      <td>{'Chimp': 'ACG', 'Bonobo': 'ACG', 'Gorilla': '...</td>\n",
       "      <td>UBE2J2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000338370.7</td>\n",
       "      <td>ATT</td>\n",
       "      <td>{'Chimp': 'ATT', 'Bonobo': 'ATT', 'Gorilla': '...</td>\n",
       "      <td>AURKAIP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000235310.7</td>\n",
       "      <td>GTG</td>\n",
       "      <td>{'Chimp': 'GTG', 'Bonobo': 'GTG', 'Gorilla': '...</td>\n",
       "      <td>MAD2L2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000356634.7</td>\n",
       "      <td>GTG</td>\n",
       "      <td>{'Chimp': 'GTG', 'Bonobo': 'GTG', 'Gorilla': '...</td>\n",
       "      <td>KDM1A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                tr_id ref_codon  \\\n",
       "0   ENST00000379389.4       ATC   \n",
       "1  ENST00000349431.10       ACG   \n",
       "2   ENST00000338370.7       ATT   \n",
       "3   ENST00000235310.7       GTG   \n",
       "4   ENST00000356634.7       GTG   \n",
       "\n",
       "                                   specie_codon_dict      gene  \n",
       "0  {'Chimp': 'ATC', 'Bonobo': 'ATC', 'Gorilla': '...     ISG15  \n",
       "1  {'Chimp': 'ACG', 'Bonobo': 'ACG', 'Gorilla': '...    UBE2J2  \n",
       "2  {'Chimp': 'ATT', 'Bonobo': 'ATT', 'Gorilla': '...  AURKAIP1  \n",
       "3  {'Chimp': 'GTG', 'Bonobo': 'GTG', 'Gorilla': '...    MAD2L2  \n",
       "4  {'Chimp': 'GTG', 'Bonobo': 'GTG', 'Gorilla': '...     KDM1A  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conservation_df[['tr_id', 'ref_codon', 'specie_codon_dict', 'gene']][0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# multiple sequence alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo_primary</th>\n",
       "      <th>PhyloCSF120score</th>\n",
       "      <th>number_of_records</th>\n",
       "      <th>ovlp</th>\n",
       "      <th>Riboseq_Summary</th>\n",
       "      <th>Coverage_value_ext</th>\n",
       "      <th>Proteomics_count_ext</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>ENST00000610821.4</td>\n",
       "      <td>LRP5L</td>\n",
       "      <td>2366</td>\n",
       "      <td>2465</td>\n",
       "      <td>33.0</td>\n",
       "      <td>chr22:25360093-25360191</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:25360093-25360191</td>\n",
       "      <td>1669.2062</td>\n",
       "      <td>110</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>ACG; 2367-2466; Rank: 3720; cov: 14.29</td>\n",
       "      <td>14.285714</td>\n",
       "      <td>0.0</td>\n",
       "      <td>basic,appris_principal_2,CCDS</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>ENST00000402859.6</td>\n",
       "      <td>LRP5L</td>\n",
       "      <td>425</td>\n",
       "      <td>524</td>\n",
       "      <td>33.0</td>\n",
       "      <td>chr22:25360093-25360191</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:25360093-25360191</td>\n",
       "      <td>1669.2062</td>\n",
       "      <td>110</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic,appris_principal_2,CCDS</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>ENST00000444995.7</td>\n",
       "      <td>LRP5L</td>\n",
       "      <td>602</td>\n",
       "      <td>701</td>\n",
       "      <td>33.0</td>\n",
       "      <td>chr22:25360093-25360191</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:25360093-25360191</td>\n",
       "      <td>1669.2062</td>\n",
       "      <td>110</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic,appris_alternative_2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>ENST00000402785.2</td>\n",
       "      <td>LRP5L</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "      <td>32.0</td>\n",
       "      <td>chr22:25360093-25360188</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:25360093-25360188</td>\n",
       "      <td>1554.5767</td>\n",
       "      <td>110</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>basic,appris_principal_2,CCDS</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                tr_id   gene  N_term_start1  N_term_end1  len_codons  \\\n",
       "68  ENST00000610821.4  LRP5L           2366         2465        33.0   \n",
       "69  ENST00000402859.6  LRP5L            425          524        33.0   \n",
       "70  ENST00000444995.7  LRP5L            602          701        33.0   \n",
       "71  ENST00000402785.2  LRP5L              1           97        32.0   \n",
       "\n",
       "     global_coo_50_and_less strand       global_coo_primary  PhyloCSF120score  \\\n",
       "68  chr22:25360093-25360191      -  chr22:25360093-25360191         1669.2062   \n",
       "69  chr22:25360093-25360191      -  chr22:25360093-25360191         1669.2062   \n",
       "70  chr22:25360093-25360191      -  chr22:25360093-25360191         1669.2062   \n",
       "71  chr22:25360093-25360188      -  chr22:25360093-25360188         1554.5767   \n",
       "\n",
       "    number_of_records   ovlp                         Riboseq_Summary  \\\n",
       "68                110  0;0;0  ACG; 2367-2466; Rank: 3720; cov: 14.29   \n",
       "69                110  0;0;0                                     NaN   \n",
       "70                110  0;0;0                                     NaN   \n",
       "71                110  0;0;0                                     NaN   \n",
       "\n",
       "    Coverage_value_ext  Proteomics_count_ext                            tag  \\\n",
       "68           14.285714                   0.0  basic,appris_principal_2,CCDS   \n",
       "69                 NaN                   NaN  basic,appris_principal_2,CCDS   \n",
       "70                 NaN                   NaN     basic,appris_alternative_2   \n",
       "71                 NaN                   NaN  basic,appris_principal_2,CCDS   \n",
       "\n",
       "   transcript_type  \n",
       "68  protein_coding  \n",
       "69  protein_coding  \n",
       "70  protein_coding  \n",
       "71  protein_coding  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P_Set[P_Set['gene'] == 'LRP5L']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "So for LRP5L theoretical extension (I included the last 3' in-frame stop and annotated ATG): \n",
    "https://data.broadinstitute.org/compbio1/cav.php?controlsState=show&intervals=chr22%3A25360090-25360194&strand=-&alnset=hg38_241mammals&spliceSites=human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_aln, full_aln, tis_aln, sliced_array_TIS_and_surroundings, ids_li = kek['ENST00000357214.5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>142</th>\n",
       "      <th>143</th>\n",
       "      <th>144</th>\n",
       "      <th>145</th>\n",
       "      <th>146</th>\n",
       "      <th>147</th>\n",
       "      <th>148</th>\n",
       "      <th>149</th>\n",
       "      <th>150</th>\n",
       "      <th>151</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115 rows × 152 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    0   1   2   3   4   5   6   7   8   9    ... 142 143 144 145 146 147 148  \\\n",
       "0     G   C   C   G   C   C   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "1     G   C   C   G   C   C   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "2     G   C   C   G   C   C   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "3     G   C   C   G   C   C   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "4     G   C   C   G   C   C   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..  ..   \n",
       "110   G   C   C   G   C   C   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "111   G   C   C   G   C   C   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "112   G   G   G   G   C   T   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "113   G   G   T   G   C   T   T   G   T   G  ...   -   -   -   -   -   -   -   \n",
       "114   G   G   C   G   C   T   T   G   T   A  ...   -   -   -   -   -   -   -   \n",
       "\n",
       "    149 150 151  \n",
       "0     -   -   -  \n",
       "1     -   -   -  \n",
       "2     -   -   -  \n",
       "3     -   -   -  \n",
       "4     -   -   -  \n",
       "..   ..  ..  ..  \n",
       "110   -   -   -  \n",
       "111   -   -   -  \n",
       "112   -   -   -  \n",
       "113   -   -   -  \n",
       "114   -   -   -  \n",
       "\n",
       "[115 rows x 152 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(full_aln)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
