{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import collections\n",
    "from collections import OrderedDict, Counter, defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import glob\n",
    "\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "\n",
    "import re\n",
    "\n",
    "import pickle\n",
    "\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 24/59 genes in [Ivanov et al, 2011] have been annotated in gencode35. \n",
    "\n",
    "# What if we remove their annotation and try to predict them using PhyloCSF score? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes from 2011 paper\n",
    "genes_42 = 'RASL10B,FGFR1,YPEL2,ENOX2,UFSP1,WDR26,TIAL1,YPEL4,NGF,TRPV6,KCTD11,STARD10,YPEL1,R3HCC1,ZFP62,NHLRC4,EPHX3,C1QL4,TLE3,HDGF,ANKRD42,FAM217B,NFKBID,PTEN,RNF187,METTL23,TMEM8B,VANGL2,FNDC5,C1QL1,CITED1,EIF4G3,HELZ2,CYTH2,NTF3,MFSD4B,C1QL3,KCNN4,IFT46,RASD2,CITED2,C1QL2'.split(',')\n",
    "genes_17 = 'GTF3A, EIF4G2, SP3, PRPS1L1, TEAD4, TEAD3, CACNG8, OAZ3, TEAD1, DDX17, VEGFA, NR1I2, HCK, WT1, BAG1, NPW, MYC'.split(', ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(94359, 14)\n",
      "(101486, 14)\n"
     ]
    }
   ],
   "source": [
    "# open metadata with scores and metrics\n",
    "meta = pd.read_csv('tmp_res/local_and_global_df_g25ovlp_g35ovlp_g38ovlp_refseqovlp_strand_FRAMES_PhyloCSF.txt', sep='\\t')\n",
    "\n",
    "metadata_pc_g25 = pd.read_csv('tmp_res/metadata_pc_g25.txt', sep='\\t')\n",
    "print (metadata_pc_g25.shape)\n",
    "\n",
    "metadata_pc_g35 = pd.read_csv('tmp_res/metadata_pc_g35.txt', sep='\\t')\n",
    "print (metadata_pc_g35.shape)\n",
    "\n",
    "\n",
    "# open SET1 (PhyloSET) and SET2 (RiboSET)\n",
    "SET1 = pd.read_csv('tmp_res/SET1.txt', sep='\\t')\n",
    "SET2 = pd.read_csv('tmp_res/SET2.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# open file with extension sequences from 2011 paper \n",
    "# there is .pdf file in supplementary data -> I took sequences manually from them \n",
    "EXT_SEQ_DF = pd.read_csv('data/extensions_from_2011.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes24 = ['FNDC5','OAZ3','RNF187', 'SP3', 'NR1I2', 'TEAD3',\n",
    " 'VEGFA', 'PRPS1L1', 'TRPV6', 'R3HCC1', 'MYC', 'BAG1', 'EIF4G2', \n",
    "           'WT1', 'TEAD4', 'GTF3A', 'NPW', 'KCTD11', 'YPEL2', 'NFKBID', 'CACNG8', 'HCK', 'YPEL1', 'DDX17']\n",
    "\n",
    "len(genes24)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>refseq_id</th>\n",
       "      <th>cds_find</th>\n",
       "      <th>nte_find</th>\n",
       "      <th>biotype</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373471.8</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>NM_153756.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>non_ATG_start,basic,CCDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000479764.6</td>\n",
       "      <td>OAZ3</td>\n",
       "      <td>NM_016178.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>non_canonical_other,non_ATG_start,basic,exp_conf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id   gene    refseq_id  cds_find  nte_find         biotype  \\\n",
       "0  ENST00000373471.8  FNDC5  NM_153756.3         0        -1  protein_coding   \n",
       "1  ENST00000479764.6   OAZ3  NM_016178.2         0        -1  protein_coding   \n",
       "\n",
       "                                                tag  \n",
       "0                          non_ATG_start,basic,CCDS  \n",
       "1  non_canonical_other,non_ATG_start,basic,exp_conf  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann_ext_g35 = pd.read_csv('tmp_res/ANNOTATED_EXT_FOUND_df_g35_pc_complete_CDS.txt', sep='\\t')\n",
    "ann_ext_g35[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXT_SEQ_DF = pd.read_csv('data/extensions_from_2011.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We are going to extract their nonAUG proteoforms from gencode35 and 'remove' their nonAUG annotation back to prevAUG one (downstream)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "interesting that for 2 genes, nonAUG proteoforms are annotated without 5'UTRs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'YPEL1', 'YPEL2'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(genes24) - set(metadata_pc_g35[(metadata_pc_g35['gene'].isin(genes24)) & \n",
    "                (metadata_pc_g35['cds_start_codon'] != 'ATG') & \n",
    "                 metadata_pc_g35['5UTR_start_seq'].notna()].gene.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plan\n",
    "\n",
    "* get a position of 1st downstream in-frame AUG (in CDS)\n",
    "* check whether it's the 1st downstream AUG was annotated before as the start (maybe the next one for some reason?)\n",
    "* 're-annotate' transcripts -> local coordinates of potential NTE -> global -> PCSF score calculation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_tr</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>cds_seq</th>\n",
       "      <th>cds_start_codon</th>\n",
       "      <th>cds_stop_codon</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>cds_stop_pos</th>\n",
       "      <th>utr5_start</th>\n",
       "      <th>utr5_end</th>\n",
       "      <th>record_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1997</th>\n",
       "      <td>ENST00000373471.8</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>FNDC5-201</td>\n",
       "      <td>AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...</td>\n",
       "      <td>AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...</td>\n",
       "      <td>ACC</td>\n",
       "      <td>ATACACCCCGGGTCGCCGAGCGCCTGGCCGCCCCGCGCCCGCGCCG...</td>\n",
       "      <td>ATA</td>\n",
       "      <td>TGA</td>\n",
       "      <td>166</td>\n",
       "      <td>805</td>\n",
       "      <td>0</td>\n",
       "      <td>166</td>\n",
       "      <td>ENST00000373471.8|ENSG00000160097.19|OTTHUMG00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>ENST00000649537.1</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>FNDC5-207</td>\n",
       "      <td>CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...</td>\n",
       "      <td>CCCCGGGCCTGCCGGCCGGAGGAGCCACC</td>\n",
       "      <td>ACC</td>\n",
       "      <td>ATACACCCCGGGTCGCCGAGCGCCTGGCCGCCCCGCGCCCGCGCCG...</td>\n",
       "      <td>ATA</td>\n",
       "      <td>TGA</td>\n",
       "      <td>29</td>\n",
       "      <td>635</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>ENST00000649537.1|ENSG00000160097.19|OTTHUMG00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tr_id   gene    gene_tr  \\\n",
       "1997  ENST00000373471.8  FNDC5  FNDC5-201   \n",
       "1998  ENST00000649537.1  FNDC5  FNDC5-207   \n",
       "\n",
       "                                         transcript_seq  \\\n",
       "1997  AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...   \n",
       "1998  CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...   \n",
       "\n",
       "                                         5UTR_start_seq CDS_start  \\\n",
       "1997  AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...       ACC   \n",
       "1998                      CCCCGGGCCTGCCGGCCGGAGGAGCCACC       ACC   \n",
       "\n",
       "                                                cds_seq cds_start_codon  \\\n",
       "1997  ATACACCCCGGGTCGCCGAGCGCCTGGCCGCCCCGCGCCCGCGCCG...             ATA   \n",
       "1998  ATACACCCCGGGTCGCCGAGCGCCTGGCCGCCCCGCGCCCGCGCCG...             ATA   \n",
       "\n",
       "     cds_stop_codon  cds_start_pos  cds_stop_pos  utr5_start  utr5_end  \\\n",
       "1997            TGA            166           805           0       166   \n",
       "1998            TGA             29           635           0        29   \n",
       "\n",
       "                                              record_id  \n",
       "1997  ENST00000373471.8|ENSG00000160097.19|OTTHUMG00...  \n",
       "1998  ENST00000649537.1|ENSG00000160097.19|OTTHUMG00...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([metadata_pc_g35[(metadata_pc_g35['gene'].isin(genes24)) & \n",
    "                (metadata_pc_g35['cds_start_codon'] != 'ATG') & \n",
    "                 metadata_pc_g35['5UTR_start_seq'].notna()],\n",
    "            metadata_pc_g35[metadata_pc_g35['gene'].isin(['YPEL1', 'YPEL2']) &\n",
    "                           (metadata_pc_g35['cds_start_codon'] != 'ATG')]])\n",
    "\n",
    "print (df.gene.nunique())\n",
    "\n",
    "df[0:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### iterate over 'CDS': find the 1st ATG "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "li = []\n",
    "\n",
    "for el in df[['tr_id', 'gene', 'cds_start_pos', 'transcript_seq']].to_numpy():\n",
    "    tr_id = el[0]\n",
    "    gene = el[1]\n",
    "    cds_start_pos = el[2]\n",
    "    tr_seq = el[3]\n",
    "    \n",
    "    CDS_seq = tr_seq[cds_start_pos:]\n",
    "    \n",
    "    #iterate over CDS seq by 3\n",
    "    for pos, codon in enumerate([CDS_seq[i:i+3] for i in range(0, len(CDS_seq), 3)]):\n",
    "        if codon == 'ATG':\n",
    "            first_dATG_pos = cds_start_pos + pos*3\n",
    "            \n",
    "            len_of_real_ext = (first_dATG_pos - cds_start_pos)/3\n",
    "            \n",
    "            li.append([tr_id, gene, tr_seq, cds_start_pos, first_dATG_pos, len_of_real_ext])\n",
    "            \n",
    "            #print (tr_id, tr_seq[first_dATG_pos:first_dATG_pos+3])\n",
    "            break \n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>tr_seq</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>first_dATG_pos</th>\n",
       "      <th>len_of_real_ext_in_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373471.8</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...</td>\n",
       "      <td>166</td>\n",
       "      <td>391</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000649537.1</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...</td>\n",
       "      <td>29</td>\n",
       "      <td>386</td>\n",
       "      <td>119.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000479764.6</td>\n",
       "      <td>OAZ3</td>\n",
       "      <td>GAAAGTATGGGGAGCGGGCCGGGGAACAGATGGCACCGCGGCTTGC...</td>\n",
       "      <td>3144</td>\n",
       "      <td>3288</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000400999.6</td>\n",
       "      <td>OAZ3</td>\n",
       "      <td>GTTGCCTAAACCTCTGCCACCCACCTGTGAACTTCACTTTGCCACA...</td>\n",
       "      <td>67</td>\n",
       "      <td>211</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id   gene  \\\n",
       "0  ENST00000373471.8  FNDC5   \n",
       "1  ENST00000649537.1  FNDC5   \n",
       "2  ENST00000479764.6   OAZ3   \n",
       "3  ENST00000400999.6   OAZ3   \n",
       "\n",
       "                                              tr_seq  cds_start_pos  \\\n",
       "0  AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...            166   \n",
       "1  CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...             29   \n",
       "2  GAAAGTATGGGGAGCGGGCCGGGGAACAGATGGCACCGCGGCTTGC...           3144   \n",
       "3  GTTGCCTAAACCTCTGCCACCCACCTGTGAACTTCACTTTGCCACA...             67   \n",
       "\n",
       "   first_dATG_pos  len_of_real_ext_in_codons  \n",
       "0             391                       75.0  \n",
       "1             386                      119.0  \n",
       "2            3288                       48.0  \n",
       "3             211                       48.0  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dATG_df = pd.DataFrame(li, columns = ['tr_id', 'gene', 'tr_seq', \n",
    "                                      'cds_start_pos', 'first_dATG_pos', 'len_of_real_ext_in_codons'])\n",
    "\n",
    "print (dATG_df.gene.nunique())\n",
    "\n",
    "# theoretical extension would be [STOP, first_dATG_pos]\n",
    "dATG_df[0:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pos_of_1st_stop_inframe(seq):\n",
    "    '''\n",
    "    seq = seq of 5'leader\n",
    "    '''\n",
    "    \n",
    "    # reverse seq\n",
    "    seq_r = seq[::-1]\n",
    "    \n",
    "    i = 0  \n",
    "    li = []\n",
    "    while i <= len(seq_r):    \n",
    "    \n",
    "        codon = seq_r[i:i+3][::-1] # read by codon and reverse it like it's read on forward strand\n",
    "        if (codon == 'TAG') | (codon == 'TGA') | (codon == 'TAA'):\n",
    "            li.append(i) \n",
    "        i = i + 3\n",
    "        \n",
    "    if li:\n",
    "        pos = (np.min(li))\n",
    "        pos_f = len(seq) - pos\n",
    "    else:\n",
    "        # check triplets \n",
    "        if len(seq) % 3 == 0:\n",
    "            pos_f = 0\n",
    "        elif len(seq) % 3 == 1:\n",
    "            pos_f = 1\n",
    "        else:\n",
    "            pos_f = 2\n",
    "    \n",
    "    return (pos_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['tr_id', 'gene', 'tr_seq', 'cds_start_pos', 'first_dATG_pos',\n",
       "       'len_of_real_ext_in_codons'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dATG_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>tr_seq</th>\n",
       "      <th>new_N_term_end</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>new_utr5_seq</th>\n",
       "      <th>new_N_term_start</th>\n",
       "      <th>new_NTE_seq</th>\n",
       "      <th>new_NTE_len_codons</th>\n",
       "      <th>len_of_real_ext_in_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373471.8</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...</td>\n",
       "      <td>391</td>\n",
       "      <td>166</td>\n",
       "      <td>AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...</td>\n",
       "      <td>1</td>\n",
       "      <td>GTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCAG...</td>\n",
       "      <td>130.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000649537.1</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...</td>\n",
       "      <td>386</td>\n",
       "      <td>29</td>\n",
       "      <td>CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...</td>\n",
       "      <td>2</td>\n",
       "      <td>CCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCCGA...</td>\n",
       "      <td>128.0</td>\n",
       "      <td>119.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id   gene  \\\n",
       "0  ENST00000373471.8  FNDC5   \n",
       "1  ENST00000649537.1  FNDC5   \n",
       "\n",
       "                                              tr_seq  new_N_term_end  \\\n",
       "0  AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...             391   \n",
       "1  CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...             386   \n",
       "\n",
       "   cds_start_pos                                       new_utr5_seq  \\\n",
       "0            166  AGTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCA...   \n",
       "1             29  CCCCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCC...   \n",
       "\n",
       "   new_N_term_start                                        new_NTE_seq  \\\n",
       "0                 1  GTCGCTGTCCGCGGAGCCGACATGCAGGCGGCTCGGGGCGGCGCAG...   \n",
       "1                 2  CCGGGCCTGCCGGCCGGAGGAGCCACCATACACCCCGGGTCGCCGA...   \n",
       "\n",
       "   new_NTE_len_codons  len_of_real_ext_in_codons  \n",
       "0               130.0                       75.0  \n",
       "1               128.0                      119.0  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the last 3'in-frame STOP\n",
    "li = []\n",
    "\n",
    "for el in dATG_df.to_numpy():\n",
    "    tr_id = el[0]\n",
    "    gene = el[1]\n",
    "    cds_start_pos = el[3]\n",
    "    tr_seq = el[2]\n",
    "    first_dATG_pos = el[4]\n",
    "    len_of_real_ext_in_codons = el[-1]\n",
    "    \n",
    "    utr5_seq = tr_seq[0:first_dATG_pos]\n",
    "    \n",
    "    pos_stop = get_pos_of_1st_stop_inframe(utr5_seq)\n",
    "    \n",
    "    nte_seq = tr_seq[pos_stop:first_dATG_pos]\n",
    "    \n",
    "    nte_ext_len = len(nte_seq) / 3\n",
    "    \n",
    "    li.append([tr_id, gene, tr_seq, first_dATG_pos, cds_start_pos, utr5_seq, \n",
    "               pos_stop, nte_seq, nte_ext_len, len_of_real_ext_in_codons])\n",
    "    \n",
    "nte_local_df = pd.DataFrame(li, columns = ['tr_id', 'gene', 'tr_seq', 'new_N_term_end',\n",
    "                                          'cds_start_pos', 'new_utr5_seq', 'new_N_term_start',\n",
    "                                          'new_NTE_seq', 'new_NTE_len_codons', 'len_of_real_ext_in_codons'])\n",
    "\n",
    "print (nte_local_df.gene.nunique())\n",
    "\n",
    "nte_local_df[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## add 50-codons long or less local coordinate "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_new_NTE_start_50_codons(x):\n",
    "    '''\n",
    "    return N_term_start_50\n",
    "    if there are >50 codons = all 50\n",
    "    if less = all we have \n",
    "    '''\n",
    "    if x['new_NTE_len_codons'] >= 50:\n",
    "        New_nte_start = x['new_N_term_end'] - 150  \n",
    "    else:\n",
    "        New_nte_start = x['new_N_term_start']\n",
    "        \n",
    "    return New_nte_start\n",
    "        \n",
    "\n",
    "nte_local_df['50len_flag'] = nte_local_df.apply(add_new_NTE_start_50_codons, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>new_N_term_start</th>\n",
       "      <th>new_N_term_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>1</td>\n",
       "      <td>391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3138</td>\n",
       "      <td>3135</td>\n",
       "      <td>3288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>61</td>\n",
       "      <td>58</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   50len_flag  new_N_term_start  new_N_term_end\n",
       "0         241                 1             391\n",
       "1         236                 2             386\n",
       "2        3138              3135            3288\n",
       "3          61                58             211"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nte_local_df[['50len_flag', 'new_N_term_start', 'new_N_term_end']][0:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "nte_local_df.to_csv('tmp_res/reprod_24genes_2011_local_coo_df.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Open global coo file and parse "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_global_coo(path_to_file_with_global_coo, colname):\n",
    "    global_coo_g25 = pd.read_csv(path_to_file_with_global_coo, sep='\\t')\n",
    "    global_coo_g25_f = global_coo_g25[global_coo_g25['hit'] == True]\n",
    "    global_coo_g25_f = global_coo_g25_f.sort_values(by=['seqnames', 'start'])\n",
    "    global_coo_g25_f['global_coo'] = global_coo_g25_f['seqnames']+':'+global_coo_g25_f['start'].astype(str) +'-'+global_coo_g25_f['end'].astype(str)\n",
    "    global_coo_g25_f = global_coo_g25_f[['group_name', 'strand', 'global_coo']].groupby(['group_name', 'strand']).agg('+'.join).reset_index()\n",
    "    global_coo_g25_f.columns = ['tr_id', 'strand', colname]\n",
    "    return global_coo_g25_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    }
   ],
   "source": [
    "codons50_global = prepare_global_coo(path_to_file_with_global_coo='tmp_res/reprod_24genes_2011_global_coo_df.txt',\n",
    "                                    colname = 'global_coo_50_and_less')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 13) 24\n"
     ]
    }
   ],
   "source": [
    "# merge it to local df\n",
    "\n",
    "nte_local_global_df = nte_local_df.merge(codons50_global, on='tr_id', how='inner')\n",
    "\n",
    "print (nte_local_global_df.shape, nte_local_global_df.gene.nunique())\n",
    "\n",
    "nte_local_global_df.to_csv('tmp_res/reprod_24genes_2011_local_global_coo_df.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# send it to server to calculated 50-codons PCSF score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse PhyloCSF results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "li = []\n",
    "\n",
    "for path in glob.glob('data/PCSF_24genes/*/*.txt'):\n",
    "    try: \n",
    "        tmp = pd.read_csv(path, sep='\\t', header=None)\n",
    "        li.append(tmp)\n",
    "        \n",
    "    except Exception as e:     \n",
    "        with open(path, 'r') as f:\n",
    "            li1 = []\n",
    "            for l in f.readlines():\n",
    "                if (l.find('Failure(\"invalid MFA alignment: Invalid_argument(\\\\\"Option.get\\\\\")\")') == -1) | (l.find('Failure(\"no regions successfully evaluated\")') == -1):\n",
    "                    li1.append(l.strip().split('\\t'))\n",
    "                else:\n",
    "                    print (l.split('\\t')[0])\n",
    "            tmp2 = pd.DataFrame(li1)\n",
    "            li.append(tmp2)\n",
    "\n",
    "PhyloCSF_50 = pd.concat(li)[[0, 1, 2]]\n",
    "\n",
    "PhyloCSF_50['tr_id'] = [x.split('/')[-1].split('.fasta')[0] for x in PhyloCSF_50[0].tolist()]\n",
    "\n",
    "PhyloCSF_50 = PhyloCSF_50[(PhyloCSF_50[2] != 'Failure(\"no regions successfully evaluated\")') & \n",
    "                          (PhyloCSF_50[2] != 'Failure(\"invalid MFA alignment: Invalid_argument(\\\\\"Option.get\\\\\")\")') & \n",
    "                          (PhyloCSF_50[1] != 'abort')]\n",
    "\n",
    "\n",
    "PhyloCSF_50 = PhyloCSF_50[['tr_id', 2]] \n",
    "\n",
    "\n",
    "PhyloCSF_50[2] = PhyloCSF_50[2].astype(float)\n",
    "\n",
    "PhyloCSF_50.columns = ['tr_id', 'PhyloCSF120score']\n",
    "\n",
    "PhyloCSF_50['tr_id1'] = [x.split('.')[0] for x in PhyloCSF_50['tr_id'].tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>PhyloCSF120score</th>\n",
       "      <th>tr_id1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000634734.2</td>\n",
       "      <td>-1026.5729</td>\n",
       "      <td>ENST00000634734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000396821.7</td>\n",
       "      <td>421.9530</td>\n",
       "      <td>ENST00000396821</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  PhyloCSF120score           tr_id1\n",
       "0  ENST00000634734.2        -1026.5729  ENST00000634734\n",
       "0  ENST00000396821.7          421.9530  ENST00000396821"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PhyloCSF_50[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nte_local_global_df[nte_local_global_df['tr_id'].isin(ann_ext_g35.tr_id.tolist())].gene.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['tr_id', 'gene', 'tr_seq', 'new_N_term_end', 'cds_start_pos',\n",
       "       'new_utr5_seq', 'new_N_term_start', 'new_NTE_seq', 'new_NTE_len_codons',\n",
       "       'len_of_real_ext_in_codons', '50len_flag', 'strand',\n",
       "       'global_coo_50_and_less', 'PhyloCSF120score', 'tr_id1'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nte_local_global_PCSF_df = nte_local_global_df.merge(PhyloCSF_50,  on=['tr_id'], how='left')\n",
    "nte_local_global_PCSF_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nte_local_global_PCSF_df[nte_local_global_PCSF_df['PhyloCSF120score'] > 0].gene.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "nte_local_global_PCSF_df.to_csv('tmp_res/PCSF_for_24_ann.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### these transcripts contain NTE-sequence predicted in [2011]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nte_local_global_df[(nte_local_global_df['tr_id'].isin(ann_ext_g35.tr_id.tolist())) \n",
    "                   & (nte_local_global_PCSF_df['PhyloCSF120score'] > 0)].gene.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>PhyloCSF120score</th>\n",
       "      <th>len_of_real_ext_in_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373471.8</td>\n",
       "      <td>FNDC5</td>\n",
       "      <td>chr1:32868374-32868388+chr1:32868882-32868997+...</td>\n",
       "      <td>-</td>\n",
       "      <td>2051.0659</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000479764.6</td>\n",
       "      <td>OAZ3</td>\n",
       "      <td>chr1:151766716-151766865</td>\n",
       "      <td>+</td>\n",
       "      <td>-139.7141</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000400999.6</td>\n",
       "      <td>OAZ3</td>\n",
       "      <td>chr1:151766716-151766865</td>\n",
       "      <td>+</td>\n",
       "      <td>-139.7141</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000305943.8</td>\n",
       "      <td>RNF187</td>\n",
       "      <td>chr1:228487666-228487815</td>\n",
       "      <td>+</td>\n",
       "      <td>-742.6165</td>\n",
       "      <td>109.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000418194.6</td>\n",
       "      <td>SP3</td>\n",
       "      <td>chr2:173955657-173955806</td>\n",
       "      <td>-</td>\n",
       "      <td>350.1512</td>\n",
       "      <td>217.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000652005.1</td>\n",
       "      <td>SP3</td>\n",
       "      <td>chr2:173955657-173955806</td>\n",
       "      <td>-</td>\n",
       "      <td>350.1512</td>\n",
       "      <td>217.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000393716.6</td>\n",
       "      <td>NR1I2</td>\n",
       "      <td>chr3:119807266-119807415</td>\n",
       "      <td>+</td>\n",
       "      <td>-625.4103</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000466380.5</td>\n",
       "      <td>NR1I2</td>\n",
       "      <td>chr3:119807266-119807415</td>\n",
       "      <td>+</td>\n",
       "      <td>-625.4103</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000338863.12</td>\n",
       "      <td>TEAD3</td>\n",
       "      <td>chr6:35486468-35486617</td>\n",
       "      <td>-</td>\n",
       "      <td>1626.5325</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000639578.2</td>\n",
       "      <td>TEAD3</td>\n",
       "      <td>chr6:35486468-35486617</td>\n",
       "      <td>-</td>\n",
       "      <td>1626.5325</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000372067.7</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>chr6:43771097-43771246</td>\n",
       "      <td>+</td>\n",
       "      <td>-708.6328</td>\n",
       "      <td>180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000672860.2</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>chr6:43771097-43771246</td>\n",
       "      <td>+</td>\n",
       "      <td>-708.6328</td>\n",
       "      <td>180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000506618.4</td>\n",
       "      <td>PRPS1L1</td>\n",
       "      <td>chr7:18027582-18027731</td>\n",
       "      <td>-</td>\n",
       "      <td>2024.7065</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000359396.8</td>\n",
       "      <td>TRPV6</td>\n",
       "      <td>chr7:142885517-142885666</td>\n",
       "      <td>-</td>\n",
       "      <td>-1512.0518</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000265806.11</td>\n",
       "      <td>R3HCC1</td>\n",
       "      <td>chr8:23290029-23290178</td>\n",
       "      <td>+</td>\n",
       "      <td>-885.1802</td>\n",
       "      <td>187.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENST00000621592.7</td>\n",
       "      <td>MYC</td>\n",
       "      <td>chr8:127736591-127736623+chr8:127738248-127738262</td>\n",
       "      <td>+</td>\n",
       "      <td>-164.9417</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENST00000634734.2</td>\n",
       "      <td>BAG1</td>\n",
       "      <td>chr9:33264462-33264611</td>\n",
       "      <td>-</td>\n",
       "      <td>-1026.5729</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENST00000526148.5</td>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>chr11:10803983-10804050+chr11:10804137-1080420...</td>\n",
       "      <td>-</td>\n",
       "      <td>522.5676</td>\n",
       "      <td>206.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENST00000525681.5</td>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>chr11:10803983-10804050+chr11:10804137-1080420...</td>\n",
       "      <td>-</td>\n",
       "      <td>522.5676</td>\n",
       "      <td>206.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENST00000339995.10</td>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>chr11:10803983-10804050+chr11:10804137-1080420...</td>\n",
       "      <td>-</td>\n",
       "      <td>522.5676</td>\n",
       "      <td>206.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENST00000396525.6</td>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>chr11:10803983-10804050+chr11:10804137-1080420...</td>\n",
       "      <td>-</td>\n",
       "      <td>522.5676</td>\n",
       "      <td>206.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENST00000452863.9</td>\n",
       "      <td>WT1</td>\n",
       "      <td>chr11:32435142-32435291</td>\n",
       "      <td>-</td>\n",
       "      <td>-877.2413</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENST00000332351.8</td>\n",
       "      <td>WT1</td>\n",
       "      <td>chr11:32435142-32435291</td>\n",
       "      <td>-</td>\n",
       "      <td>-877.2413</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ENST00000448076.8</td>\n",
       "      <td>WT1</td>\n",
       "      <td>chr11:32435142-32435291</td>\n",
       "      <td>-</td>\n",
       "      <td>-877.2413</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ENST00000358409.6</td>\n",
       "      <td>TEAD4</td>\n",
       "      <td>chr12:2994836-2994985</td>\n",
       "      <td>+</td>\n",
       "      <td>2394.7639</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ENST00000359864.7</td>\n",
       "      <td>TEAD4</td>\n",
       "      <td>chr12:2994836-2994985</td>\n",
       "      <td>+</td>\n",
       "      <td>2394.7639</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>ENST00000381140.9</td>\n",
       "      <td>GTF3A</td>\n",
       "      <td>chr13:27432798-27432804+chr13:27434139-2743421...</td>\n",
       "      <td>+</td>\n",
       "      <td>1635.4710</td>\n",
       "      <td>235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ENST00000566435.3</td>\n",
       "      <td>NPW</td>\n",
       "      <td>chr16:2019908-2020057</td>\n",
       "      <td>+</td>\n",
       "      <td>-111.5428</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>ENST00000333751.6</td>\n",
       "      <td>KCTD11</td>\n",
       "      <td>chr17:7352793-7352942</td>\n",
       "      <td>+</td>\n",
       "      <td>109.0243</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>ENST00000641389.1</td>\n",
       "      <td>NFKBID</td>\n",
       "      <td>chr19:35897657-35897806</td>\n",
       "      <td>-</td>\n",
       "      <td>-397.7879</td>\n",
       "      <td>142.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ENST00000270458.2</td>\n",
       "      <td>CACNG8</td>\n",
       "      <td>chr19:53963095-53963244</td>\n",
       "      <td>+</td>\n",
       "      <td>1223.7617</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>ENST00000375862.6</td>\n",
       "      <td>HCK</td>\n",
       "      <td>chr20:32052338-32052486+chr20:32071662-32071662</td>\n",
       "      <td>+</td>\n",
       "      <td>-1389.3606</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ENST00000375852.3</td>\n",
       "      <td>HCK</td>\n",
       "      <td>chr20:32052338-32052486+chr20:32071662-32071662</td>\n",
       "      <td>+</td>\n",
       "      <td>-1389.3606</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ENST00000403230.2</td>\n",
       "      <td>DDX17</td>\n",
       "      <td>chr22:38506001-38506150</td>\n",
       "      <td>-</td>\n",
       "      <td>421.9530</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENST00000396821.7</td>\n",
       "      <td>DDX17</td>\n",
       "      <td>chr22:38506001-38506150</td>\n",
       "      <td>-</td>\n",
       "      <td>421.9530</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENST00000672916.1</td>\n",
       "      <td>YPEL2</td>\n",
       "      <td>chr17:59353311-59353409</td>\n",
       "      <td>+</td>\n",
       "      <td>521.8397</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENST00000672036.1</td>\n",
       "      <td>YPEL1</td>\n",
       "      <td>chr22:21710745-21710849</td>\n",
       "      <td>-</td>\n",
       "      <td>-756.5811</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tr_id     gene  \\\n",
       "0    ENST00000373471.8    FNDC5   \n",
       "2    ENST00000479764.6     OAZ3   \n",
       "3    ENST00000400999.6     OAZ3   \n",
       "4    ENST00000305943.8   RNF187   \n",
       "5    ENST00000418194.6      SP3   \n",
       "6    ENST00000652005.1      SP3   \n",
       "7    ENST00000393716.6    NR1I2   \n",
       "8    ENST00000466380.5    NR1I2   \n",
       "10  ENST00000338863.12    TEAD3   \n",
       "11   ENST00000639578.2    TEAD3   \n",
       "12   ENST00000372067.7    VEGFA   \n",
       "13   ENST00000672860.2    VEGFA   \n",
       "14   ENST00000506618.4  PRPS1L1   \n",
       "15   ENST00000359396.8    TRPV6   \n",
       "18  ENST00000265806.11   R3HCC1   \n",
       "24   ENST00000621592.7      MYC   \n",
       "25   ENST00000634734.2     BAG1   \n",
       "26   ENST00000526148.5   EIF4G2   \n",
       "27   ENST00000525681.5   EIF4G2   \n",
       "28  ENST00000339995.10   EIF4G2   \n",
       "29   ENST00000396525.6   EIF4G2   \n",
       "35   ENST00000452863.9      WT1   \n",
       "37   ENST00000332351.8      WT1   \n",
       "38   ENST00000448076.8      WT1   \n",
       "39   ENST00000358409.6    TEAD4   \n",
       "40   ENST00000359864.7    TEAD4   \n",
       "43   ENST00000381140.9    GTF3A   \n",
       "47   ENST00000566435.3      NPW   \n",
       "48   ENST00000333751.6   KCTD11   \n",
       "49   ENST00000641389.1   NFKBID   \n",
       "51   ENST00000270458.2   CACNG8   \n",
       "52   ENST00000375862.6      HCK   \n",
       "54   ENST00000375852.3      HCK   \n",
       "56   ENST00000403230.2    DDX17   \n",
       "57   ENST00000396821.7    DDX17   \n",
       "58   ENST00000672916.1    YPEL2   \n",
       "59   ENST00000672036.1    YPEL1   \n",
       "\n",
       "                               global_coo_50_and_less strand  \\\n",
       "0   chr1:32868374-32868388+chr1:32868882-32868997+...      -   \n",
       "2                            chr1:151766716-151766865      +   \n",
       "3                            chr1:151766716-151766865      +   \n",
       "4                            chr1:228487666-228487815      +   \n",
       "5                            chr2:173955657-173955806      -   \n",
       "6                            chr2:173955657-173955806      -   \n",
       "7                            chr3:119807266-119807415      +   \n",
       "8                            chr3:119807266-119807415      +   \n",
       "10                             chr6:35486468-35486617      -   \n",
       "11                             chr6:35486468-35486617      -   \n",
       "12                             chr6:43771097-43771246      +   \n",
       "13                             chr6:43771097-43771246      +   \n",
       "14                             chr7:18027582-18027731      -   \n",
       "15                           chr7:142885517-142885666      -   \n",
       "18                             chr8:23290029-23290178      +   \n",
       "24  chr8:127736591-127736623+chr8:127738248-127738262      +   \n",
       "25                             chr9:33264462-33264611      -   \n",
       "26  chr11:10803983-10804050+chr11:10804137-1080420...      -   \n",
       "27  chr11:10803983-10804050+chr11:10804137-1080420...      -   \n",
       "28  chr11:10803983-10804050+chr11:10804137-1080420...      -   \n",
       "29  chr11:10803983-10804050+chr11:10804137-1080420...      -   \n",
       "35                            chr11:32435142-32435291      -   \n",
       "37                            chr11:32435142-32435291      -   \n",
       "38                            chr11:32435142-32435291      -   \n",
       "39                              chr12:2994836-2994985      +   \n",
       "40                              chr12:2994836-2994985      +   \n",
       "43  chr13:27432798-27432804+chr13:27434139-2743421...      +   \n",
       "47                              chr16:2019908-2020057      +   \n",
       "48                              chr17:7352793-7352942      +   \n",
       "49                            chr19:35897657-35897806      -   \n",
       "51                            chr19:53963095-53963244      +   \n",
       "52    chr20:32052338-32052486+chr20:32071662-32071662      +   \n",
       "54    chr20:32052338-32052486+chr20:32071662-32071662      +   \n",
       "56                            chr22:38506001-38506150      -   \n",
       "57                            chr22:38506001-38506150      -   \n",
       "58                            chr17:59353311-59353409      +   \n",
       "59                            chr22:21710745-21710849      -   \n",
       "\n",
       "    PhyloCSF120score  len_of_real_ext_in_codons  \n",
       "0          2051.0659                       75.0  \n",
       "2          -139.7141                       48.0  \n",
       "3          -139.7141                       48.0  \n",
       "4          -742.6165                      109.0  \n",
       "5           350.1512                      217.0  \n",
       "6           350.1512                      217.0  \n",
       "7          -625.4103                       55.0  \n",
       "8          -625.4103                       55.0  \n",
       "10         1626.5325                       65.0  \n",
       "11         1626.5325                       65.0  \n",
       "12         -708.6328                      180.0  \n",
       "13         -708.6328                      180.0  \n",
       "14         2024.7065                       67.0  \n",
       "15        -1512.0518                       40.0  \n",
       "18         -885.1802                      187.0  \n",
       "24         -164.9417                       15.0  \n",
       "25        -1026.5729                       71.0  \n",
       "26          522.5676                      206.0  \n",
       "27          522.5676                      206.0  \n",
       "28          522.5676                      206.0  \n",
       "29          522.5676                      206.0  \n",
       "35         -877.2413                       73.0  \n",
       "37         -877.2413                       73.0  \n",
       "38         -877.2413                       73.0  \n",
       "39         2394.7639                       73.0  \n",
       "40         2394.7639                       73.0  \n",
       "43         1635.4710                      235.0  \n",
       "47         -111.5428                       52.0  \n",
       "48          109.0243                       39.0  \n",
       "49         -397.7879                      142.0  \n",
       "51         1223.7617                       34.0  \n",
       "52        -1389.3606                       21.0  \n",
       "54        -1389.3606                       21.0  \n",
       "56          421.9530                       79.0  \n",
       "57          421.9530                       79.0  \n",
       "58          521.8397                       33.0  \n",
       "59         -756.5811                       35.0  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nte_local_global_PCSF_df[(nte_local_global_PCSF_df['tr_id'].isin(ann_ext_g35.tr_id.tolist()))][['tr_id', 'gene', 'global_coo_50_and_less', \n",
    "                                                                                      'strand', 'PhyloCSF120score', 'len_of_real_ext_in_codons']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remaining genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    }
   ],
   "source": [
    "remaining_genes = nte_local_global_PCSF_df[(~nte_local_global_PCSF_df['gene'].isin(list(nte_local_global_PCSF_df[nte_local_global_PCSF_df['PhyloCSF120score'] > 0].gene.unique()))) &\n",
    "                                          (nte_local_global_df['tr_id'].isin(ann_ext_g35.tr_id.tolist()))]\n",
    "\n",
    "\n",
    "print (remaining_genes.gene.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene, tr_id TRPV6 ENST00000359396.8\n",
      "coo:  chr7:142885517-142885666 -\n",
      "\n",
      "real ext len 40.0\n",
      "\n",
      "-1512.0518\n",
      "\n",
      "theor.nte-seq GAGTCCTGGCTGGCTCTGCCAAGTGTAACAAACTCACAGCCCTCTCCAAACTGGCTGGGGCTGCTGGGAGACTCCCAAGGAACTCGTCAGGAAGGCAGGAGACAGGAGACGGGACCTCTACAGGGAGACGGTGGGCCGGCCCTTGGGGGGGCTGATGTGGCCCCAAGGCTGAGTCCCGTCAGGGTCTGGCCTCGGCCTCAGGCCCCCAAGGAGCCGGCCCTACACCCC\n",
      "\n",
      "ext pred in paper\n",
      "ACGGGACCTCTACAGGGAGACGGTGGGCCGGCCCTTGGGGGGGCTGATGTGGCCCCAAGGCTGAGTCCCGTCAGGGTCTGGCCTCGGCCTCAGGCCCCCAAGGAGCCGGCCCTACACCCC\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "#remaining_genes[['tr_id', 'gene', 'global_coo_50_and_less',\n",
    "                 #'strand', 'PhyloCSF120score', 'len_of_real_ext_in_codons']]\n",
    "    \n",
    "i = 7\n",
    "\n",
    "print ('gene, tr_id', remaining_genes.iloc[i].gene, remaining_genes.iloc[i].tr_id)\n",
    "\n",
    "print ('coo: ', remaining_genes.iloc[i].global_coo_50_and_less, remaining_genes.iloc[i].strand)\n",
    "\n",
    "print ()\n",
    "\n",
    "print ('real ext len', remaining_genes.iloc[i].len_of_real_ext_in_codons)\n",
    "\n",
    "print ()\n",
    "\n",
    "print (remaining_genes.iloc[i].PhyloCSF120score)\n",
    "\n",
    "print ()\n",
    "\n",
    "print ('theor.nte-seq', nte_local_global_PCSF_df[nte_local_global_PCSF_df['tr_id'] == remaining_genes.iloc[i].tr_id].iloc[0].new_NTE_seq)\n",
    "\n",
    "print ()\n",
    "\n",
    "print ('ext pred in paper')\n",
    "print (EXT_SEQ_DF[EXT_SEQ_DF['gene'] == remaining_genes.iloc[i].gene].iloc[0].ext_seq[:-3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>strand</th>\n",
       "      <th>PhyloCSF120score</th>\n",
       "      <th>len_of_real_ext_in_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000479764.6</td>\n",
       "      <td>OAZ3</td>\n",
       "      <td>chr1:151766716-151766865</td>\n",
       "      <td>+</td>\n",
       "      <td>-139.7141</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000400999.6</td>\n",
       "      <td>OAZ3</td>\n",
       "      <td>chr1:151766716-151766865</td>\n",
       "      <td>+</td>\n",
       "      <td>-139.7141</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000305943.8</td>\n",
       "      <td>RNF187</td>\n",
       "      <td>chr1:228487666-228487815</td>\n",
       "      <td>+</td>\n",
       "      <td>-742.6165</td>\n",
       "      <td>109.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000393716.6</td>\n",
       "      <td>NR1I2</td>\n",
       "      <td>chr3:119807266-119807415</td>\n",
       "      <td>+</td>\n",
       "      <td>-625.4103</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000466380.5</td>\n",
       "      <td>NR1I2</td>\n",
       "      <td>chr3:119807266-119807415</td>\n",
       "      <td>+</td>\n",
       "      <td>-625.4103</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000372067.7</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>chr6:43771097-43771246</td>\n",
       "      <td>+</td>\n",
       "      <td>-708.6328</td>\n",
       "      <td>180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000672860.2</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>chr6:43771097-43771246</td>\n",
       "      <td>+</td>\n",
       "      <td>-708.6328</td>\n",
       "      <td>180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000359396.8</td>\n",
       "      <td>TRPV6</td>\n",
       "      <td>chr7:142885517-142885666</td>\n",
       "      <td>-</td>\n",
       "      <td>-1512.0518</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000265806.11</td>\n",
       "      <td>R3HCC1</td>\n",
       "      <td>chr8:23290029-23290178</td>\n",
       "      <td>+</td>\n",
       "      <td>-885.1802</td>\n",
       "      <td>187.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENST00000621592.7</td>\n",
       "      <td>MYC</td>\n",
       "      <td>chr8:127736591-127736623+chr8:127738248-127738262</td>\n",
       "      <td>+</td>\n",
       "      <td>-164.9417</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENST00000634734.2</td>\n",
       "      <td>BAG1</td>\n",
       "      <td>chr9:33264462-33264611</td>\n",
       "      <td>-</td>\n",
       "      <td>-1026.5729</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENST00000452863.9</td>\n",
       "      <td>WT1</td>\n",
       "      <td>chr11:32435142-32435291</td>\n",
       "      <td>-</td>\n",
       "      <td>-877.2413</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENST00000332351.8</td>\n",
       "      <td>WT1</td>\n",
       "      <td>chr11:32435142-32435291</td>\n",
       "      <td>-</td>\n",
       "      <td>-877.2413</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ENST00000448076.8</td>\n",
       "      <td>WT1</td>\n",
       "      <td>chr11:32435142-32435291</td>\n",
       "      <td>-</td>\n",
       "      <td>-877.2413</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ENST00000566435.3</td>\n",
       "      <td>NPW</td>\n",
       "      <td>chr16:2019908-2020057</td>\n",
       "      <td>+</td>\n",
       "      <td>-111.5428</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>ENST00000641389.1</td>\n",
       "      <td>NFKBID</td>\n",
       "      <td>chr19:35897657-35897806</td>\n",
       "      <td>-</td>\n",
       "      <td>-397.7879</td>\n",
       "      <td>142.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>ENST00000375862.6</td>\n",
       "      <td>HCK</td>\n",
       "      <td>chr20:32052338-32052486+chr20:32071662-32071662</td>\n",
       "      <td>+</td>\n",
       "      <td>-1389.3606</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ENST00000375852.3</td>\n",
       "      <td>HCK</td>\n",
       "      <td>chr20:32052338-32052486+chr20:32071662-32071662</td>\n",
       "      <td>+</td>\n",
       "      <td>-1389.3606</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENST00000672036.1</td>\n",
       "      <td>YPEL1</td>\n",
       "      <td>chr22:21710745-21710849</td>\n",
       "      <td>-</td>\n",
       "      <td>-756.5811</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tr_id    gene  \\\n",
       "2    ENST00000479764.6    OAZ3   \n",
       "3    ENST00000400999.6    OAZ3   \n",
       "4    ENST00000305943.8  RNF187   \n",
       "7    ENST00000393716.6   NR1I2   \n",
       "8    ENST00000466380.5   NR1I2   \n",
       "12   ENST00000372067.7   VEGFA   \n",
       "13   ENST00000672860.2   VEGFA   \n",
       "15   ENST00000359396.8   TRPV6   \n",
       "18  ENST00000265806.11  R3HCC1   \n",
       "24   ENST00000621592.7     MYC   \n",
       "25   ENST00000634734.2    BAG1   \n",
       "35   ENST00000452863.9     WT1   \n",
       "37   ENST00000332351.8     WT1   \n",
       "38   ENST00000448076.8     WT1   \n",
       "47   ENST00000566435.3     NPW   \n",
       "49   ENST00000641389.1  NFKBID   \n",
       "52   ENST00000375862.6     HCK   \n",
       "54   ENST00000375852.3     HCK   \n",
       "59   ENST00000672036.1   YPEL1   \n",
       "\n",
       "                               global_coo_50_and_less strand  \\\n",
       "2                            chr1:151766716-151766865      +   \n",
       "3                            chr1:151766716-151766865      +   \n",
       "4                            chr1:228487666-228487815      +   \n",
       "7                            chr3:119807266-119807415      +   \n",
       "8                            chr3:119807266-119807415      +   \n",
       "12                             chr6:43771097-43771246      +   \n",
       "13                             chr6:43771097-43771246      +   \n",
       "15                           chr7:142885517-142885666      -   \n",
       "18                             chr8:23290029-23290178      +   \n",
       "24  chr8:127736591-127736623+chr8:127738248-127738262      +   \n",
       "25                             chr9:33264462-33264611      -   \n",
       "35                            chr11:32435142-32435291      -   \n",
       "37                            chr11:32435142-32435291      -   \n",
       "38                            chr11:32435142-32435291      -   \n",
       "47                              chr16:2019908-2020057      +   \n",
       "49                            chr19:35897657-35897806      -   \n",
       "52    chr20:32052338-32052486+chr20:32071662-32071662      +   \n",
       "54    chr20:32052338-32052486+chr20:32071662-32071662      +   \n",
       "59                            chr22:21710745-21710849      -   \n",
       "\n",
       "    PhyloCSF120score  len_of_real_ext_in_codons  \n",
       "2          -139.7141                       48.0  \n",
       "3          -139.7141                       48.0  \n",
       "4          -742.6165                      109.0  \n",
       "7          -625.4103                       55.0  \n",
       "8          -625.4103                       55.0  \n",
       "12         -708.6328                      180.0  \n",
       "13         -708.6328                      180.0  \n",
       "15        -1512.0518                       40.0  \n",
       "18         -885.1802                      187.0  \n",
       "24         -164.9417                       15.0  \n",
       "25        -1026.5729                       71.0  \n",
       "35         -877.2413                       73.0  \n",
       "37         -877.2413                       73.0  \n",
       "38         -877.2413                       73.0  \n",
       "47         -111.5428                       52.0  \n",
       "49         -397.7879                      142.0  \n",
       "52        -1389.3606                       21.0  \n",
       "54        -1389.3606                       21.0  \n",
       "59         -756.5811                       35.0  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remaining_genes[['tr_id', 'gene', 'global_coo_50_and_less', 'strand', 'PhyloCSF120score', 'len_of_real_ext_in_codons']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
